<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="pyVIA core" href="pyVIA%20Core.html" /><link rel="prev" title="Parameters and Attributes" href="../Parameters%20and%20Attributes.html" />

    <!-- Generated with Sphinx 5.1.1 and Furo 2023.03.27 -->
        <title>API - pyvia documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?digest=fad236701ea90a88636c2a8c73b44ae642ed2a53" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">pyvia  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">pyvia  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">General</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../pyVia-home.html">StaVia - Multi-Omic Single-Cell Cartography for Spatial and Temporal Atlases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Release%20History.html">Release History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Tutorial%20Video.html">VIA Tutorial Videos</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Parameters%20and%20Attributes.html">Parameters and Attributes</a></li>
<li class="toctree-l1 current has-children current-page"><a class="current reference internal" href="#">API</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="pyVIA%20Core.html">pyVIA core</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="Plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.animate_atlas.html">pyVIA.plotting_via.animate_atlas</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.animate_streamplot.html">pyVIA.plotting_via.animate_streamplot</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.get_gene_expression.html">pyVIA.plotting_via.get_gene_expression</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_atlas_view.html">pyVIA.plotting_via.plot_atlas_view</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_differentiation_flow.html">pyVIA.plotting_via.plot_differentiation_flow</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_gene_trend_heatmaps.html">pyVIA.plotting_via.plot_gene_trend_heatmaps</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_piechart_viagraph.html">pyVIA.plotting_via.plot_piechart_viagraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_population_composition.html">pyVIA.plotting_via.plot_population_composition</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_sc_lineage_probability.html">pyVIA.plotting_via.plot_sc_lineage_probability</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_scatter.html">pyVIA.plotting_via.plot_scatter</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_trajectory_curves.html">pyVIA.plotting_via.plot_trajectory_curves</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_viagraph.html">pyVIA.plotting_via.plot_viagraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_streamplot.html">pyVIA.plotting_via.via_streamplot</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="Utils.html">Util Functions for plotting</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.make_edgebundle_milestone.html">pyVIA.plotting_via.make_edgebundle_milestone</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_atlas_emb.html">pyVIA.plotting_via.via_atlas_emb</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_forcelayout.html">pyVIA.plotting_via.via_forcelayout</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_mds.html">pyVIA.plotting_via.via_mds</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="Datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.datasets_via/VIA.datasets_via.cell_cycle.html">VIA.datasets_via.cell_cycle</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.datasets_via/VIA.datasets_via.cell_cycle_cyto_data.html">VIA.datasets_via.cell_cycle_cyto_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.datasets_via/VIA.datasets_via.embryoid_body.html">VIA.datasets_via.embryoid_body</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.datasets_via/VIA.datasets_via.scATAC_hematopoiesis.html">VIA.datasets_via.scATAC_hematopoiesis</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.datasets_via/VIA.datasets_via.scRNA_hematopoiesis.html">VIA.datasets_via.scRNA_hematopoiesis</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.datasets_via/VIA.datasets_via.toy_disconnected.html">VIA.datasets_via.toy_disconnected</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/VIA.datasets_via/VIA.datasets_via.toy_multifurcating.html">VIA.datasets_via.toy_multifurcating</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../Basic%20Example%20Code.html">Examples for installation checking</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Via 2.0 Atlas View Gallery</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Atlas%20view%20examples.html">Via2.0 Atlas views</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mouse_to_pup_atlas.html">Large Mouse Embryo to Pup Developmental Atlas</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Via2%20Atlas%20Animation.html">Via 2.0 Atlas Animations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Via%20Atlas%20View%20for%20Spatial%20omics.html">Via2.0 Atlas views for Spatial Omics data</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Via 2.0 Cartography</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/Via2.0%20Cartographic%20Mouse%20Gastrualation.html">1. Via 2.0 Cartography on Mouse Gastrulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/Zebrahub%20TI%20tutorial.html">2. Via 2.0 Cartography on Zebrahub (Trajectory Inference)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/Zebrahub_tutorial_visualization.html">3. Via 2.0 Cartography on Zebrahub (Visualization)</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials Via</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/ViaJupyter_Toy_Multifurcating.html">1. Basic workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/ViaJupyter_Toy_Disconnected.html">2. Disconnected Trajectories</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/ViaJupyter_scRNA_Hematopoiesis.html">3. Via Human Hematopoiesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/Imaging%20Cytometry%20%28cell%20cycle%29.html">4. Imaging cytometry</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/mESC_timeseries.html">5. Using time-series metadata (mESC Cytof)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/ViaJupyter_Pancreas_RNAvelocity.html">6. Using RNA-velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/ViaJupyter_scRNAVelocity_hematopoiesis.html">7. Bone marrow with RNA-velocity</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="api">
<h1>API<a class="headerlink" href="#api" title="Permalink to this heading"></a></h1>
<p>Import pyvia as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pyVIA.core</span> <span class="k">as</span> <span class="nn">via</span>
</pre></div>
</div>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyVIA%20Core.html">pyVIA core</a></li>
<li class="toctree-l1"><a class="reference internal" href="Plotting.html">Plotting</a></li>
<li class="toctree-l1"><a class="reference internal" href="Utils.html">Util Functions for plotting</a></li>
<li class="toctree-l1"><a class="reference internal" href="Datasets.html">Datasets</a></li>
</ul>
</div>
<section id="full-api">
<h2>Full API<a class="headerlink" href="#full-api" title="Permalink to this heading"></a></h2>
<section id="pyvia-core">
<h3>pyVIA core<a class="headerlink" href="#pyvia-core" title="Permalink to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="VIA.core.VIA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">VIA.core.</span></span><span class="sig-name descname"><span class="pre">VIA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">true_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgepruning_clustering_resolution_local</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgepruning_clustering_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_all_local_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">too_big_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution_parameter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partition_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ModularityVP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">small_pop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jac_weighted_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">knn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_iter_leiden</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'l2'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_smallpop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">super_cluster_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">super_node_degree_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">super_terminal_cells</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_lazy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.99</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_teleport</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.99</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">root_user</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preserve_disconnected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">super_terminal_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_coarse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">csr_full_graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">csr_array_locally_pruned</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ig_full_graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_neighbor_array</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_distance_array</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">embedding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_annot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preserve_disconnected_after_pruning</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">secondary_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudotime_threshold_TS</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_graph_pruning</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">visual_cluster_graph_pruning</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neighboring_terminal_states_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_mcmc_simulations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">piegraph_arrow_head_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">piegraph_edgeweight_scalingfactor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_visual_outgoing_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">via_coarse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">velocity_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">velo_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgebundle_pruning</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">A_velo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CSM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgebundle_pruning_twice</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pca_loadings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_series</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_series_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">knn_sequential</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">knn_sequential_reverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t_diff_step</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">single_cell_transition_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">embedding_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'via-mds'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">do_compute_embedding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_defined_terminal_cell</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_defined_terminal_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">do_gaussian_kernel_edgeweights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RW2_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">working_dir_fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">viagraph_decay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_memory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/core.html#VIA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.core.VIA" title="Permalink to this definition"></a></dt>
<dd><p>A class to represent the VIA analysis</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>ndarray</em>) – input matrix of size n_cells x n_dims. Expects the PCs or features that will be used in the TI computation. Can be e.g. adata.obsm[‘X_pca][:,0:20]</p></li>
<li><p><strong>true_label</strong> (<em>list</em>) – list of str/int that correspond to the ground truth or reference annotations. Can also be None when no labels are available</p></li>
<li><p><strong>labels</strong> (<em>ndarray</em><em> (</em><em>nsamples</em><em>, </em><em>)</em>) – default is None. and PARC clusters are used for the viagraph. alternatively provide a list of clustermemberships that are integer values (not strings) to construct the viagraph using another clustering method or available annotations</p></li>
<li><p><strong>edgepruning_clustering_resolution_local</strong> (<em>float</em>) – default = 2
local level of pruning for PARC graph clustering stage. Range (0.1,3) higher numbers mean more edge retention. For large datasets can stick to just tuning edgepruning_clustering_resolution</p></li>
<li><p><strong>edgepruning_clustering_resolution</strong> (<em>float</em>) – (optional, default = 0.15, can also set as ‘median’) graph pruning for PARC clustering stage. Higher value keeps more edges, results in fewer clusters. Smaller value removes more edges and results in more clusters. Number of standard deviations below the network’s mean-jaccard-weighted edges. 0.1-1 provide reasonable pruning.
higher value means less pruning (more edges retained). e.g. a value of 0.15 means all edges that are above mean(edgeweight)-0.15*std(edge-weights) are retained.
We find both 0.15 and ‘median’ to yield good results/starting point and resulting in pruning away ~ 50-60% edges</p></li>
<li><p><strong>keep_all_local_dist</strong> (<em>bool</em><em>, </em><em>str</em>) – default value of ‘auto’ means that for smaller datasets local-pruning is done prior to clustering, but for large datasets local pruning is set to False for speed.
can also set to be bool of True or False</p></li>
<li><p><strong>too_big_factor</strong> (<em>float</em>) – (optional, default=0.4). Forces clusters &gt; 0.4*n_cells to be re-clustered</p></li>
<li><p><strong>resolution_parameter</strong> (<em>float</em>) – (default =1) larger value means more and smaller clusters</p></li>
<li><p><strong>partition_type</strong> (<em>str</em>) – (default “ModularityVP”) Options</p></li>
<li><p><strong>small_pop</strong> (<em>int</em>) – (default 10) Via attempts to merge Clusters with a population &lt; 10 cells with larger clusters. If you have a very small dataset (e.g. few hundred cells), then consider lowering to e.g. 5</p></li>
<li><p><strong>jac_weighted_edges</strong> (<em>bool</em>) – (default = True) Use weighted edges in the PARC clustering step</p></li>
<li><p><strong>knn</strong> (<em>int</em>) – (optional, default = 30) number of K-Nearest Neighbors for HNSWlib KNN graph. Larger knn means more graph connectivity. Lower knn means more loosely connected clusters/cells</p></li>
<li><p><strong>n_iter_leiden</strong> (<em>int</em>) – </p></li>
<li><p><strong>random_seed</strong> (<em>int</em>) – Random seed to pass to clustering</p></li>
<li><p><strong>num_threads</strong> – </p></li>
<li><p><strong>distance</strong> (<em>str</em>) – (default ‘l2’) Euclidean distance ‘l2’ by default; other options ‘ip’ and ‘cosine’ for graph construction and similarity</p></li>
<li><p><strong>visual_cluster_graph_pruning</strong> (<em>float</em>) – (optional, default = 0.15) This only comes into play if the user deliberately chooses not to use the default edge-bundling method of visualizating edges (draw_piechart_graph()) and instead calls draw_piechart_graph_nobundle().
It is often set to the same value as the PARC clustering level of edgepruning_clustering_resolution. This does not impact computation of terminal states, pseudotime or lineage likelihoods.
It controls the number of edges plotted for visual effect</p></li>
<li><p><strong>cluster_graph_pruning</strong> (<em>float</em>) – (optional, default =0.15) Pruning level of the cluster graph (does not impact number of clusters). Only impacts the connectivity of the clustergraph. Often set to the same value as the PARC clustering level of edgepruning_clustering_resolution.Reasonable range [0.1,1]
To retain more connectivity in the clustergraph underlying the trajectory computations, increase the value</p></li>
<li><p><strong>time_smallpop</strong> (<em>max time to be allowed handling singletons</em>) – </p></li>
<li><p><strong>x_lazy</strong> (<em>float</em>) – (default =0.95) 1-x = probability of staying in same node (lazy). Values between 0.9-0.99 are reasonable</p></li>
<li><p><strong>alpha_teleport</strong> (<em>float</em>) – (default = 0.99) 1-alpha is probability of jumping. Values between 0.95-0.99 are reasonable unless prior knowledge of teleportation</p></li>
<li><p><strong>root_user</strong> (<em>list</em><em>, </em><em>None</em>) – can be a list of strings, a list of int or None
(default is None) When the root_user is set as None and an RNA velocity matrix is available, a root will be automatically computed
if the root_user is None and not velocity matrix is provided, then an arbitrary root is selected
if the root_user is [‘celltype_earlystage’] where the str corresponds to an item in true_label, then a suitable starting point will be selected corresponding to this group
if the root_user is [678], where 678 is the index of the cell chosen as a start cell, then this will be the designated starting cell.
It is possible to give a list of root indices and groups. [120, 699] or [‘traj1_earlystage’, ‘traj2_earlystage’] when there are more than one trajectories</p></li>
<li><p><strong>preserve_disconnected</strong> (<em>bool</em>) – (default = True) If you believe there may be disconnected trajectories then set this to False</p></li>
<li><p><strong>dataset</strong> (<em>str</em>) – Can be set to ‘group’ or ‘’ (default). this refers to the type of root label (group level root or single cell index) you are going to provide.
if your true_label has a sensible group of cells for a root then you can set dataset to ‘group’ and make the root parameter [‘labelname_root_cell_type’]
if your root corresponds to one particular cell then set dataset = ‘’ (default)</p></li>
<li><p><strong>embedding</strong> (<em>ndarray</em>) – (optional, default = None) embedding (e.g. precomputed tsne, umap, phate, via-umap) for plotting data. Size n_cells x 2
If an embedding is provided when running VIA, then a scatterplot colored by pseudotime, highlighting terminal fates</p></li>
<li><p><strong>velo_weight</strong> (<em>float</em>) – (optional, default = 0.5) #float between [0,1]. the weight assigned to directionality and connectivity derived from scRNA-velocity</p></li>
<li><p><strong>neighboring_terminal_states_threshold</strong> (<em>int</em>) – (default = 3). Candidates for terminal states that are neighbors of each other may be removed from the list if they have this number of more of terminal states as neighbors</p></li>
<li><p><strong>knn_sequential</strong> (<em>int</em>) – (default =10) number of knn in the adjacent time-point for time-series data (t_i and t_i+1)</p></li>
<li><p><strong>knn_sequential_reverse</strong> (<em>int</em>) – (default = 0) number of knn enforced from current to previous time point</p></li>
<li><p><strong>t_diff_step</strong> (<em>int</em>) – (default =1) Number of permitted temporal intervals between connected nodes. If time data is labeled as [0,25,50,75,100,..]
then t_diff_step=1 corresponds to ‘25’ and only edges within t_diff_steps are retained</p></li>
<li><p><strong>is_coarse</strong> (<em>bool</em>) – (default = True) If running VIA in two iterations where you wish to link the second fine-grained iteration with the initial iteration, then you set to False</p></li>
<li><p><strong>via_coarse</strong> (<a class="reference internal" href="pyVIA%20Core.html#VIA.core.VIA" title="VIA.core.VIA"><em>VIA</em></a>) – (default = None) If instantiating a second iteration of VIA that needs to be linked to a previous iteration (e.g. via0), then set via_coarse to the previous via0 object</p></li>
<li><p><strong>df_annot</strong> (<em>DataFrame</em>) – (default None) used for the Mouse Organ data</p></li>
<li><p><strong>preserve_disconnected_after_pruning</strong> (<em>bool</em>) – (default = False) If you believe there are disconnected trajectories then set this to True and test your hypothesis</p></li>
<li><p><strong>A_velo</strong> (<em>ndarray</em>) – Cluster Graph Transition matrix based on rna velocity [n_clus x n_clus]</p></li>
<li><p><strong>velocity_matrix</strong> (<em>matrix</em>) – (default None) matrix of size [n_samples x n_genes]. this is the velocity matrix computed by scVelo (or similar package) and stored in adata.layers[‘velocity’]. The genes used for computing velocity should correspond to those useing in gene_matrix
Requires gene_matrix to be provided too.</p></li>
<li><p><strong>gene_matrix</strong> (<em>matrix</em>) – (default None) Only used if Velocity_matrix is available. matrix of size [n_samples x n_genes]. We recommend using a subset like HVGs rather than full set of genes. (need to densify input if taking from adata = adata.X.todense())</p></li>
<li><p><strong>time_series</strong> (<em>bool</em>) – (default False) if the data has time-series labels then set to True</p></li>
<li><p><strong>time_series_labels</strong> (<em>list</em>) – (default None) list of integer values of temporal annoataions corresponding to e.g. hours (post fert), days, or sequential ordering</p></li>
<li><p><strong>pca_loadings</strong> (<em>array</em>) – (default None) the loadings of the pcs used to project the cells (to projected euclidean location based on velocity). n_cells x n_pcs</p></li>
<li><p><strong>secondary_annotations</strong> (<em>None</em>) – (default None)</p></li>
<li><p><strong>edgebundle_pruning</strong> (<em>float</em>) – (default=None) will by default be set to the same as the cluster_graph_pruning and influences the visualized level of pruning of edges.
Typical values can be between [0,1] with higher numbers retaining more edges</p></li>
<li><p><strong>edgebundle_pruning_twice</strong> (<em>bool</em>) – <p>default: False. When True, the edgebundling is applied to a further visually pruned (visual_cluster_graph_pruning) and can sometimes simplify the visualization. it does not impact the pseudotime and lineage computations
piegraph_arrow_head_width: float</p>
<blockquote>
<div><p>(default = 0.1) size of arrow heads in via cluster graph</p>
</div></blockquote>
</p></li>
<li><p><strong>piegraph_edgeweight_scalingfactor</strong> – (defaulf = 1.5) scaling factor for edge thickness in via cluster graph</p></li>
<li><p><strong>max_visual_outgoing_edges</strong> (<em>int</em>) – (default =2) Only allows max_visual_outgoing_edges to come out of any given node. Used in differentiation_flow()</p></li>
<li><p><strong>edgebundle_pruning</strong> – (default=None) will by default be set to the same as the cluster_graph_pruning and influences the visualized level of pruning of edges.
Typical values can be between [0,1] with higher numbers retaining more edges</p></li>
<li><p><strong>edgebundle_pruning_twice</strong> – default: False. When True, the edgebundling is applied to a further visually pruned (visual_cluster_graph_pruning) and can sometimes simplify the visualization for very cluttered graphs. it does not impact the pseudotime and lineage computations</p></li>
<li><p><strong>pseudotime_threshold_TS</strong> (<em>int</em>) – (default = 30) corresponds to the criteria for a state to be considered a candidate terminal cell fate to be 30% or later of the computed psuedotime range</p></li>
<li><p><strong>num_mcmc_simulations</strong> (<em>int</em>) – (default = 1300) number of random walk simulations conducted</p></li>
<li><p><strong>embedding_type</strong> (<em>str</em>) – (default = ‘via-mds’, other options are ‘via-atlas’ and ‘via-force’</p></li>
<li><p><strong>do_compute_embedding</strong> (<em>bool</em>) – (default = False) If you want an embedding (n_samples x2) to be computed on the basis of the via sc graph then set this to True</p></li>
<li><p><strong>do_gaussian_kernel_edgeweights</strong> (<em>bool</em>) – (default = False) Type of edgeweighting on the graph edges</p></li>
<li><p><strong>memory</strong> (<em>1/q * edge weight to a next-node that is not a neighbor of previous node. larger number means more memory and more introspective walk. small number &lt;1 means more exploration</em>) – (default = 2) higher q means more memory, more retrospective/inwards randomwalk. memory = 2 means run using the non-memory Via 1.0 mode</p></li>
<li><p><strong>viagraph_decay</strong> (<em>float</em>) – (default = 0.9) increasing decay causes more edges to merge</p></li>
<li><p><strong>memory</strong> – </p></li>
<li><p><strong>p_memory</strong> (<em>1/p * edge weight to next node = previous node. large value means more exploration</em>) – </p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.labels">
<span class="sig-name descname"><span class="pre">labels</span></span><a class="headerlink" href="#VIA.core.VIA.labels" title="Permalink to this definition"></a></dt>
<dd><p>length (n_samples, ) of cluster labels ndarray pre determined cluster labels user defined. #np.asarray(pre_labels).flatten()</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.single_cell_pt_markov">
<span class="sig-name descname"><span class="pre">single_cell_pt_markov</span></span><a class="headerlink" href="#VIA.core.VIA.single_cell_pt_markov" title="Permalink to this definition"></a></dt>
<dd><p>length n_samples of pseudotime</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.single_cell_bp">
<span class="sig-name descname"><span class="pre">single_cell_bp</span></span><a class="headerlink" href="#VIA.core.VIA.single_cell_bp" title="Permalink to this definition"></a></dt>
<dd><p>[n_lineages x n_samples] array of single cell branching probabilities towards each lineage (lineage normalized).
Each column corresponds to a terminal state, in the order presented by the terminal_clusters attribute</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.single_cell_bp_rownormed">
<span class="sig-name descname"><span class="pre">single_cell_bp_rownormed</span></span><a class="headerlink" href="#VIA.core.VIA.single_cell_bp_rownormed" title="Permalink to this definition"></a></dt>
<dd><p>[n_lineages x n_samples] array of single cell branching probabilities towards each lineage (cell normalized).
Each column corresponds to a terminal state, in the order presented by the terminal_clusters attribute</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.terminal_clusters">
<span class="sig-name descname"><span class="pre">terminal_clusters</span></span><a class="headerlink" href="#VIA.core.VIA.terminal_clusters" title="Permalink to this definition"></a></dt>
<dd><p>list of clusters that are cell fates/ unique lineages</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.cluster_bp">
<span class="sig-name descname"><span class="pre">cluster_bp</span></span><a class="headerlink" href="#VIA.core.VIA.cluster_bp" title="Permalink to this definition"></a></dt>
<dd><p>[n_clusters x n_terminal_states]. Lineage probability of cluster towards a particular terminal cluster state</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.CSM">
<span class="sig-name descname"><span class="pre">CSM</span></span><a class="headerlink" href="#VIA.core.VIA.CSM" title="Permalink to this definition"></a></dt>
<dd><p>[n_cluster x n_clusters] array of cosine similarity used to weight the cluster graph transition matrix by velocity</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.single_cell_transition_matrix">
<span class="sig-name descname"><span class="pre">single_cell_transition_matrix</span></span><a class="headerlink" href="#VIA.core.VIA.single_cell_transition_matrix" title="Permalink to this definition"></a></dt>
<dd><p>[n_samples x n_samples]</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id0">
<span class="sig-name descname"><span class="pre">terminal_clusters</span></span><a class="headerlink" href="#id0" title="Permalink to this definition"></a></dt>
<dd><p>(default None) list of terminal clusters</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.csr_full_graph">
<span class="sig-name descname"><span class="pre">csr_full_graph</span></span><a class="headerlink" href="#VIA.core.VIA.csr_full_graph" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>csr matrix of single-cell graph (augmented with sequential data when providing time_series information)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.csr_array_locally_pruned">
<span class="sig-name descname"><span class="pre">csr_array_locally_pruned</span></span><a class="headerlink" href="#VIA.core.VIA.csr_array_locally_pruned" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>csr matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.ig_full_graph">
<span class="sig-name descname"><span class="pre">ig_full_graph</span></span><a class="headerlink" href="#VIA.core.VIA.ig_full_graph" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.full_neighbor_array">
<span class="sig-name descname"><span class="pre">full_neighbor_array</span></span><a class="headerlink" href="#VIA.core.VIA.full_neighbor_array" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.user_defined_terminal_cell">
<span class="sig-name descname"><span class="pre">user_defined_terminal_cell</span></span><a class="headerlink" href="#VIA.core.VIA.user_defined_terminal_cell" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list=[]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.user_defined_terminal_group">
<span class="sig-name descname"><span class="pre">user_defined_terminal_group</span></span><a class="headerlink" href="#VIA.core.VIA.user_defined_terminal_group" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list=[]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.n_milestones">
<span class="sig-name descname"><span class="pre">n_milestones</span></span><a class="headerlink" href="#VIA.core.VIA.n_milestones" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int = None Number of milestones in the via-mds computation (anything more than 10,000 can be computationally heavy and time consuming) Typically auto-determined within the via-mds function</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="VIA.core.VIA.embedding">
<span class="sig-name descname"><span class="pre">embedding</span></span><a class="headerlink" href="#VIA.core.VIA.embedding" title="Permalink to this definition"></a></dt>
<dd><p>[n_cells x 2] provided by user or autocomputed with via-mds or via-umap</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="VIA.core.VIA.sc_transition_matrix">
<span class="sig-name descname"><span class="pre">sc_transition_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">smooth_transition</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_sequentially_augmented</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/core.html#VIA.sc_transition_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.core.VIA.sc_transition_matrix" title="Permalink to this definition"></a></dt>
<dd><p>#computes the single cell level transition directions that are later used to calculate velocity of embedding
#based on changes at single cell level in genes and single cell level velocity</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>smooth_transition</strong> – </p></li>
<li><p><strong>b</strong> – slope of logistic function</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-VIA.plotting_via">
<span id="plotting"></span><h3>Plotting<a class="headerlink" href="#module-VIA.plotting_via" title="Permalink to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.animate_atlas">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">animate_atlas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hammerbundle_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">via_object</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth_bundle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_milestones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'white'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plasma_r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_title_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveto</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/user/Trajectory/Datasets/animation_default.gif'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_series_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_labels_numeric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#animate_atlas"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.animate_atlas" title="Permalink to this definition"></a></dt>
<dd><p>Plot animated atlas view visualising the growth of edgebudle and single cell embedding based on given time_series_labels.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> – axis to plot on</p></li>
<li><p><strong>hammer_bundle</strong> – hammerbundle object with coordinates of all the edges to draw</p></li>
<li><p><strong>layout</strong> – coords of cluster nodes and optionally also contains the numeric value associated with each cluster (such as time-stamp) layout[[‘x’,’y’,’numeric label’]] sc/cluster/milestone level</p></li>
<li><p><strong>CSM</strong> – cosine similarity matrix. cosine similarity between the RNA velocity between neighbors and the change in gene expression between these neighbors. Only used when available</p></li>
<li><p><strong>velocity_weight</strong> – percentage weightage given to the RNA velocity based transition matrix</p></li>
<li><p><strong>time_series_labels</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – if not given, by default it will be set to pseudotime</p></li>
<li><p><strong>alpha_bundle</strong> – alpha when drawing lines</p></li>
<li><p><strong>linewidth_bundle</strong> – linewidth of bundled lines</p></li>
<li><p><strong>edge_color</strong> – </p></li>
<li><p><strong>frame_interval</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – smaller number, faster refresh and video</p></li>
<li><p><strong>facecolor</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – default = white</p></li>
<li><p><strong>headwidth_bundle</strong> – headwidth of arrows used in bundled edges</p></li>
<li><p><strong>arrow_frequency</strong> – min dist between arrows (bundled edges otherwise have overcrowding of arrows)</p></li>
<li><p><strong>show_direction</strong> – True will draw arrows along the lines to indicate direction</p></li>
<li><p><strong>milestone_edges</strong> – pandas DataFrame milestone_edges[[‘source’,’target’]]</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>axis with bundled edges plotted</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.animate_streamplot">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">animate_streamplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">embedding</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">density_grid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mass</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff_perc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_edgewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth_transition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth_grid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_scheme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'annotation'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b_bias</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_neighbors_velocity_grid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_animate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_scatter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rainbow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_stream</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Blues'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">segment_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveto</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/shobi/Trajectory/Datasets/animation.gif'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_sequentially_augmented</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'white'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#animate_streamplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.animate_streamplot" title="Permalink to this definition"></a></dt>
<dd><p>Draw Animated vector plots. the Saved .gif file saved at the saveto address, is the best for viewing the animation as the fig, ax output can be slow</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – viaobject</p></li>
<li><p><strong>embedding</strong> – ndarray (nsamples,2) umap, tsne, via-umap, via-mds</p></li>
<li><p><strong>density_grid</strong> – </p></li>
<li><p><strong>linewidth</strong> – </p></li>
<li><p><strong>min_mass</strong> – </p></li>
<li><p><strong>cutoff_perc</strong> – </p></li>
<li><p><strong>scatter_size</strong> – </p></li>
<li><p><strong>scatter_alpha</strong> – </p></li>
<li><p><strong>marker_edgewidth</strong> – </p></li>
<li><p><strong>smooth_transition</strong> – </p></li>
<li><p><strong>smooth_grid</strong> – </p></li>
<li><p><strong>color_scheme</strong> – ‘annotation’, ‘cluster’, ‘other’</p></li>
<li><p><strong>add_outline_clusters</strong> – </p></li>
<li><p><strong>cluster_outline_edgewidth</strong> – </p></li>
<li><p><strong>gp_color</strong> – </p></li>
<li><p><strong>bg_color</strong> – </p></li>
<li><p><strong>title</strong> – </p></li>
<li><p><strong>b_bias</strong> – </p></li>
<li><p><strong>n_neighbors_velocity_grid</strong> – </p></li>
<li><p><strong>fontsize</strong> – </p></li>
<li><p><strong>alpha_animate</strong> – </p></li>
<li><p><strong>cmap_scatter</strong> – </p></li>
<li><p><strong>cmap_stream</strong> – string of a cmap for streamlines, default = ‘Blues’ (for dark blue lines) . Consider ‘Blues_r’ for white lines OR ‘Greys/_r’ ‘gist_yard/_r’</p></li>
<li><p><strong>color_stream</strong> – string like ‘white’. will override cmap_stream</p></li>
<li><p><strong>segment_length</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>fig, ax.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.get_gene_expression">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">get_gene_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_exp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'jet'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_splines</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spline_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_lineages</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optional_title_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#get_gene_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.get_gene_expression" title="Permalink to this definition"></a></dt>
<dd><p>Line graph of gene expression vs pseudotime along lineages.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – via object</p></li>
<li><p><strong>gene_exp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – dataframe where columns are features (gene) and rows are single cells</p></li>
<li><p><strong>cmap</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – default: ‘jet’</p></li>
<li><p><strong>dpi</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – default:150</p></li>
<li><p><strong>marker_genes</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – Default is to use all genes in gene_exp. other provide a list of marker genes that will be used from gene_exp.</p></li>
<li><p><strong>linewidth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – default:2</p></li>
<li><p><strong>n_slines</strong> – default:10 Note n_splines must be &gt; spline_order.</p></li>
<li><p><strong>spline_order</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – default:4 n_splines must be &gt; spline_order.</p></li>
<li><p><strong>marker_lineages</strong> – Default is to use all lineage pathways. other provide a list of lineage number (terminal cluster number).</p></li>
<li><p><strong>cmap_dict</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code>]) – {lineage number: ‘color’}</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>fig, axs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.make_edgebundle_milestone">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">make_edgebundle_milestone</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">embedding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">via_object</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_pt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_bandwidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.03</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_milestones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">milestone_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_labels_numeric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weighted</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_visual_pruning</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">terminal_cluster_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">single_cell_lineage_prob</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#make_edgebundle_milestone"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.make_edgebundle_milestone" title="Permalink to this definition"></a></dt>
<dd><p>Perform Edgebundling of edges in a milestone level to return a hammer bundle of milestone-level edges. This is more granular than the original parc-clusters but less granular than single-cell level and hence also less computationally expensive
requires some type of embedding (n_samples x 2) to be available</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>embedding</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – optional (not required if via_object is provided) embedding single cell. also looks nice when done on via_mds as more streamlined continuous diffused graph structure. Umap is a but “clustery”</p></li>
<li><p><strong>graph</strong> – optional (not required if via_object is provided) igraph single cell graph level</p></li>
<li><p><strong>via_object</strong> – via_object (best way to run this function by simply providing via_object)</p></li>
<li><p><strong>sc_graph</strong> – igraph graph set as the via attribute self.ig_full_graph (affinity graph)</p></li>
<li><p><strong>initial_bandwidth</strong> – increasing bw increases merging of minor edges</p></li>
<li><p><strong>decay</strong> – increasing decay increases merging of minor edges #https://datashader.org/user_guide/Networks.html</p></li>
<li><p><strong>milestone_labels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – default list=[]. Usually autocomputed. but can provide as single-cell level labels (clusters, groups, which function as milestone groupings of the single cells)</p></li>
<li><p><strong>sc_labels_numeric</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – default is None which automatically chooses via_object’s pseudotime or time_series_labels (when available). otherwise set to a list of numerical values representing some sequential/chronological information</p></li>
<li><p><strong>terminal_cluster_list</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – default list [] and automatically uses all terminal clusters. otherwise set to any of the terminal cluster numbers within a list</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dictionary containing keys: hb_dict[‘hammerbundle’] = hb hammerbundle class with hb.x and hb.y containing the coords
hb_dict[‘milestone_embedding’] dataframe with ‘x’ and ‘y’ columns for each milestone and hb_dict[‘edges’] dataframe with columns [‘source’,’target’] milestone for each each and [‘cluster_pop’], hb_dict[‘sc_milestone_labels’] is a list of milestone label for each single cell</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_atlas_view">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_atlas_view</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hammerbundle_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">via_object</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_bundle_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth_bundle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'white'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plasma'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_title_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_milestones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">headwidth_bundle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">headwidth_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arrow_frequency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_arrow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_labels_sequential</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_labels_expression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_bandwidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.03</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_milestones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_scatter_size_pop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_milestones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lineage_pathway</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_visual_pruning</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_sc_labels_sequential_for_direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_scatter_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_scatter_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_sc_embedding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_milestones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colorbar_legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pseudotime'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_atlas_view"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_atlas_view" title="Permalink to this definition"></a></dt>
<dd><p>Draws atlas view. Integrated visualisation of edgebundle and single cell embedding.</p>
<p>Edges can be colored by time-series numeric labels, pseudotime, lineage pathway probabilities,  or gene expression. If not specificed then time-series is chosen if available, otherwise falls back to pseudotime. to use gene expression the sc_labels_expression is provided as a list.
To specify other numeric sequential data provide a list of sc_labels_sequential = [] n_samples in length. via_object.embedding must be an ndarray of shape (nsamples,2)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hammer_bundle_dict</strong> – dictionary with keys: hammerbundle object with coordinates of all the edges to draw. If hammer_bundle and layout are None, then this will be computed internally</p></li>
<li><p><strong>via_object</strong> – type via object, if hammerbundle_dict is None, then you must provide a via_object. Ensure that via_object has embedding attribute</p></li>
<li><p><strong>layout</strong> – coords of cluster nodes and optionally also contains the numeric value associated with each cluster (such as time-stamp) layout[[‘x’,’y’,’numeric label’]] sc/cluster/milestone level</p></li>
<li><p><strong>CSM</strong> – cosine similarity matrix. cosine similarity between the RNA velocity between neighbors and the change in gene expression between these neighbors. Only used when available</p></li>
<li><p><strong>velocity_weight</strong> – percentage weightage given to the RNA velocity based transition matrix</p></li>
<li><p><strong>pt</strong> – cluster-level pseudotime</p></li>
<li><p><strong>alpha_bundle</strong> – alpha when drawing lines</p></li>
<li><p><strong>linewidth_bundle</strong> – linewidth of bundled lines</p></li>
<li><p><strong>edge_color</strong> – </p></li>
<li><p><strong>alpha_milestones</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – float 0.3 alpha of milestones</p></li>
<li><p><strong>size_milestones</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – scatter size of the milestones (use sc_size_scatter to control single cell scatter when using in conjunction with lineage probs/ sc embeddings)</p></li>
<li><p><strong>arrow_frequency</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – min dist between arrows (bundled edges otherwise have overcrowding of arrows)</p></li>
<li><p><strong>show_direction</strong> – True will draw arrows along the lines to indicate direction</p></li>
<li><p><strong>milestone_edges</strong> – pandas DataFrame milestoone_edges[[‘source’,’target’]]</p></li>
<li><p><strong>milestone_numeric_values</strong> – the milestone average of numeric values such as time (days, hours), location (position), or other numeric value used for coloring edges in a sequential manner
if this is None then the edges are colored by length to distinguish short and long range edges</p></li>
<li><p><strong>arrow_frequency</strong> – 0.05. higher means fewer arrows</p></li>
<li><p><strong>n_milestones</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – int  None. if no hammerbundle_dict is provided, but via_object is provided, then the user can specify level of granularity by setting the n_milestones. otherwise it will be automatically selected</p></li>
<li><p><strong>scale_scatter_size_pop</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – bool default False</p></li>
<li><p><strong>sc_labels_expression</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – list single cell numeric values used for coloring edges and nodes of corresponding milestones mean expression levels (len n_single_cell samples)
edges can be colored by time-series numeric labels, pseudotime, or gene expression. If not specificed then time-series is chosen if available, otherwise falls back to pseudotime. to use gene expression the sc_labels_expression is provided as a list</p></li>
<li><p><strong>sc_labels_sequential</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – list single cell numeric sequential values used for directionality inference as replacement for  pseudotime or via_object.time_series_labels (len n_samples single cell)</p></li>
<li><p><strong>sc_labels</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – list None list of single-cell level labels (categorial or discrete set of numerical values) to label the nodes</p></li>
<li><p><strong>text_labels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – bool False if you want to label the nodes based on sc_labels (or true_label if via_object is provided)</p></li>
<li><p><strong>lineage_pathway</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – list of terminal states to plot lineage pathways</p></li>
<li><p><strong>use_sc_labels_sequential_for_direction</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – use the sequential data (timeseries labels or other provided by user) to direct the arrows</p></li>
</ul>
</dd>
</dl>
<p>:param lineage_alpha_threshold number representing the percentile (0-100) of lineage likelikhood in a particular lineage pathway, below which edges will be drawn with lower alpha transparency factor
:type sc_scatter_alpha: <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>
:param sc_scatter_alpha: transparency of the background singlecell scatter when plotting lineages
:type add_sc_embedding: <code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>
:param add_sc_embedding: add background of single cell scatter plot for Atlas
:param scatter_size_sc_embedding
:param colorbar_legend str title of colorbar
:return: fig, axis with bundled edges plotted</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_differentiation_flow">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_differentiation_flow</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_lineages</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_node</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">do_log_flow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">majority_cluster_population_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_sankey</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rainbow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title_str</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Differentiation</span> <span class="pre">Flow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">root_cluster_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_differentiation_flow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_differentiation_flow" title="Permalink to this definition"></a></dt>
<dd><p>Draws sankey plots visualising lineage and cell fate.</p>
<p>#SANKEY PLOTS
G is the igraph knn (low K) used for shortest path in high dim space. no idx needed as it’s made on full sample
knn_hnsw is the knn made in the embedded space used for query to find the nearest point in the downsampled embedding
that corresponds to the single cells in the full graph</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – </p></li>
<li><p><strong>embedding</strong> – n_samples x 2. embedding is 2D representation of the full dataset.</p></li>
<li><p><strong>idx</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – if one uses a downsampled embedding of the original data, then idx is the selected indices of the downsampled samples used in the visualization</p></li>
<li><p><strong>cmap_name</strong> – </p></li>
<li><p><strong>dpi</strong> – </p></li>
</ul>
</dd>
</dl>
<p>:param do_log_flow bool True (default) take the natural log (1+edge flow value)
:param label_node list of labels for each cell (could be cell type, stage level) length is n_cells
:param scatter_size: if None, then auto determined based on n_cells
:param marker_lineages: Default is to use all lineage pathways. other provide a list of lineage number (terminal cluster number).
:type alpha_factor: <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>
:param alpha_factor: float transparency
:type root_cluster_list: <code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]
:param root_cluster_list: list of roots by cluster number e.g. [5] means a good root is cluster number 5
:return: fig, axs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_gene_trend_heatmaps">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_gene_trend_heatmaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_gene_exp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_lineages</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ytick_labelrotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_gene_trend_heatmaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_gene_trend_heatmaps" title="Permalink to this definition"></a></dt>
<dd><p>Plot the gene trends on heatmap: a heatmap is generated for each lineage (identified by terminal cluster number). Default selects all lineages</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – </p></li>
<li><p><strong>df_gene_exp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>) – pandas DataFrame single-cell level expression [cells x genes]</p></li>
<li><p><strong>marker_lineages</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – list default = None and plots all detected all lineages. Optionally provide a list of integers corresponding to the cluster number of terminal cell fates</p></li>
<li><p><strong>fontsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – int default = 8</p></li>
<li><p><strong>cmap</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – str default = ‘viridis’</p></li>
<li><p><strong>normalize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – bool = True</p></li>
<li><p><strong>ytick_labelrotation</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – int default = 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>fig and list of axes</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_piechart_viagraph">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_piechart_viagraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_exp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_piechart</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rainbow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">headwidth_arrow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_edge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth_edge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'darkblue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pie_size_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pt_visual_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">99</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highlight_terminal_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_node_notpiechart</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_piechart_viagraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_piechart_viagraph" title="Permalink to this definition"></a></dt>
<dd><p>Plot two subplots with a clustergraph level representation of the viagraph showing true-label composition (lhs) and pseudotime/gene expression (rhs)</p>
<p>Returns matplotlib figure with two axes that plot the clustergraph using edge bundling
left axis shows the clustergraph with each node colored by annotated ground truth membership.
right axis shows the same clustergraph with each node colored by the pseudotime or gene expression</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – is class VIA (the same function also exists as a method of the class and an external plotting function</p></li>
<li><p><strong>type_data</strong> – string  default ‘pt’ for pseudotime colored nodes. or ‘gene’</p></li>
<li><p><strong>gene_exp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – list of values (or column of dataframe) corresponding to feature or gene expression to be used to color nodes at CLUSTER level</p></li>
<li><p><strong>cmap_piechart</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – str cmap for piechart categories</p></li>
<li><p><strong>title</strong> – string</p></li>
<li><p><strong>cmap</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – default None. automatically chooses coolwarm for gene expression or viridis_r for pseudotime</p></li>
<li><p><strong>ax_text</strong> – Bool default= True. Annotates each node with cluster number and population of membership</p></li>
<li><p><strong>dpi</strong> – int default = 150</p></li>
<li><p><strong>headwidth_arrow</strong> – default = 0.1. width of arrowhead used to directed edges</p></li>
<li><p><strong>reference_labels</strong> – None or list. list of categorical (str) labels for cluster composition of the piecharts (LHS subplot) length = n_samples.</p></li>
<li><p><strong>pie_size_scale</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – float default=0.8 scaling factor of the piechart nodes</p></li>
<li><p><strong>pt_visual_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – int (percentage) default = 95 corresponding to rescaling the visual color scale by clipping outlier cluster pseudotimes</p></li>
</ul>
</dd>
</dl>
<p>:param highlight_terminal_clusters:bool = True (red border around terminal clusters)
:type size_node_notpiechart: <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>
:param size_node_notpiechart: scaling factor for node size of the viagraph (not the piechart part)
:return: f, ax, ax1</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_population_composition">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_population_composition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">celltype_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rainbow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_intervals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'time'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Cell</span> <span class="pre">populations'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_population_composition"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_population_composition" title="Permalink to this definition"></a></dt>
<dd><p>Plot population compositions along known time stamps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – optional. this is required unless both time_labels and cell_labels are provided as arguments to the function</p></li>
<li><p><strong>time_labels</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – list length n_cells of pseudotime or known stage numeric labels</p></li>
<li><p><strong>cell_labels</strong> – list of cell type or cluster length n_cells</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>ax</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_sc_lineage_probability">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_sc_lineage_probability</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">embedding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'plasma'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_lineages</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">majority_cluster_population_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_sankey</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rainbow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">do_sankey</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_sc_lineage_probability"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_sc_lineage_probability" title="Permalink to this definition"></a></dt>
<dd><p>Plot single cell embedding labelled with lieage likelihood.</p>
<p>G is the igraph knn (low K) used for shortest path in high dim space. no idx needed as it’s made on full sample
knn_hnsw is the knn made in the embedded space used for query to find the nearest point in the downsampled embedding
that corresponds to the single cells in the full graph</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – </p></li>
<li><p><strong>embedding</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – n_samples x 2. embedding is either the full or downsampled 2D representation of the full dataset.</p></li>
<li><p><strong>idx</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – if one uses a downsampled embedding of the original data, then idx is the selected indices of the downsampled samples used in the visualization</p></li>
<li><p><strong>cmap_name</strong> – </p></li>
<li><p><strong>dpi</strong> – </p></li>
<li><p><strong>scatter_size</strong> – if None, then auto determined based on n_cells</p></li>
<li><p><strong>marker_lineages</strong> – Default is to use all lineage pathways. other provide a list of lineage number (terminal cluster number).</p></li>
<li><p><strong>alpha_factor</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – float transparency</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>fig, axs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_scatter">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_scatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">embedding</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rainbow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'None'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">via_object</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_index_terminal_states</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">true_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_scatter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_scatter" title="Permalink to this definition"></a></dt>
<dd><p>General scatter plotting tool for numeric and categorical labels on the single-cell level</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>embedding</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>) – ndarray n_samples x 2</p></li>
<li><p><strong>labels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – list single cell labels list of number or strings</p></li>
<li><p><strong>cmap</strong> – str default = ‘rainbow’</p></li>
<li><p><strong>s</strong> – int size of scatter dot</p></li>
<li><p><strong>alpha</strong> – float with 0 transparent to 1 opaque default =0.3</p></li>
<li><p><strong>edgecolors</strong> – </p></li>
<li><p><strong>title</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – str</p></li>
<li><p><strong>text_labels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – bool default =True</p></li>
<li><p><strong>via_object</strong> – </p></li>
<li><p><strong>sc_index_terminal_states</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – list of integers corresponding to one cell in each of the terminal states</p></li>
<li><p><strong>color_dict</strong> – {‘true_label_group_1’: #COLOR,’true_label_group_2’: #COLOR2,….} where the dictionary keys correspond to the provided labels</p></li>
<li><p><strong>true_labels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – list of single cell labels used to annotate the terminal states</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>matplotlib pyplot fig, ax</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_trajectory_curves">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_trajectory_curves</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">embedding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title_str</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Pseudotime'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">draw_all_curves</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arrow_width_scale_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_edgewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_pseudotime</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis_r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highlight_terminal_states</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_maxout_edgelist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_trajectory_curves"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_trajectory_curves" title="Permalink to this definition"></a></dt>
<dd><p>Projects the graph based coarse trajectory onto a umap/tsne embedding</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – via object</p></li>
<li><p><strong>embedding</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – 2d array [n_samples x 2] with x and y coordinates of all n_samples. Umap, tsne, pca OR use the via computed embedding via_object.embedding</p></li>
<li><p><strong>idx</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – default: None. Or List. if you had previously computed a umap/tsne (embedding) only on a subset of the total n_samples (subsampled as per idx), then the via objects and results will be indexed according to idx too</p></li>
<li><p><strong>title_str</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – title of figure</p></li>
<li><p><strong>draw_all_curves</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – if the clustergraph has too many edges to project in a visually interpretable way, set this to False to get a simplified view of the graph pathways</p></li>
<li><p><strong>arrow_width_scale_factor</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>scatter_size</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>scatter_alpha</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>linewidth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>marker_edgewidth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>cmap_pseudotime</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – </p></li>
<li><p><strong>dpi</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – int default = 150. Use 300 for paper figures</p></li>
<li><p><strong>highlight_terminal_states</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – whether or not to highlight/distinguish the clusters which are detected as the terminal states by via</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>f, ax1, ax2</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_viagraph">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_viagraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arrow_head</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgeweight_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_factor_node</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_viagraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_viagraph" title="Permalink to this definition"></a></dt>
<dd><p>Cluster level expression of gene/feature intensity</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – </p></li>
<li><p><strong>type_data</strong> – </p></li>
<li><p><strong>gene_exp</strong> – pd.Dataframe size n_cells x genes. Otherwise defaults to plotting pseudotime</p></li>
<li><p><strong>gene_list</strong> – list of gene names corresponding to the column name</p></li>
<li><p><strong>arrow_head</strong> – </p></li>
<li><p><strong>edgeweight_scale</strong> – </p></li>
<li><p><strong>cmap</strong> – </p></li>
<li><p><strong>label_text</strong> – bool to add numeric values of the gene exp level</p></li>
</ul>
</dd>
</dl>
<p>:param size_factor_node size of graph nodes
:return: fig, axs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.plot_viagraph_">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">plot_viagraph_</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hammer_bundle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CSM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">velocity_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_bundle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth_bundle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'darkblue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">headwidth_bundle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arrow_frequency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">via_object</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#plot_viagraph_"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.plot_viagraph_" title="Permalink to this definition"></a></dt>
<dd><p>Plots the edgebundles on the via clustergraph level and also adds the relevant arrow directions based on the TI directionality</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> – axis to plot on</p></li>
<li><p><strong>hammer_bundle</strong> – hammerbundle object with coordinates of all the edges to draw. self.hammer</p></li>
<li><p><strong>layout</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – coords of cluster nodes</p></li>
<li><p><strong>CSM</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – cosine similarity matrix. cosine similarity between the RNA velocity between neighbors and the change in gene expression between these neighbors. Only used when available</p></li>
<li><p><strong>velocity_weight</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – percentage weightage given to the RNA velocity based transition matrix</p></li>
<li><p><strong>pt</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – cluster-level pseudotime (or other intensity level of features at average-cluster level)</p></li>
<li><p><strong>alpha_bundle</strong> – alpha when drawing lines</p></li>
<li><p><strong>linewidth_bundle</strong> – linewidth of bundled lines</p></li>
<li><p><strong>edge_color</strong> – </p></li>
<li><p><strong>headwidth_bundle</strong> – headwidth of arrows used in bundled edges</p></li>
<li><p><strong>arrow_frequency</strong> – min dist between arrows (bundled edges otherwise have overcrowding of arrows)</p></li>
<li><p><strong>show_direction</strong> – bool default True. will draw arrows along the lines to indicate direction</p></li>
<li><p><strong>plot_clusters</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – bool default False. When this function is called on its own (and not from within draw_piechart_graph() then via_object must be provided</p></li>
<li><p><strong>ax_text</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – bool default True. Show labels of the clusters with the cluster population and PARC cluster label</p></li>
<li><p><strong>fontsize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – float default 9 Font size of labels</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>fig, ax with bundled edges plotted</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.via_atlas_emb">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">via_atlas_emb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_input</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negative_sample_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gamma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spread</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'via'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_epochs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'euclidean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_membership</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveto</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#via_atlas_emb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.via_atlas_emb" title="Permalink to this definition"></a></dt>
<dd><p>Run dimensionality reduction using the StaVia modified HNSW graph using via cluster graph initialization when Via_object is provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – if via_object is provided then X_input and graph are ignored</p></li>
<li><p><strong>X_input</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – ndarray nsamples x features (PCs)</p></li>
<li><p><strong>graph</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">csr_matrix</span></code>]) – csr_matrix of knngraph. This usually is via’s pruned, sequentially augmented sc-knn graph accessed as an attribute of via via_object.csr_full_graph</p></li>
<li><p><strong>n_components</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>alpha</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>negative_sample_rate</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>gamma</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Weight to apply to negative samples.</p></li>
<li><p><strong>spread</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – The effective scale of embedded points. In combination with min_dist this determines how clustered/clumped the embedded points are.</p></li>
<li><p><strong>min_dist</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – The effective minimum distance between embedded points. Smaller values will result in a more clustered/clumped embedding where nearby points on the manifold are drawn closer together, while larger values will result on a more even dispersal of points</p></li>
<li><p><strong>init_pos</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – either a string (default) ‘via’ (uses via graph to initialize), or ‘spectral’. Or a n_cellx2 dimensional ndarray with initial coordinates</p></li>
<li><p><strong>random_state</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>n_epochs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The number of training epochs to be used in optimizing the low dimensional embedding. Larger values result in more accurate embeddings. If 0 is specified a value will be selected based on the size of the input dataset (200 for large datasets, 500 for small).</p></li>
<li><p><strong>distance_metric</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – </p></li>
<li><p><strong>layout</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – ndarray . custom initial layout. (n_cells x2). also requires cluster_membership labels</p></li>
<li><p><strong>cluster_membership</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – via_object.labels (cluster level labels of length n_samples corresponding to the layout)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray of shape (nsamples,n_components)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.via_forcelayout">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">via_forcelayout</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X_pca</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">viagraph_full</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_milestones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_series_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">knn_seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveto</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#via_forcelayout"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.via_forcelayout" title="Permalink to this definition"></a></dt>
<dd><p>Compute force directed layout. #TODO not complete</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X_pca</strong> – </p></li>
<li><p><strong>viagraph_full</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">csr_matrix</span></code>]) – optional. if calling before via, then None. if calling after or from within via, then we can use the via-graph to reinforce the layout</p></li>
<li><p><strong>k</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>random_seed</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>t_diffusion</strong> – </p></li>
<li><p><strong>n_milestones</strong> – </p></li>
<li><p><strong>time_series_labels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – </p></li>
<li><p><strong>knn_seq</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.via_mds">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">via_mds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_pca</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">viagraph_full</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">diffusion_op</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_milestones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_series_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">knn_seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k_project_milestones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t_difference</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveto</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">embedding_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mds'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double_diffusion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#via_mds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.via_mds" title="Permalink to this definition"></a></dt>
<dd><p>Fast computation of a 2D embedding based on previous StaVia analysis.</p>
<p>FOR EXAMPLE:
via_object.embedding = via.via_mds(via_object = v0)
plot_scatter(embedding = via_object.embedding, labels = via_object.true_labels)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – </p></li>
<li><p><strong>X_pca</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – dimension reduced (only if via_object is not passed)</p></li>
<li><p><strong>viagraph_full</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">csr_matrix</span></code>]) – optional. if calling before or without via, then None and a milestone graph will be computed. if calling after or from within via, then we can use the via-graph to reinforce the layout of the milestone graph</p></li>
<li><p><strong>k</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – number of knn for the via_mds reinforcement graph on milestones. default =15. integers 5-20 are reasonable</p></li>
<li><p><strong>random_seed</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – randomseed integer</p></li>
<li><p><strong>t_diffusion</strong> – default integer value = 1 with higher values generate more smoothing</p></li>
<li><p><strong>n_milestones</strong> – number of milestones used to generate the initial embedding</p></li>
<li><p><strong>time_series_labels</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – numerical values in list form representing some sequentual information</p></li>
<li><p><strong>knn_seq</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – if time-series data is available, this will augment the knn with sequential neighbors (2-10 are reasonable values) default =5</p></li>
<li><p><strong>embedding_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – default = ‘mds’ or set to ‘umap’</p></li>
<li><p><strong>double_diffusion</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – default is False. To achieve sharper strokes/lineages, set to True</p></li>
<li><p><strong>k_project_milestones</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – number of milestones in the milestone-knngraph used to compute the single-cell projection</p></li>
<li><p><strong>n_iterations</strong> – number of iterations to run</p></li>
<li><p><strong>neighbors_distances</strong> – array of distances of each neighbor for each cell (n_cells x knn) used when called from within via.run() for autocompute via-mds</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy array of size n_samples x 2</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.plotting_via.via_streamplot">
<span class="sig-prename descclassname"><span class="pre">VIA.plotting_via.</span></span><span class="sig-name descname"><span class="pre">via_streamplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">via_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">embedding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">density_grid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arrow_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arrow_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'k'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arrow_style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-|&gt;'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mass</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff_perc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_edgewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">density_stream</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth_transition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth_grid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_scheme</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'annotation'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_outline_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_outline_edgewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gp_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'white'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bg_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Streamplot'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b_bias</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_neighbors_velocity_grid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_sequentially_augmented</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rainbow'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_text_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/plotting_via.html#via_streamplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.plotting_via.via_streamplot" title="Permalink to this definition"></a></dt>
<dd><p>Construct vector streamplot on the embedding to show a fine-grained view of inferred directions in the trajectory</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>via_object</strong> – </p></li>
<li><p><strong>embedding</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – np.ndarray of shape (n_samples, 2) umap or other 2-d embedding on which to project the directionality of cells</p></li>
<li><p><strong>density_grid</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>arrow_size</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>arrow_color</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – </p></li>
<li><p><strong>arrow_style</strong> – </p></li>
<li><p><strong>max_length</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>linewidth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – width of  lines in streamplot, default = 1</p></li>
<li><p><strong>min_mass</strong> – </p></li>
<li><p><strong>cutoff_perc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>scatter_size</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – size of scatter points default =500</p></li>
<li><p><strong>scatter_alpha</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – transpsarency of scatter points</p></li>
<li><p><strong>marker_edgewidth</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – width of outline arround each scatter point, default = 0.1</p></li>
<li><p><strong>density_stream</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>smooth_transition</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>smooth_grid</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>color_scheme</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – str, default = ‘annotation’ corresponds to self.true_labels. Other options are ‘time’ (uses single-cell pseudotime) and ‘cluster’ (via cluster graph) and ‘other’. Alternatively provide labels as a list</p></li>
<li><p><strong>add_outline_clusters</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – </p></li>
<li><p><strong>cluster_outline_edgewidth</strong> – </p></li>
<li><p><strong>gp_color</strong> – </p></li>
<li><p><strong>bg_color</strong> – </p></li>
<li><p><strong>dpi</strong> – </p></li>
<li><p><strong>title</strong> – </p></li>
<li><p><strong>b_bias</strong> – default = 20. higher value makes the forward bias of pseudotime stronger</p></li>
<li><p><strong>n_neighbors_velocity_grid</strong> – </p></li>
<li><p><strong>labels</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>]) – list (will be used for the color scheme) or if a color_dict is provided these labels should match</p></li>
<li><p><strong>use_sequentially_augmented</strong> – </p></li>
<li><p><strong>cmap</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>fig, ax</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-VIA.datasets_via">
<span id="datasets"></span><h3>Datasets<a class="headerlink" href="#module-VIA.datasets_via" title="Permalink to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="VIA.datasets_via.cell_cycle">
<span class="sig-prename descclassname"><span class="pre">VIA.datasets_via.</span></span><span class="sig-name descname"><span class="pre">cell_cycle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">foldername</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/datasets_via.html#cell_cycle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.datasets_via.cell_cycle" title="Permalink to this definition"></a></dt>
<dd><p>Load cell cycle data as AnnData object</p>
<dl class="simple">
<dt>Args:</dt><dd><p>foldername (string): Directory of dataset</p>
</dd>
<dt>Returns:</dt><dd><p>AnnData object</p>
</dd>
</dl>
<img alt="https://github.com/ShobiStassen/VIA/blob/master/Figures/mb231_overall_300dpi.png?raw=true:width=&quot;200px&quot;" src="https://github.com/ShobiStassen/VIA/blob/master/Figures/mb231_overall_300dpi.png?raw=true:width=&quot;200px&quot;" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.datasets_via.cell_cycle_cyto_data">
<span class="sig-prename descclassname"><span class="pre">VIA.datasets_via.</span></span><span class="sig-name descname"><span class="pre">cell_cycle_cyto_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">foldername</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/datasets_via.html#cell_cycle_cyto_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.datasets_via.cell_cycle_cyto_data" title="Permalink to this definition"></a></dt>
<dd><p>Load cell cycle imagine based flow-cyto features</p>
<p>AnnData object with n_obs × n_vars = 2036 × 38
obs: ‘cell_cycle_phase’
:param foldername (string) Default current directory. path to directory where you want to store the dataset</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.datasets_via.embryoid_body">
<span class="sig-prename descclassname"><span class="pre">VIA.datasets_via.</span></span><span class="sig-name descname"><span class="pre">embryoid_body</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">foldername</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/datasets_via.html#embryoid_body"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.datasets_via.embryoid_body" title="Permalink to this definition"></a></dt>
<dd><p>Load embryoid body data as AnnData object</p>
<dl class="simple">
<dt>Args:</dt><dd><p>foldername (string): Directory to save dataset</p>
</dd>
<dt>Returns:</dt><dd><p>AnnData object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.datasets_via.scATAC_hematopoiesis">
<span class="sig-prename descclassname"><span class="pre">VIA.datasets_via.</span></span><span class="sig-name descname"><span class="pre">scATAC_hematopoiesis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">foldername</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/datasets_via.html#scATAC_hematopoiesis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.datasets_via.scATAC_hematopoiesis" title="Permalink to this definition"></a></dt>
<dd><p>Load scATAC seq Hematopoiesis data as AnnData object</p>
<dl class="simple">
<dt>Args:</dt><dd><p>foldername (string): Directory of dataset</p>
</dd>
<dt>Returns:</dt><dd><p>AnnData object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.datasets_via.scRNA_hematopoiesis">
<span class="sig-prename descclassname"><span class="pre">VIA.datasets_via.</span></span><span class="sig-name descname"><span class="pre">scRNA_hematopoiesis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">foldername</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/datasets_via.html#scRNA_hematopoiesis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.datasets_via.scRNA_hematopoiesis" title="Permalink to this definition"></a></dt>
<dd><p>Load scRNA seq Hematopoiesis data as AnnData object</p>
<dl class="simple">
<dt>Args:</dt><dd><p>foldername (string): Directory of dataset</p>
</dd>
<dt>Returns:</dt><dd><p>AnnData object</p>
</dd>
</dl>
<img alt="https://github.com/ShobiStassen/VIA/blob/master/Figures/humancd34_streamplot.png?raw=true:width=&quot;200px&quot;" src="https://github.com/ShobiStassen/VIA/blob/master/Figures/humancd34_streamplot.png?raw=true:width=&quot;200px&quot;" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.datasets_via.toy_disconnected">
<span class="sig-prename descclassname"><span class="pre">VIA.datasets_via.</span></span><span class="sig-name descname"><span class="pre">toy_disconnected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">foldername</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/datasets_via.html#toy_disconnected"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.datasets_via.toy_disconnected" title="Permalink to this definition"></a></dt>
<dd><p>Load Toy_Disconnected data as AnnData object</p>
<p>To access obs (label) as list, use AnnData.obs[‘group_id’].values.tolist()</p>
<dl class="simple">
<dt>Args:</dt><dd><p>foldername (string): Default current directory. path to directory where you want to store the dataset</p>
</dd>
<dt>Returns:</dt><dd><p>AnnData object</p>
</dd>
</dl>
<img alt="https://github.com/ShobiStassen/VIA/blob/master/Figures/stream_plot_toy4.png?raw=true:width=&quot;200px&quot;" src="https://github.com/ShobiStassen/VIA/blob/master/Figures/stream_plot_toy4.png?raw=true:width=&quot;200px&quot;" />
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="VIA.datasets_via.toy_multifurcating">
<span class="sig-prename descclassname"><span class="pre">VIA.datasets_via.</span></span><span class="sig-name descname"><span class="pre">toy_multifurcating</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">foldername</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/VIA/datasets_via.html#toy_multifurcating"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#VIA.datasets_via.toy_multifurcating" title="Permalink to this definition"></a></dt>
<dd><p>Load Toy_Multifurcating data as AnnData object</p>
<p>To access obs (label) as list, use AnnData.obs[‘group_id’].values.tolist()</p>
<dl class="simple">
<dt>Args:</dt><dd><p>foldername (string): foldername (string): path to directory where you want to store the dataset ‘./’ current directory is default</p>
</dd>
<dt>Returns:</dt><dd><p>AnnData object</p>
</dd>
</dl>
<img alt="https://github.com/ShobiStassen/VIA/blob/master/Figures/toy3_streamvia.png?raw=true:width=&quot;200px&quot;" src="https://github.com/ShobiStassen/VIA/blob/master/Figures/toy3_streamvia.png?raw=true:width=&quot;200px&quot;" />
</dd></dl>

</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="pyVIA%20Core.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">pyVIA core</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../Parameters%20and%20Attributes.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Parameters and Attributes</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2022, shobana stassen
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">API</a><ul>
<li><a class="reference internal" href="#full-api">Full API</a><ul>
<li><a class="reference internal" href="#pyvia-core">pyVIA core</a></li>
<li><a class="reference internal" href="#module-VIA.plotting_via">Plotting</a></li>
<li><a class="reference internal" href="#module-VIA.datasets_via">Datasets</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/scripts/furo.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script defer="defer" src="https://unpkg.com/@popperjs/core@2"></script>
    <script defer="defer" src="https://unpkg.com/tippy.js@6"></script>
    <script defer="defer" src="../_static/tippy\api\index.047a68f7-3b52-4b2c-9c73-32037b13ce5c.js"></script>
    </body>
</html>