<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" />

    <!-- Generated with Sphinx 5.1.1 and Furo 2023.03.27 -->
        <title>1. Basic workflow - pyvia documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?digest=fad236701ea90a88636c2a8c73b44ae642ed2a53" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">pyvia  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">pyvia  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">General</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../pyVia-home.html">StaVia - Multi-Omic Single-Cell Cartography for Spatial and Temporal Atlases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Release%20History.html">Release History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Tutorial%20Video.html">VIA Tutorial Videos</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Parameters%20and%20Attributes.html">Parameters and Attributes</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../api/index.html">API</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../api/pyVIA%20Core.html">pyVIA core</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/Plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.animate_atlas.html">pyVIA.plotting_via.animate_atlas</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.animate_streamplot.html">pyVIA.plotting_via.animate_streamplot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.get_gene_expression.html">pyVIA.plotting_via.get_gene_expression</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_atlas_view.html">pyVIA.plotting_via.plot_atlas_view</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_differentiation_flow.html">pyVIA.plotting_via.plot_differentiation_flow</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_gene_trend_heatmaps.html">pyVIA.plotting_via.plot_gene_trend_heatmaps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_piechart_viagraph.html">pyVIA.plotting_via.plot_piechart_viagraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_population_composition.html">pyVIA.plotting_via.plot_population_composition</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_sc_lineage_probability.html">pyVIA.plotting_via.plot_sc_lineage_probability</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_scatter.html">pyVIA.plotting_via.plot_scatter</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_trajectory_curves.html">pyVIA.plotting_via.plot_trajectory_curves</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_viagraph.html">pyVIA.plotting_via.plot_viagraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_streamplot.html">pyVIA.plotting_via.via_streamplot</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/Utils.html">Util Functions for plotting</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.make_edgebundle_milestone.html">pyVIA.plotting_via.make_edgebundle_milestone</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_atlas_emb.html">pyVIA.plotting_via.via_atlas_emb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_forcelayout.html">pyVIA.plotting_via.via_forcelayout</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_mds.html">pyVIA.plotting_via.via_mds</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/Datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.cell_cycle.html">VIA.datasets_via.cell_cycle</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.cell_cycle_cyto_data.html">VIA.datasets_via.cell_cycle_cyto_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.embryoid_body.html">VIA.datasets_via.embryoid_body</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.scATAC_hematopoiesis.html">VIA.datasets_via.scATAC_hematopoiesis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.scRNA_hematopoiesis.html">VIA.datasets_via.scRNA_hematopoiesis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.toy_disconnected.html">VIA.datasets_via.toy_disconnected</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.toy_multifurcating.html">VIA.datasets_via.toy_multifurcating</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../Basic%20Example%20Code.html">Examples for installation checking</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Via 2.0 Atlas View Gallery</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Atlas%20view%20examples.html">Via2.0 Atlas views</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mouse_to_pup_atlas.html">Large Mouse Embryo to Pup Developmental Atlas</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Via2%20Atlas%20Animation.html">Via 2.0 Atlas Animations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Via%20Atlas%20View%20for%20Spatial%20omics.html">Via2.0 Atlas views for Spatial Omics data</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Via 2.0 Cartography</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Via2.0%20Cartographic%20Mouse%20Gastrualation.html">1. Via 2.0 Cartography on Mouse Gastrulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Zebrahub%20TI%20tutorial.html">2. Via 2.0 Cartography on Zebrahub (Trajectory Inference)</a></li>
<li class="toctree-l1"><a class="reference internal" href="Zebrahub_tutorial_visualization.html">3. Via 2.0 Cartography on Zebrahub (Visualization)</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials Via</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_Toy_Multifurcating.html">1. Basic workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_Toy_Disconnected.html">2. Disconnected Trajectories</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_scRNA_Hematopoiesis.html">3. Via Human Hematopoiesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="Imaging%20Cytometry%20%28cell%20cycle%29.html">4. Imaging cytometry</a></li>
<li class="toctree-l1"><a class="reference internal" href="mESC_timeseries.html">5. Using time-series metadata (mESC Cytof)</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_Pancreas_RNAvelocity.html">6. Using RNA-velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_scRNAVelocity_hematopoiesis.html">7. Bone marrow with RNA-velocity</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section class="tex2jax_ignore mathjax_ignore" id="basic-workflow">
<h1>1. Basic workflow<a class="headerlink" href="#basic-workflow" title="Permalink to this heading"></a></h1>
<p>This vignette shows the basic workflow of Via and how to access the different visualizations after running the computations. The dataset is a toy multifurcation.
We start by importing modules.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pyVIA.core</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span> <span class="nn">pyVIA.datasets_via</span> <span class="k">as</span> <span class="nn">datasets_via</span>
<span class="c1">#from core_working_ import *</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span> 
<span class="kn">import</span> <span class="nn">umap</span> 
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/shobi/anaconda3/envs/Via2Env/lib/python3.7/site-packages/phate/__init__.py
</pre></div>
</div>
</div>
</div>
<p>First load all the data. Below we use UMAP to create a 2D embedding which is later used for visualization. Umap can be substituted with tsne/phate/PCA etc. VIA can also automatically compute an embedding (‘via-mds’ or ‘via-umap’) which is optionally generated when running VIA.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_counts</span> <span class="o">=</span> <span class="n">datasets_via</span><span class="o">.</span><span class="n">toy_multifurcating</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_counts</span><span class="p">)</span>
<span class="n">true_label</span> <span class="o">=</span> <span class="n">adata_counts</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;group_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">ncomps</span> <span class="o">=</span> <span class="mi">30</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_counts</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="n">ncomps</span><span class="p">)</span>

<span class="n">embedding</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">adata_counts</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/shobi/anaconda3/envs/Via2Env/lib/python3.7/site-packages/anndata/_core/anndata.py:120: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 1000 × 1000
    obs: &#39;group_id&#39;, &#39;true_time&#39;
inside umap_ fuzzy simplicial n_neig 15 /home/shobi/anaconda3/envs/Via2Env/lib/python3.7/site-packages/umap/umap_
knn indices size 15000 (1000, 15) knn dists shape (1000, 15)
local connectivity 1.0
n_neighbors 15
sigmas 1000
rhos 1000
coo matrix after membership strength 14000
before fuzyy symmetry result.data.size, max, mean 14000 1.0 0.2790636
applying symmetry in umap_
size of symmetry only 19066
after fuzzy symm result.data.size, max, mean 19066 1.0 0.3331308
graph.data.size 19066 0.333 1.0
graph.data.size 19058 19058
</pre></div>
</div>
</div>
</div>
<section id="expected-outputs">
<h2>Expected Outputs<a class="headerlink" href="#expected-outputs" title="Permalink to this heading"></a></h2>
<ol class="arabic simple">
<li><p>Cluster graph level trajectory highlighting cell type composition, pseudotime and start/end states</p></li>
<li><p>Trajectory drawn onto UMAP embedding.</p></li>
<li><p>Lineage pathways depicting lineage probabilities towards each of the detected terminal states</p></li>
<li><p>Gene trends for each lineage for marker genes</p></li>
<li><p>Vector fields for fine-grained trajectory</p></li>
<li><p>Cluster graph trajectory colored by gene expression level</p></li>
</ol>
</section>
<section id="parameters">
<h2>Parameters<a class="headerlink" href="#parameters" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">knn</span></code> number of nearest neighbors for the KNN-graph</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">true_label</span></code>: list of annotations length nsamples. if these are unavailable then leave as default None</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">clustergraph_pruning</span></code>: typically 0-3, the number of standard deviations for clustergraph pruning. level of pruning applied to the clustergraph carried forward for pseudotime and lineage probability calculations. Larger means less pruning</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">jac_std_global</span></code>: Controls granularity of clustering. range 0-1 is reasonable. values closer to 0 will result in more and smaller clusters</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">too_big_factor</span></code>: default 0.3. Clusters that are bigger than 30% of the entire cell population will be re-clustered. Values 0.1-0.3 work well</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">preserve_disconnected</span></code>: default = True. Attempts to reconnect fragmentation incurred by pruning. if you suspect disconnected trajectories, set this to False</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">dataset</span></code> = <code class="docutils literal notranslate"><span class="pre">(''</span> <span class="pre">or</span> <span class="pre">'group')</span></code> and is set based on the type of <code class="docutils literal notranslate"><span class="pre">root_user</span></code> parameter provided : <code class="docutils literal notranslate"><span class="pre">'group'</span></code> works for any dataset where the root is provided as a group ID in the true_label. e.g. if root_user = [‘HSC’], then the root will be selected as a cluster which has root like properties and is largely made up of ‘HSC’ cells. <code class="docutils literal notranslate"><span class="pre">''</span></code> is suitable when <code class="docutils literal notranslate"><span class="pre">'root_user'</span></code> is a single-cell index or None</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">root_user</span></code>: default None. Can be provided at the group or single-cell index level. e.g. [‘GroupX’] corresponding to a cell type in true_label or [1326] corresponding to a single cell index. If left as None, then in the case where velocity_matrix is available then a shortlist of 3 roots will be suggested. When root_user is None and velocity_matrix is None, then a root is chosen at random</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">edgebundle_pruning_twice</span></code>: default = False. When True, can sometimes further simplify the visualization if the edges are too numerous/cluttered. it does not impact the pseudotime and lineage computations</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#define parameters</span>
<span class="n">ncomps</span><span class="p">,</span> <span class="n">knn</span><span class="p">,</span> <span class="n">random_seed</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">root_user</span>  <span class="o">=</span><span class="mi">30</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span> <span class="mi">42</span><span class="p">,</span><span class="s1">&#39;toy&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;M1&#39;</span><span class="p">]</span>

<span class="n">v0</span> <span class="o">=</span> <span class="n">VIA</span><span class="p">(</span><span class="n">adata_counts</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="n">ncomps</span><span class="p">],</span> <span class="n">true_label</span><span class="p">,</span> <span class="n">jac_std_global</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">dist_std_local</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
             <span class="n">knn</span><span class="o">=</span><span class="n">knn</span><span class="p">,</span> <span class="n">cluster_graph_pruning_std</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">too_big_factor</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">root_user</span><span class="o">=</span><span class="n">root_user</span><span class="p">,</span> <span class="n">preserve_disconnected</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span>
             <span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">)</span><span class="c1">#, piegraph_arrow_head_width=0.2,             piegraph_edgeweight_scalingfactor=1.0)  </span>
<span class="n">v0</span><span class="o">.</span><span class="n">run_VIA</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-06 14:59:46.714859	Running VIA over input data of 1000 (samples) x 30 (features)
2023-01-06 14:59:46.715028	Knngraph has 20 neighbors
2023-01-06 14:59:46.992497	Finished global pruning of 20-knn graph used for clustering at level of 0.15. Kept 46.7 % of edges. 
2023-01-06 14:59:46.997641	Number of connected components used for clustergraph  is 1
2023-01-06 14:59:47.024562	Commencing community detection
2023-01-06 14:59:47.061412	Finished running Leiden algorithm. Found 43 clusters.
2023-01-06 14:59:47.062116	Merging 30 very small clusters (&lt;10)
2023-01-06 14:59:47.063432	Finished detecting communities. Found 13 communities
2023-01-06 14:59:47.063631	Making cluster graph. Global cluster graph pruning level: 1
2023-01-06 14:59:47.067875	Graph has 1 connected components before pruning
2023-01-06 14:59:47.069733	Graph has 1 connected components before pruning
2023-01-06 14:59:47.070323	0.0% links trimmed from local pruning relative to start
2023-01-06 14:59:47.071805	Starting make edgebundle viagraph...
2023-01-06 14:59:47.071841	Make via clustergraph edgebundle
2023-01-06 14:59:48.597443	Hammer dims: Nodes shape: (13, 2) Edges shape: (26, 3)
2023-01-06 14:59:48.603146	New root is 3 and majority M1
2023-01-06 14:59:48.603528	New root is 6 and majority M1
2023-01-06 14:59:48.603985	Computing lazy-teleporting expected hitting times
2023-01-06 14:59:49.110905	Identifying terminal clusters corresponding to unique lineages...
2023-01-06 14:59:49.111047	Closeness:[3, 4, 6, 7, 8, 10]
2023-01-06 14:59:49.111075	Betweenness:[3, 4, 6, 7, 8, 9, 10]
2023-01-06 14:59:49.111095	Out Degree:[3, 4, 6, 7, 8, 9]
2023-01-06 14:59:49.112261	Terminal clusters corresponding to unique lineages in this component are [4, 7, 8, 9, 10] 
2023-01-06 14:59:49.331527	From root 6,  the Terminal state 4 is reached 75 times.
2023-01-06 14:59:49.521343	From root 6,  the Terminal state 7 is reached 426 times.
2023-01-06 14:59:49.731108	From root 6,  the Terminal state 8 is reached 249 times.
2023-01-06 14:59:49.929187	From root 6,  the Terminal state 9 is reached 235 times.
2023-01-06 14:59:50.126257	From root 6,  the Terminal state 10 is reached 394 times.
2023-01-06 14:59:50.145610	Terminal clusters corresponding to unique lineages are {4: &#39;M6&#39;, 7: &#39;M2&#39;, 8: &#39;M4&#39;, 9: &#39;M5&#39;, 10: &#39;M8&#39;}
2023-01-06 14:59:50.145751	Begin projection of pseudotime and lineage likelihood
2023-01-06 14:59:50.281107	Graph has 1 connected components before pruning
2023-01-06 14:59:50.282593	Graph has 1 connected components before pruning
2023-01-06 14:59:50.283161	11.5% links trimmed from local pruning relative to start
2023-01-06 14:59:50.283212	50.0% links trimmed from global pruning relative to start
2023-01-06 14:59:50.287447	Start making edgebundle milestone...
2023-01-06 14:59:50.291364	Time elapsed 3.5 seconds
</pre></div>
</div>
</div>
</div>
</section>
<section id="via-graph">
<h2>VIA graph<a class="headerlink" href="#via-graph" title="Permalink to this heading"></a></h2>
<p>To visualize the results of the Trajectory inference in various ways. Via offers various plotting functions.We first show the cluster-graph level trajectory abstraction consisting of two subplots colored by annotated (true_label) composition and by pseudotime</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># draw the piechart clustergraph beside a clustergraph colored by the pseudotime (&#39;pt&#39;) or gene expression (&#39;gene&#39;).</span>
<span class="c1"># type_pt: &#39;pt&#39; automatically takes the cluster level pseudotimes. &#39;gene&#39; requires user to also pass gene_exp </span>
<span class="c1"># gene_exp =&#39;&#39;. In order to plot gene intensity, a list of cluster level gene/feature values should be provided as a list</span>
<span class="c1"># cmap = None. Can be specified and influences the cmap of the pt/gene graph</span>
<span class="c1"># ax_text = True. Can set to false if you don&#39;t want the cluster and populations labels </span>
<span class="c1"># other features like edge width, color, alpha can be set here too</span>
<span class="c1"># Terminal lineage clusters have red outline (in rhs graph)</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">ax2</span><span class="o">=</span> <span class="n">draw_piechart_graph</span><span class="p">(</span><span class="n">via0</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">type_data</span><span class="o">=</span><span class="s1">&#39;pt&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Toy multifurcation&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">ax_text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">gene_exp</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">alpha_edge</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">linewidth_edge</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">edge_color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">headwidth_arrow</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/0f87fc4500b1a513e1a4b795f8830759c8c4b6bd6498ddcbb0746205f026b1dd.png" src="../_images/0f87fc4500b1a513e1a4b795f8830759c8c4b6bd6498ddcbb0746205f026b1dd.png" />
</div>
</div>
</section>
<section id="scatter-plots">
<h2>Scatter Plots<a class="headerlink" href="#scatter-plots" title="Permalink to this heading"></a></h2>
<section id="colored-by-true-label">
<h3>colored by true_label<a class="headerlink" href="#colored-by-true-label" title="Permalink to this heading"></a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_scatter</span><span class="p">(</span><span class="n">embedding</span> <span class="o">=</span> <span class="n">embedding</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">v0</span><span class="o">.</span><span class="n">true_label</span><span class="p">)</span>
<span class="n">f</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_scatter</span><span class="p">(</span><span class="n">embedding</span> <span class="o">=</span> <span class="n">embedding</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">v0</span><span class="o">.</span><span class="n">single_cell_pt_markov</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>No handles with labels found to put in legend.
</pre></div>
</div>
<img alt="../_images/0407ace27eb42297e8fc3fb5229a1f3885db73554a2931de89dc9393b989234b.png" src="../_images/0407ace27eb42297e8fc3fb5229a1f3885db73554a2931de89dc9393b989234b.png" />
<img alt="../_images/56fe7cf1e2c63840a6bd547329eae9906af5c4c1f545e2a1397a3628668d2909.png" src="../_images/56fe7cf1e2c63840a6bd547329eae9906af5c4c1f545e2a1397a3628668d2909.png" />
</div>
</div>
</section>
</section>
<section id="fast-visualization-using-underlying-single-cell-via-graphs">
<h2>Fast visualization using underlying single-cell VIA-graphs<a class="headerlink" href="#fast-visualization-using-underlying-single-cell-via-graphs" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>VIA offers two embedding options generated together with the trajectory by combining the underlying trajectory graph with different manifold projections:</p>
<ul>
<li><p>via-mds</p></li>
<li><p>via-umap</p></li>
</ul>
</li>
<li><p>For automated computation (probably easiest): simply initialize VIA with parameters <code class="docutils literal notranslate"><span class="pre">do_compute_embedding</span> <span class="pre">=</span> <span class="pre">True</span></code>, and <code class="docutils literal notranslate"><span class="pre">embedding_type</span> <span class="pre">=</span> <span class="pre">'via-mds'</span> <span class="pre">or</span> <span class="pre">'via-umap'</span></code>.</p></li>
</ul>
<p>For more control after running VIA, simply call the <code class="docutils literal notranslate"><span class="pre">via_mds()</span></code> function. When using time-series data consider using via-mds and via-umap automatically for improved visualizations that are well integrated with the Trajectory.</p>
<section id="via-mds-parameters">
<h3>VIA-mds Parameters<a class="headerlink" href="#via-mds-parameters" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">n_milestones</span></code> (default 2000) larger number takes longer time and more RAM</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">double_diffusion</span></code> (default False) setting True can sharpen the “strokes”</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">t_diffusion</span></code> (default 1) higher values can incur more smoothing</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">time_series_labels</span></code> (list of numeric values, length n_samples representing some kind of sequential/temporal information, such as sampling times).</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">via_mds_embedding1</span> <span class="o">=</span> <span class="n">via_mds</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">)</span>
<span class="n">f</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_scatter</span><span class="p">(</span><span class="n">embedding</span><span class="o">=</span><span class="n">via_mds_embedding1</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">v0</span><span class="o">.</span><span class="n">true_label</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;via-mds&#39;</span><span class="p">)</span>

<span class="n">via_mds_embedding2</span> <span class="o">=</span> <span class="n">via_mds</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">double_diffusion</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">f</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_scatter</span><span class="p">(</span><span class="n">embedding</span><span class="o">=</span><span class="n">via_mds_embedding2</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">v0</span><span class="o">.</span><span class="n">true_label</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;via-mds double diffusion&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-06 15:00:06.517881	Commencing Via-MDS
2023-01-06 15:00:06.518114	Resetting n_milestones to 1000 as n_samples &gt; original n_milestones
2023-01-06 15:00:06.889333	Start computing with diffusion power:1
2023-01-06 15:00:06.904533	Starting MDS on milestone
2023-01-06 15:00:07.642632	End computing mds with diffusion power:1
2023-01-06 15:00:07.672491	Commencing Via-MDS
2023-01-06 15:00:07.672565	Resetting n_milestones to 1000 as n_samples &gt; original n_milestones
2023-01-06 15:00:08.026065	Start computing with diffusion power:1
2023-01-06 15:00:08.040905	Starting MDS on milestone
2023-01-06 15:00:08.714692	End computing mds with diffusion power:1
</pre></div>
</div>
<img alt="../_images/c01633474a87a589f66ac1eec0ece45721a7bbad51e09f8a186899f0eed51409.png" src="../_images/c01633474a87a589f66ac1eec0ece45721a7bbad51e09f8a186899f0eed51409.png" />
<img alt="../_images/cf27ebf74783fd7d17c753b12854194bb8b9f9b2462c81addfa3fb234ac7b2c4.png" src="../_images/cf27ebf74783fd7d17c753b12854194bb8b9f9b2462c81addfa3fb234ac7b2c4.png" />
</div>
</div>
</section>
</section>
<section id="via-umap-based-on-underlying-via-single-cell-knn-graphs">
<h2>VIA-umap based on underlying via single cell knn graphs<a class="headerlink" href="#via-umap-based-on-underlying-via-single-cell-knn-graphs" title="Permalink to this heading"></a></h2>
<p>This is encouraged especially when VIA is run with timeseries data <code class="docutils literal notranslate"><span class="pre">time_series_labels</span> <span class="pre">=</span> <span class="pre">[list</span> <span class="pre">of</span> <span class="pre">numeric</span> <span class="pre">values]</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">via_umap1</span> <span class="o">=</span> <span class="n">via_umap</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span><span class="n">min_dist</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">f</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_scatter</span><span class="p">(</span><span class="n">embedding</span><span class="o">=</span><span class="n">via_umap1</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">v0</span><span class="o">.</span><span class="n">true_label</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;via-umap&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-05 13:05:49.632704	Computing umap on sc-Viagraph
graph.data.size 19000 28.264 38.715
graph.data.size 19000 19000
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/shobi/anaconda3/envs/Via2Env/lib/python3.7/site-packages/umap/spectral.py:317: RuntimeWarning: divide by zero encountered in true_divide
  1.0 / np.sqrt(diag_data), 0, graph.shape[0], graph.shape[0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>	completed  0  /  500 epochs
	completed  50  /  500 epochs
	completed  100  /  500 epochs
	completed  150  /  500 epochs
	completed  200  /  500 epochs
	completed  250  /  500 epochs
	completed  300  /  500 epochs
	completed  350  /  500 epochs
	completed  400  /  500 epochs
	completed  450  /  500 epochs
</pre></div>
</div>
<img alt="../_images/6200757d470dc727d3a6e2ed0bbfc6232e589641ab94e2130d289e0f9774c48d.png" src="../_images/6200757d470dc727d3a6e2ed0bbfc6232e589641ab94e2130d289e0f9774c48d.png" />
</div>
</div>
<section id="vector-field">
<h3>Vector field<a class="headerlink" href="#vector-field" title="Permalink to this heading"></a></h3>
<p>The vector field of trajectories is projected onto the embedding</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#default parameters:</span>
<span class="c1"># scatter_size=500</span>
<span class="c1"># scatter_alpha=0.5</span>
<span class="c1"># marker_edgewidth=0.1, outline of scatter points</span>
<span class="c1"># density_stream = 2, spacing between field lines</span>
<span class="c1"># smooth_transition=1,smoothing of transition matrix over neighbors of neighbors</span>
<span class="c1"># color_scheme = &#39;annotation&#39;, corresponds to true_labels  (cell types) provided. Can also be set to &#39;cluster&#39; and &#39;time&#39; and &#39;other&#39;</span>
<span class="c1"># add_outline_clusters=False</span>
<span class="c1"># other_labels = [] provide a list of values for each cell corresponding to the value / annotation you wish to plot</span>

<span class="n">via_streamplot</span><span class="p">(</span><span class="n">v0</span><span class="p">,</span> <span class="n">embedding</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/shobi/anaconda3/envs/ViaEnv/lib/python3.7/site-packages/scipy/sparse/_index.py:124: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.
  self._set_arrayXarray(i, j, x)
</pre></div>
</div>
<img alt="../_images/5e3ddca36fdc8089964d5953920c7df681c31602ddf42b25c6a9a24b2bdde807.png" src="../_images/5e3ddca36fdc8089964d5953920c7df681c31602ddf42b25c6a9a24b2bdde807.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">animated_streamplot</span><span class="p">(</span><span class="n">v0</span><span class="p">,</span> <span class="n">embedding</span><span class="p">,</span> <span class="n">scatter_size</span><span class="o">=</span><span class="mi">800</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">density_grid</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>  <span class="n">saveto</span><span class="o">=</span><span class="s1">&#39;/home/shobi/Trajectory/Datasets/Toy3/test_toy3.gif&#39;</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2022-08-24 20:32:55.315504	Inside animated. File will be saved to location /home/shobi/Trajectory/Datasets/Toy3/test_toy3.gif
total number of stream lines 2329
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  0%|          | 0/27 [00:00&lt;?, ?it/s]
</pre></div>
</div>
<img alt="../_images/fdc5a593f6024972095ef26fb1a60ac9bda2b9ae4a0d751af19cdac25146b64e.png" src="../_images/fdc5a593f6024972095ef26fb1a60ac9bda2b9ae4a0d751af19cdac25146b64e.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">Image</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;/home/shobi/Trajectory/Datasets/Toy3/test_toy3.gif&#39;</span><span class="p">,</span><span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">display</span><span class="p">(</span><span class="n">Image</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">read</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/8d2c20347a2d9b36903bc7196d66364b213c6a1d21c1f3eb2bc1adda49232a14.png" src="../_images/8d2c20347a2d9b36903bc7196d66364b213c6a1d21c1f3eb2bc1adda49232a14.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span><span class="n">via_streamplot</span><span class="p">(</span><span class="n">v0</span><span class="p">,</span> <span class="n">embedding</span><span class="p">,</span> <span class="n">color_scheme</span><span class="o">=</span><span class="s1">&#39;time&#39;</span><span class="p">,</span><span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">min_mass</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">cutoff_perc</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span><span class="n">scatter_size</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span><span class="n">marker_edgewidth</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">density_stream</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">smooth_transition</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">smooth_grid</span><span class="o">=</span><span class="mf">.5</span><span class="p">)</span><span class="c1"># other args: add_outline_clusters=True, cluster_outline_edgewidth =0.005</span>
<span class="n">f</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b00cb839812d8bc6d9a0fd4381511ad65574283a17aee1d6561485793652ecbf.png" src="../_images/b00cb839812d8bc6d9a0fd4381511ad65574283a17aee1d6561485793652ecbf.png" />
</div>
</div>
</section>
<section id="overall-trajectory">
<h3>Overall trajectory<a class="headerlink" href="#overall-trajectory" title="Permalink to this heading"></a></h3>
<p>Draw overall pseudotime and main trajectories on single cell embedding</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># draw overall pseudotime and main trajectories</span>
<span class="n">draw_trajectory_gams</span><span class="p">(</span><span class="n">via_coarse</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">via_fine</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2022-08-26 13:29:19.041954	Super cluster 4 is a super terminal with sub_terminal cluster 4
2022-08-26 13:29:19.042299	Super cluster 7 is a super terminal with sub_terminal cluster 7
2022-08-26 13:29:19.042495	Super cluster 8 is a super terminal with sub_terminal cluster 8
2022-08-26 13:29:19.042690	Super cluster 9 is a super terminal with sub_terminal cluster 9
2022-08-26 13:29:19.042879	Super cluster 10 is a super terminal with sub_terminal cluster 10
</pre></div>
</div>
<img alt="../_images/15a21ca412c8b4bd40dcf04c002c1452f676709d860b44c9060fb7c92df8bdd1.png" src="../_images/15a21ca412c8b4bd40dcf04c002c1452f676709d860b44c9060fb7c92df8bdd1.png" />
</div>
</div>
</section>
<section id="lineage-probabilities">
<h3>Lineage probabilities<a class="headerlink" href="#lineage-probabilities" title="Permalink to this heading"></a></h3>
<p>Plot the lineage probabilities for each lineage (associated with a terminal state)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">draw_sc_lineage_probability</span><span class="p">(</span><span class="n">via_coarse</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">via_fine</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span><span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">f</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-04 14:25:46.714135	Marker_lineages: [4, 7, 8, 9, 10]
2023-01-04 14:25:46.715970	The number of components in the original full graph is 1
2023-01-04 14:25:46.716086	For downstream visualization purposes we are also constructing a low knn-graph 
2023-01-04 14:25:47.090044	Check sc pb [0.444 0.023 0.17  0.339 0.023]
2023-01-04 14:25:47.108323	Cluster path on clustergraph starting from Root Cluster 6 to Terminal Cluster 4: [6, 3, 1, 11, 2, 4]
2023-01-04 14:25:47.108384	Cluster path on clustergraph starting from Root Cluster 6 to Terminal Cluster 7: [6, 3, 1, 12, 0, 5, 7]
2023-01-04 14:25:47.108405	Cluster path on clustergraph starting from Root Cluster 6 to Terminal Cluster 8: [6, 3, 1, 11, 2, 9, 8]
2023-01-04 14:25:47.108423	Cluster path on clustergraph starting from Root Cluster 6 to Terminal Cluster 9: [6, 3, 1, 11, 2, 9]
2023-01-04 14:25:47.108441	Cluster path on clustergraph starting from Root Cluster 6 to Terminal Cluster 10: [6, 3, 1, 12, 0, 5, 10]
2023-01-04 14:25:47.190811	Revised Cluster level path on sc-knnGraph from Root Cluster 6 to Terminal Cluster 4 along path: [6, 6, 3, 11, 2, 4, 4, 4, 4]
2023-01-04 14:25:47.212205	Revised Cluster level path on sc-knnGraph from Root Cluster 6 to Terminal Cluster 7 along path: [6, 6, 3, 12, 0, 5, 7, 7]
2023-01-04 14:25:47.232647	Revised Cluster level path on sc-knnGraph from Root Cluster 6 to Terminal Cluster 8 along path: [6, 6, 3, 11, 2, 9, 8, 8, 8, 8, 8, 8]
2023-01-04 14:25:47.253150	Revised Cluster level path on sc-knnGraph from Root Cluster 6 to Terminal Cluster 9 along path: [6, 6, 3, 1, 11, 2, 9, 9]
2023-01-04 14:25:47.274188	Revised Cluster level path on sc-knnGraph from Root Cluster 6 to Terminal Cluster 10 along path: [6, 6, 3, 12, 0, 5, 10, 10, 10]
</pre></div>
</div>
<img alt="../_images/9d3de51a7267499afd28117faad1a2482539de7af2c8186700c442356e873481.png" src="../_images/9d3de51a7267499afd28117faad1a2482539de7af2c8186700c442356e873481.png" />
</div>
</div>
</section>
<section id="lineage-pathways-edge-views">
<h3>Lineage Pathways - edge views<a class="headerlink" href="#lineage-pathways-edge-views" title="Permalink to this heading"></a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">plot_edge_bundle</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">n_milestones</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth_bundle</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">alpha_bundle_factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                     <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;plasma&#39;</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">size_scatter</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">alpha_scatter</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">scale_scatter_size_pop</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                     <span class="n">extra_title_text</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">headwidth_bundle</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">lineage_pathway</span> <span class="o">=</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">],</span> <span class="n">text_labels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sc_labels</span><span class="o">=</span><span class="n">true_label</span><span class="p">)</span>
<span class="n">f</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-06 15:02:46.482837	Computing Edges
2023-01-06 15:02:46.482945	Start finding milestones
2023-01-06 15:02:46.806557	End milestones
2023-01-06 15:02:46.808889	Recompute weights
2023-01-06 15:02:46.812384	pruning milestone graph based on recomputed weights
2023-01-06 15:02:46.813313	Graph has 1 connected components before pruning
2023-01-06 15:02:46.813637	Graph has 1 connected components before pruning
2023-01-06 15:02:46.814280	regenerate igraph on pruned edges
2023-01-06 15:02:46.818722	Setting numeric label as time_series_labels or other sequential metadata for coloring edges
2023-01-06 15:02:46.823656	Making smooth edges
</pre></div>
</div>
<img alt="../_images/112d8236366f06f19e60478c18526a819b4e08912de850a0cfd98975f86b71ff.png" src="../_images/112d8236366f06f19e60478c18526a819b4e08912de850a0cfd98975f86b71ff.png" />
</div>
</div>
</section>
<section id="gene-dynamics">
<h3>Gene Dynamics<a class="headerlink" href="#gene-dynamics" title="Permalink to this heading"></a></h3>
<p>Plot the gene expression levels along each lineage for marker genes. Here we use dummy values for the Genes which are just the values of the initial PCs for illustrative purposes</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_subset</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_counts</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gene0&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene1&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene2&#39;</span><span class="p">])</span>
<span class="n">f</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">get_gene_expression</span><span class="p">(</span><span class="n">via0</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">gene_exp</span><span class="o">=</span><span class="n">df_subset</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;jet&#39;</span><span class="p">,</span><span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">fontsize_</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">spline_order</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">n_splines</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span> <span class="c1">#default parameters shown here for retrieving gene expression</span>
<span class="n">f</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/4218158d2db1ff5d6b936fc63b0b5db3c44a3de80729e0a0c7b6f45a566d4184.png" src="../_images/4218158d2db1ff5d6b936fc63b0b5db3c44a3de80729e0a0c7b6f45a566d4184.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">v0</span><span class="o">.</span><span class="n">embedding</span> <span class="o">=</span> <span class="n">via_mds_embedding2</span>
<span class="n">f</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">plot_edge_bundle</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">n_milestones</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">linewidth_bundle</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">alpha_bundle_factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                     <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;plasma_r&#39;</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">size_scatter</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">alpha_scatter</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">scale_scatter_size_pop</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                     <span class="n">extra_title_text</span><span class="o">=</span><span class="s1">&#39;pseudotime&#39;</span><span class="p">,</span> <span class="n">headwidth_bundle</span><span class="o">=</span><span class="mf">2.5</span><span class="p">,</span> <span class="n">text_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sc_labels</span><span class="o">=</span><span class="n">true_label</span><span class="p">)</span>
<span class="n">f</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">plot_edge_bundle</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">n_milestones</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth_bundle</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">alpha_bundle_factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                     <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;plasma_r&#39;</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">size_scatter</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">alpha_scatter</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">scale_scatter_size_pop</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                     <span class="n">extra_title_text</span><span class="o">=</span><span class="s1">&#39;Gene0 Expression&#39;</span><span class="p">,</span> <span class="n">headwidth_bundle</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">sc_labels_expression</span> <span class="o">=</span> <span class="n">adata_counts</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">text_labels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sc_labels</span><span class="o">=</span><span class="n">true_label</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-06 15:01:56.908671	Computing Edges
2023-01-06 15:01:56.908944	Start finding milestones
2023-01-06 15:01:57.161464	End milestones
2023-01-06 15:01:57.163345	Recompute weights
2023-01-06 15:01:57.165011	pruning milestone graph based on recomputed weights
2023-01-06 15:01:57.165856	Graph has 1 connected components before pruning
2023-01-06 15:01:57.166166	Graph has 1 connected components before pruning
2023-01-06 15:01:57.167271	regenerate igraph on pruned edges
2023-01-06 15:01:57.172146	Setting numeric label as time_series_labels or other sequential metadata for coloring edges
2023-01-06 15:01:57.177309	Making smooth edges
2023-01-06 15:01:57.510906	Computing Edges
2023-01-06 15:01:57.511040	Start finding milestones
2023-01-06 15:01:57.654583	End milestones
2023-01-06 15:01:57.656575	Recompute weights
2023-01-06 15:01:57.659783	pruning milestone graph based on recomputed weights
2023-01-06 15:01:57.660657	Graph has 1 connected components before pruning
2023-01-06 15:01:57.660979	Graph has 1 connected components before pruning
2023-01-06 15:01:57.661612	regenerate igraph on pruned edges
2023-01-06 15:01:57.665804	Setting numeric label as time_series_labels or other sequential metadata for coloring edges
2023-01-06 15:01:57.670556	Making smooth edges
</pre></div>
</div>
<img alt="../_images/80c72d634e91730dfc9b7187abbe5549d2534468834d12d46b4bb56aaddae12c.png" src="../_images/80c72d634e91730dfc9b7187abbe5549d2534468834d12d46b4bb56aaddae12c.png" />
<img alt="../_images/b9efb723750fc87a77165e65be03f5956fa130b5bea7fc0638c02dd484d76c94.png" src="../_images/b9efb723750fc87a77165e65be03f5956fa130b5bea7fc0638c02dd484d76c94.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_counts</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gene0&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene1&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene2&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene3&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene4&#39;</span><span class="p">])</span>

<span class="n">f</span><span class="p">,</span> <span class="n">axlist</span> <span class="o">=</span> <span class="n">plot_gene_trend_heatmaps</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">df_gene_exp</span><span class="o">=</span><span class="n">df_genes</span><span class="p">,</span> <span class="n">marker_lineages</span><span class="o">=</span><span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">])</span>
<span class="n">axlist</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;pseudotime&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">f</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/434597b2e0e3f44a8772eea4aa1844f22383167e7c460fc4ac0412dcf2e0b7a1.png" src="../_images/434597b2e0e3f44a8772eea4aa1844f22383167e7c460fc4ac0412dcf2e0b7a1.png" />
</div>
</div>
</section>
</section>
<section id="pre-defined-terminal-states-and-integrated-visualization">
<h2>Pre-defined terminal states and Integrated Visualization<a class="headerlink" href="#pre-defined-terminal-states-and-integrated-visualization" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>In some cases you may have a good idea of what the terminal states are, in this case you can set them when initalized VIA based on group level corresponding to true_label or single-cell level as a list of string or integers</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">user_defined_terminal_group=['M6','M8','M2','M7']</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">user_defined_terminal_cell</span> <span class="pre">=</span> <span class="pre">[89,560,808]</span></code></p></li>
</ul>
</li>
<li><p>You may not have a precomputed embedding, doing this as part of the VIA analysis is very fast. The embeddings can also be refined after running VIA once you have a via_object available</p></li>
<li><p>If you have timeseries data then this is strongly recommended as you can get superior visualization that make use of the temporal information provided by <code class="docutils literal notranslate"><span class="pre">time_series</span> <span class="pre">=</span> <span class="pre">True</span></code> and <code class="docutils literal notranslate"><span class="pre">time_series_labels</span></code> (list of numeric values of length nsamples)</p></li>
<li><p>If you have very large data using via-mds is also recommended as it is very fast</p></li>
<li><p>‘via-mds’ is generally faster but can require more parameter adjustment that ‘via-umap’ to optimize the visualization</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">do_compute_embedding</span> <span class="pre">=</span> <span class="pre">True</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">embedding_type</span> <span class="pre">=</span> <span class="pre">'via-mds'</span> <span class="pre">or</span> <span class="pre">'via-umap'</span></code></p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#define parameters</span>
<span class="n">ncomps</span><span class="p">,</span> <span class="n">knn</span><span class="p">,</span> <span class="n">random_seed</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">root_user</span>  <span class="o">=</span><span class="mi">30</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span> <span class="mi">42</span><span class="p">,</span><span class="s1">&#39;toy&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;M1&#39;</span><span class="p">]</span>

<span class="n">v0</span> <span class="o">=</span> <span class="n">VIA</span><span class="p">(</span><span class="n">adata_counts</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="n">ncomps</span><span class="p">],</span> <span class="n">true_label</span><span class="p">,</span> <span class="n">jac_std_global</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">dist_std_local</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
             <span class="n">knn</span><span class="o">=</span><span class="n">knn</span><span class="p">,</span> <span class="n">cluster_graph_pruning_std</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">too_big_factor</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">root_user</span><span class="o">=</span><span class="n">root_user</span><span class="p">,</span> <span class="n">preserve_disconnected</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span>
             <span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">,</span> <span class="n">user_defined_terminal_group</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;M6&#39;</span><span class="p">,</span><span class="s1">&#39;M8&#39;</span><span class="p">,</span><span class="s1">&#39;M2&#39;</span><span class="p">,</span><span class="s1">&#39;M4&#39;</span><span class="p">],</span> <span class="n">do_compute_embedding</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">embedding_type</span> <span class="o">=</span> <span class="s1">&#39;via-mds&#39;</span><span class="p">)</span>
<span class="n">v0</span><span class="o">.</span><span class="n">run_VIA</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-05 11:39:11.107267	Running VIA over input data of 1000 (samples) x 30 (features)
2023-01-05 11:39:11.107435	Knngraph has 20 neighbors
2023-01-05 11:39:11.394790	Finished global pruning of 20-knn graph used for clustering at level of 0.15. Kept 46.7 % of edges. 
2023-01-05 11:39:11.399722	Number of connected components used for clustergraph  is 1
2023-01-05 11:39:11.425567	Run via-mds
2023-01-05 11:39:11.425667	Commencing Via-MDS
2023-01-05 11:39:11.425883	Resetting n_milestones to 1000 as n_samples &gt; original n_milestones
2023-01-05 11:39:11.781384	Start computing with diffusion power:2
2023-01-05 11:39:11.804889	Starting MDS on milestone
2023-01-05 11:39:12.599374	End computing mds with diffusion power:2
2023-01-05 11:39:12.600595	Commencing community detection
2023-01-05 11:39:12.633515	Finished running Leiden algorithm. Found 43 clusters.
2023-01-05 11:39:12.633911	Merging 30 very small clusters (&lt;10)
2023-01-05 11:39:12.634848	Finished detecting communities. Found 13 communities
2023-01-05 11:39:12.634967	Making cluster graph. Global cluster graph pruning level: 1
2023-01-05 11:39:12.637684	Graph has 1 connected components before pruning
2023-01-05 11:39:12.638783	Graph has 1 connected components before pruning
2023-01-05 11:39:12.639230	0.0% links trimmed from local pruning relative to start
2023-01-05 11:39:12.640604	Starting make edgebundle viagraph...
2023-01-05 11:39:12.640636	Make via clustergraph edgebundle
2023-01-05 11:39:12.790576	Hammer dims: Nodes shape: (13, 2) Edges shape: (26, 3)
2023-01-05 11:39:12.796419	New root is 3 and majority M1
2023-01-05 11:39:12.796802	New root is 6 and majority M1
2023-01-05 11:39:12.797174	Computing lazy-teleporting expected hitting times
2023-01-05 11:39:13.290067	Terminal cluster list based on user defined cells/groups: [(&#39;M6&#39;, 4), (&#39;M8&#39;, 10), (&#39;M2&#39;, 7), (&#39;M4&#39;, 8)]
2023-01-05 11:39:13.290393	Terminal clusters corresponding to unique lineages in this component are [4, 10, 7, 8] 
2023-01-05 11:39:13.522544	From root 6,  the Terminal state 4 is reached 75 times.
2023-01-05 11:39:13.734316	From root 6,  the Terminal state 10 is reached 394 times.
2023-01-05 11:39:13.919576	From root 6,  the Terminal state 7 is reached 426 times.
2023-01-05 11:39:14.187448	From root 6,  the Terminal state 8 is reached 249 times.
2023-01-05 11:39:14.208372	Terminal clusters corresponding to unique lineages are {4: &#39;M6&#39;, 10: &#39;M8&#39;, 7: &#39;M2&#39;, 8: &#39;M4&#39;}
2023-01-05 11:39:14.208501	Begin projection of pseudotime and lineage likelihood
2023-01-05 11:39:14.344523	Graph has 1 connected components before pruning
2023-01-05 11:39:14.345884	Graph has 1 connected components before pruning
2023-01-05 11:39:14.346360	11.5% links trimmed from local pruning relative to start
2023-01-05 11:39:14.346412	50.0% links trimmed from global pruning relative to start
2023-01-05 11:39:14.348142	Start making edgebundle milestone...
2023-01-05 11:39:14.348196	Start finding milestones
2023-01-05 11:39:15.813594	End milestones
2023-01-05 11:39:15.813709	Will use via-pseudotime for edges, otherwise consider providing a list of numeric labels (single cell level) or via_object
2023-01-05 11:39:15.816299	Recompute weights
2023-01-05 11:39:15.828333	pruning milestone graph based on recomputed weights
2023-01-05 11:39:15.829286	Graph has 1 connected components before pruning
2023-01-05 11:39:15.829595	Graph has 1 connected components before pruning
2023-01-05 11:39:15.830304	61.8% links trimmed from global pruning relative to start
2023-01-05 11:39:15.830353	regenerate igraph on pruned edges
2023-01-05 11:39:15.833813	Setting numeric label as single cell pseudotime for coloring edges
2023-01-05 11:39:15.840870	Making smooth edges
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>No handles with labels found to put in legend.
</pre></div>
</div>
<img alt="../_images/80bb1dad4f0f6f4957ab6a4f5d9e9c6d1c72b6e7b2a3451fe59e60b90000846b.png" src="../_images/80bb1dad4f0f6f4957ab6a4f5d9e9c6d1c72b6e7b2a3451fe59e60b90000846b.png" />
<img alt="../_images/40e4a3a43239a19780b41fa04772e37f0828aeb45d2fbac5d5ca5149f8965020.png" src="../_images/40e4a3a43239a19780b41fa04772e37f0828aeb45d2fbac5d5ca5149f8965020.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-05 11:39:16.675769	Time elapsed 5.4 seconds
</pre></div>
</div>
</div>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2022, shobana stassen
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">1. Basic workflow</a><ul>
<li><a class="reference internal" href="#expected-outputs">Expected Outputs</a></li>
<li><a class="reference internal" href="#parameters">Parameters</a></li>
<li><a class="reference internal" href="#via-graph">VIA graph</a></li>
<li><a class="reference internal" href="#scatter-plots">Scatter Plots</a><ul>
<li><a class="reference internal" href="#colored-by-true-label">colored by true_label</a></li>
</ul>
</li>
<li><a class="reference internal" href="#fast-visualization-using-underlying-single-cell-via-graphs">Fast visualization using underlying single-cell VIA-graphs</a><ul>
<li><a class="reference internal" href="#via-mds-parameters">VIA-mds Parameters</a></li>
</ul>
</li>
<li><a class="reference internal" href="#via-umap-based-on-underlying-via-single-cell-knn-graphs">VIA-umap based on underlying via single cell knn graphs</a><ul>
<li><a class="reference internal" href="#vector-field">Vector field</a></li>
<li><a class="reference internal" href="#overall-trajectory">Overall trajectory</a></li>
<li><a class="reference internal" href="#lineage-probabilities">Lineage probabilities</a></li>
<li><a class="reference internal" href="#lineage-pathways-edge-views">Lineage Pathways - edge views</a></li>
<li><a class="reference internal" href="#gene-dynamics">Gene Dynamics</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pre-defined-terminal-states-and-integrated-visualization">Pre-defined terminal states and Integrated Visualization</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/scripts/furo.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script defer="defer" src="https://unpkg.com/@popperjs/core@2"></script>
    <script defer="defer" src="https://unpkg.com/tippy.js@6"></script>
    <script defer="defer" src="../_static/tippy\notebooks\Basic tutorial.8eebe737-da15-49bc-a1e8-c53f66133b42.js"></script>
    </body>
</html>