<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="4. Imaging cytometry" href="Imaging%20Cytometry%20%28cell%20cycle%29.html" /><link rel="prev" title="2. Disconnected Trajectories" href="ViaJupyter_Toy_Disconnected.html" />

    <!-- Generated with Sphinx 5.1.1 and Furo 2023.03.27 -->
        <title>3. Via Human Hematopoiesis - pyvia documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?digest=fad236701ea90a88636c2a8c73b44ae642ed2a53" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">pyvia  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">pyvia  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">General</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../pyVia-home.html">StaVia - Multi-Omic Single-Cell Cartography for Spatial and Temporal Atlases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Release%20History.html">Release History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Tutorial%20Video.html">VIA Tutorial Videos</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Parameters%20and%20Attributes.html">Parameters and Attributes</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../api/index.html">API</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../api/pyVIA%20Core.html">pyVIA core</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/Plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.animate_atlas.html">pyVIA.plotting_via.animate_atlas</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.animate_streamplot.html">pyVIA.plotting_via.animate_streamplot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.get_gene_expression.html">pyVIA.plotting_via.get_gene_expression</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_atlas_view.html">pyVIA.plotting_via.plot_atlas_view</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_differentiation_flow.html">pyVIA.plotting_via.plot_differentiation_flow</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_gene_trend_heatmaps.html">pyVIA.plotting_via.plot_gene_trend_heatmaps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_piechart_viagraph.html">pyVIA.plotting_via.plot_piechart_viagraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_population_composition.html">pyVIA.plotting_via.plot_population_composition</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_sc_lineage_probability.html">pyVIA.plotting_via.plot_sc_lineage_probability</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_scatter.html">pyVIA.plotting_via.plot_scatter</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_trajectory_curves.html">pyVIA.plotting_via.plot_trajectory_curves</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_viagraph.html">pyVIA.plotting_via.plot_viagraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_streamplot.html">pyVIA.plotting_via.via_streamplot</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/Utils.html">Util Functions for plotting</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.make_edgebundle_milestone.html">pyVIA.plotting_via.make_edgebundle_milestone</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_atlas_emb.html">pyVIA.plotting_via.via_atlas_emb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_forcelayout.html">pyVIA.plotting_via.via_forcelayout</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_mds.html">pyVIA.plotting_via.via_mds</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/Datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/datasets_via.cell_cycle_cyto_data.html">datasets_via.cell_cycle_cyto_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/datasets_via.embryoid_body.html">datasets_via.embryoid_body</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/datasets_via.scATAC_hematopoiesis.html">datasets_via.scATAC_hematopoiesis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/datasets_via.scRNA_hematopoiesis.html">datasets_via.scRNA_hematopoiesis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/datasets_via.toy_disconnected.html">datasets_via.toy_disconnected</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/datasets_via.toy_multifurcating.html">datasets_via.toy_multifurcating</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../Basic%20Example%20Code.html">Examples for installation checking</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Via 2.0 Atlas View Gallery</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Atlas%20view%20examples.html">Via2.0 Atlas views</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mouse_to_pup_atlas.html">Large Mouse Embryo to Pup Developmental Atlas</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Via2%20Atlas%20Animation.html">Via 2.0 Atlas Animations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Via%20Atlas%20View%20for%20Spatial%20omics.html">Via2.0 Atlas views for Spatial Omics data</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Via 2.0 Cartography</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Via2.0%20Cartographic%20Mouse%20Gastrualation.html">1. Via 2.0 Cartography on Mouse Gastrulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Zebrahub%20TI%20tutorial.html">2. Via 2.0 Cartography on Zebrahub (Trajectory Inference)</a></li>
<li class="toctree-l1"><a class="reference internal" href="Zebrahub_tutorial_visualization.html">3. Via 2.0 Cartography on Zebrahub (Visualization)</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials Via</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_Toy_Multifurcating.html">1. Basic workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_Toy_Disconnected.html">2. Disconnected Trajectories</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">3. Via Human Hematopoiesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="Imaging%20Cytometry%20%28cell%20cycle%29.html">4. Imaging cytometry</a></li>
<li class="toctree-l1"><a class="reference internal" href="mESC_timeseries.html">5. Using time-series metadata (mESC Cytof)</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_Pancreas_RNAvelocity.html">6. Using RNA-velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_scRNAVelocity_hematopoiesis.html">7. Bone marrow with RNA-velocity</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section class="tex2jax_ignore mathjax_ignore" id="via-human-hematopoiesis">
<h1>3. Via Human Hematopoiesis<a class="headerlink" href="#via-human-hematopoiesis" title="Permalink to this heading"></a></h1>
<p>This notebook uses VIA to interpret the CD34 Hematopoiesis dataset (Setty et al.,2019)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pyVIA.core</span> <span class="k">as</span> <span class="nn">via</span>
<span class="kn">import</span> <span class="nn">pyVIA.datasets_via</span> <span class="k">as</span>   <span class="nn">datasets</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span> 
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="load-data-from-directory">
<h2>Load data from directory<a class="headerlink" href="#load-data-from-directory" title="Permalink to this heading"></a></h2>
<p>We use annotations given by SingleR which uses Novershtern Hematopoietic cell data as the reference. These are in line with the annotations given by Setty et al., 2019 but offer a slightly more granular annotation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">datasets</span><span class="o">.</span><span class="n">scRNA_hematopoiesis</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">tsnem</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">]</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Cell type CMP has  968 cells
Cell type PRE_B3 has  2 cells
Cell type B_a4 has  8 cells
Cell type pDC has  204 cells
Cell type HSC2 has  10 cells
Cell type GRAN1 has  3 cells
Cell type Nka3 has  5 cells
Cell type EOS2 has  3 cells
Cell type MONO1 has  186 cells
Cell type B_a1 has  27 cells
Cell type ERY3 has  104 cells
Cell type ERY1 has  773 cells
Cell type HSC1 has  2365 cells
Cell type MONO2 has  35 cells
Cell type ERY2 has  40 cells
Cell type mDC (cDC) has  96 cells
Cell type B_a2 has  1 cells
Cell type BASO1 has  2 cells
Cell type ERY4 has  53 cells
Cell type MEP has  339 cells
Cell type PRE_B2 has  515 cells
Cell type TCEL7 has  1 cells
Cell type GMP has  28 cells
Cell type MEGA1 has  12 cells
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;, &#39;label&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;pca&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_pca&#39;
    varm: &#39;PCs&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="user-s-choice-of-embedding">
<h2>User’s choice of embedding<a class="headerlink" href="#user-s-choice-of-embedding" title="Permalink to this heading"></a></h2>
<p>Three embedding exmaples: TSNE, UMAP, and PHATE
Alternatively allow via.VIA() to compute an embedding using the underlying graph by setting: do_embedding = True AND embedding_type = ‘via-umap’ OR ‘via-mds’</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># TSNE - same as the embedding used in the original Setty et al., publication</span>
<span class="n">embedding</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">]</span>

<span class="c1"># # UMAP</span>
<span class="c1"># ncomp = 30</span>
<span class="c1"># embedding = umap.UMAP().fit_transform(adata.obsm[&#39;X_pca&#39;][:, 0:ncomp])</span>

<span class="c1"># # PHATE</span>
<span class="c1"># ncomp = 30</span>
<span class="c1"># phate_op = phate.PHATE()</span>
<span class="c1"># embedding = phate_op.fit_transform(adata.obsm[&#39;X_pca&#39;][:, 0:ncomp])</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="run-via">
<h2>Run VIA<a class="headerlink" href="#run-via" title="Permalink to this heading"></a></h2>
<section id="key-parameters">
<h3>Key Parameters:<a class="headerlink" href="#key-parameters" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>knn</p></li>
<li><p>root_user (root index of type int/ or celltype label of type string)</p></li>
<li><p>edgepruning_clustering_resolution (lower number means smaller (and more) clusters) typical range 0-1</p></li>
<li><p>memory (higher value means more focused search pathways to cell fates) typical range 2-50</p></li>
<li><p>cluster_graph_pruning (lower number means fewer edges) typical range 0-1</p></li>
<li><p>dataset (relates to the root_user parameter. if the root is an index, the leave as dataset =’’. If the root is a celltype label found in true_label paramters, then dataset=’group’)</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ncomps</span><span class="o">=</span><span class="mi">80</span>
<span class="n">knn</span><span class="o">=</span><span class="mi">30</span>
<span class="n">v0_random_seed</span><span class="o">=</span><span class="mi">4</span>
<span class="n">root_user</span> <span class="o">=</span> <span class="p">[</span><span class="mi">4823</span><span class="p">]</span> <span class="c1">#the index of a cell belonging to the HSC cell type</span>
<span class="n">memory</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">#NOTE1, if you decide to choose a cell type as a root, then you need to set the dataset as &#39;group&#39;</span>
<span class="sd">#root_user=[&#39;HSC1&#39;]</span>
<span class="sd">#dataset = &#39;group&#39;# &#39;humanCD34&#39;</span>
<span class="sd">#NOTE2, if rna-velocity is available, considering using it to compute the root automatically- see RNA velocity tutorial</span>
<span class="sd">&#39;&#39;&#39;</span>

<span class="n">v0</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">VIA</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="n">ncomps</span><span class="p">],</span> <span class="n">true_label</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">],</span> 
             <span class="n">edgepruning_clustering_resolution</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">cluster_graph_pruning</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span>
             <span class="n">knn</span><span class="o">=</span><span class="n">knn</span><span class="p">,</span>  <span class="n">root_user</span><span class="o">=</span><span class="n">root_user</span><span class="p">,</span> <span class="n">resolution_parameter</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span>
             <span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="n">v0_random_seed</span><span class="p">,</span> <span class="n">memory</span><span class="o">=</span><span class="n">memory</span><span class="p">)</span><span class="c1">#, do_compute_embedding=True, embedding_type=&#39;via-atlas&#39;)</span>

<span class="n">v0</span><span class="o">.</span><span class="n">run_VIA</span><span class="p">()</span>

<span class="c1">#NOTE2: if you want to pre-specify terminal states you can do so by a list of group-level names corresponding to true-labels OR by a list of single cell indices</span>
<span class="c1">#E.G. user_defined_terminal_group=[&#39;pDC&#39;,&#39;ERY1&#39;, &#39;ERY3&#39;, &#39;MONO1&#39;,&#39;mDC (cDC)&#39;,&#39;PRE_B2&#39;]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-30 14:26:27.296505	Running VIA over input data of 5780 (samples) x 80 (features)
2023-10-30 14:26:27.296612	Knngraph has 30 neighbors
2023-10-30 14:26:31.954335	Finished global pruning of 30-knn graph used for clustering at level of 0.15. Kept 46.3 % of edges. 
2023-10-30 14:26:32.011295	Number of connected components used for clustergraph  is 1
2023-10-30 14:26:32.503636	Commencing community detection
2023-10-30 14:26:32.631581	Finished running Leiden algorithm. Found 212 clusters.
2023-10-30 14:26:32.637560	Merging 188 very small clusters (&lt;10)
2023-10-30 14:26:32.642614	Finished detecting communities. Found 24 communities
2023-10-30 14:26:32.643731	Making cluster graph. Global cluster graph pruning level: 0.15
2023-10-30 14:26:32.673200	Graph has 1 connected components before pruning
2023-10-30 14:26:32.676487	Graph has 2 connected components after pruning
2023-10-30 14:26:32.678096	Graph has 1 connected components after reconnecting
2023-10-30 14:26:32.678934	0.0% links trimmed from local pruning relative to start
2023-10-30 14:26:32.678958	65.5% links trimmed from global pruning relative to start
2023-10-30 14:26:32.683710	component number 0 out of  [0]
2023-10-30 14:26:32.707100	The root index, 4823 provided by the user belongs to cluster number 1 and corresponds to cell type HSC1
2023-10-30 14:26:32.709534	Computing lazy-teleporting expected hitting times
2023-10-30 14:26:33.454240	 ended all multiprocesses, will retrieve and reshape
try rw2 hitting times setup
start computing walks with rw2 method
g.indptr.size, 25
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "6100c42d75354ae185a7a6482bcad918"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>memory for rw2 hittings times  2. Using rw2 based pt
do scaling of pt
2023-10-30 14:26:40.339327	Identifying terminal clusters corresponding to unique lineages...
2023-10-30 14:26:40.339362	Closeness:[1, 7, 8, 11, 12, 13, 15, 18]
2023-10-30 14:26:40.339379	Betweenness:[1, 3, 6, 7, 11, 12, 13, 15, 18, 20, 21, 22]
2023-10-30 14:26:40.339392	Out Degree:[1, 5, 7, 8, 9, 11, 12, 13, 15, 18, 20, 22]
2023-10-30 14:26:40.340052	Terminal clusters corresponding to unique lineages in this component are [7, 11, 12, 15, 18, 20, 22] 
TESTING rw2_lineage probability at memory 10
testing rw2 lineage probability at memory 10
g.indptr.size, 25
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "83cdff46461f48c2a7f71318d0346dc1"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-30 14:26:44.688612	 Cluster or terminal cell fate 7 is reached 11.0 times
2023-10-30 14:26:44.769144	 Cluster or terminal cell fate 11 is reached 189.0 times
2023-10-30 14:26:44.857752	 Cluster or terminal cell fate 12 is reached 11.0 times
2023-10-30 14:26:44.957626	 Cluster or terminal cell fate 15 is reached 303.0 times
2023-10-30 14:26:45.048059	 Cluster or terminal cell fate 18 is reached 155.0 times
2023-10-30 14:26:45.125893	 Cluster or terminal cell fate 20 is reached 644.0 times
2023-10-30 14:26:45.201391	 Cluster or terminal cell fate 22 is reached 709.0 times
2023-10-30 14:26:45.213079	There are (7) terminal clusters corresponding to unique lineages {7: &#39;PRE_B2&#39;, 11: &#39;ERY3&#39;, 12: &#39;PRE_B2&#39;, 15: &#39;MONO1&#39;, 18: &#39;ERY1&#39;, 20: &#39;pDC&#39;, 22: &#39;mDC (cDC)&#39;}
2023-10-30 14:26:45.213129	Begin projection of pseudotime and lineage likelihood
2023-10-30 14:26:46.728132	Cluster graph layout based on forward biasing
2023-10-30 14:26:46.732186	Starting make edgebundle viagraph...
2023-10-30 14:26:46.732218	Make via clustergraph edgebundle
2023-10-30 14:26:48.990054	Hammer dims: Nodes shape: (24, 2) Edges shape: (102, 3)
2023-10-30 14:26:48.991583	Graph has 1 connected components before pruning
2023-10-30 14:26:48.994114	Graph has 5 connected components after pruning
2023-10-30 14:26:48.998647	Graph has 1 connected components after reconnecting
2023-10-30 14:26:48.999453	53.9% links trimmed from local pruning relative to start
2023-10-30 14:26:48.999478	69.6% links trimmed from global pruning relative to start
2023-10-30 14:26:49.018135	Time elapsed 20.0 seconds
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="cluster-level-trajectory-plots">
<h2>Cluster Level Trajectory Plots<a class="headerlink" href="#cluster-level-trajectory-plots" title="Permalink to this heading"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">ax1</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_piechart_viagraph</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">show_legend</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/6cbcf2b577978c4f2537be3a28904ca8800542783239e8b11cfb6da05fb86a62.png" src="../_images/6cbcf2b577978c4f2537be3a28904ca8800542783239e8b11cfb6da05fb86a62.png" />
</div>
</div>
</section>
<section id="visualise-gene-feature-graph">
<h2>Visualise gene/feature graph<a class="headerlink" href="#visualise-gene-feature-graph" title="Permalink to this heading"></a></h2>
<p>View the gene expression along the VIA graph. We use the computed HNSW small world graph in VIA to accelerate the gene imputation calculations (using similar approach to MAGIC) as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
<span class="n">df_</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>

<span class="n">gene_list_magic</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IL3RA&#39;</span><span class="p">,</span> <span class="s1">&#39;IRF8&#39;</span><span class="p">,</span> <span class="s1">&#39;GATA1&#39;</span><span class="p">,</span> <span class="s1">&#39;GATA2&#39;</span><span class="p">,</span> <span class="s1">&#39;ITGA2B&#39;</span><span class="p">,</span> <span class="s1">&#39;MPO&#39;</span><span class="p">,</span> <span class="s1">&#39;CD79B&#39;</span><span class="p">,</span> <span class="s1">&#39;SPI1&#39;</span><span class="p">,</span> <span class="s1">&#39;CD34&#39;</span><span class="p">,</span> <span class="s1">&#39;CSF1R&#39;</span><span class="p">,</span> <span class="s1">&#39;ITGAX&#39;</span><span class="p">]</span>

<span class="c1">#optional to do gene expression imputation</span>

<span class="n">df_magic</span> <span class="o">=</span> <span class="n">v0</span><span class="o">.</span><span class="n">do_impute</span><span class="p">(</span><span class="n">df_</span><span class="p">,</span> <span class="n">magic_steps</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list_magic</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_viagraph</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">type_data</span><span class="o">=</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="n">df_genes</span><span class="o">=</span><span class="n">df_magic</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list_magic</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">],</span> <span class="n">arrow_head</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>shape of transition matrix raised to power 3 (5780, 5780)
</pre></div>
</div>
<img alt="../_images/17950f8b701ad73325290792309f8f43cb204b9954a637995d0120ce623f43a3.png" src="../_images/17950f8b701ad73325290792309f8f43cb204b9954a637995d0120ce623f43a3.png" />
</div>
</div>
</section>
<section id="trajectory-projection">
<h2>Trajectory projection<a class="headerlink" href="#trajectory-projection" title="Permalink to this heading"></a></h2>
<p>Visualize the overall VIA trajectory projected onto a 2D embedding (UMAP, Phate, TSNE etc) in different ways.</p>
<ol class="arabic simple">
<li><p>Draw the high-level clustergraph abstraction onto the embedding;</p></li>
<li><p>Draw high-edge resolution directed graph</p></li>
<li><p>Draw a vector field/stream plot of the more fine-grained directionality of cells along the trajectory projected onto an embedding.</p></li>
</ol>
<section id="id1">
<h3>Key Parameters:<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>scatter_size</p></li>
<li><p>scatter_alpha</p></li>
<li><p>linewidth</p></li>
<li><p>draw_all_curves (if too crowded, set to False)</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">ax1</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_trajectory_curves</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span><span class="n">embedding</span><span class="o">=</span><span class="n">tsnem</span><span class="p">,</span> <span class="n">draw_all_curves</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:01 Time:  0:00:01
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-30 13:55:26.039883	Super cluster 7 is a super terminal with sub_terminal cluster 7
2023-10-30 13:55:26.040203	Super cluster 9 is a super terminal with sub_terminal cluster 9
2023-10-30 13:55:26.040240	Super cluster 11 is a super terminal with sub_terminal cluster 11
2023-10-30 13:55:26.040271	Super cluster 12 is a super terminal with sub_terminal cluster 12
2023-10-30 13:55:26.040304	Super cluster 15 is a super terminal with sub_terminal cluster 15
2023-10-30 13:55:26.040351	Super cluster 18 is a super terminal with sub_terminal cluster 18
2023-10-30 13:55:26.040382	Super cluster 20 is a super terminal with sub_terminal cluster 20
2023-10-30 13:55:26.040412	Super cluster 21 is a super terminal with sub_terminal cluster 21
2023-10-30 13:55:26.040441	Super cluster 22 is a super terminal with sub_terminal cluster 22
</pre></div>
</div>
<img alt="../_images/df3ecfbb545bbb1899f57726752facff65d0ed9d289ac570c01b3c66ffb73173.png" src="../_images/df3ecfbb545bbb1899f57726752facff65d0ed9d289ac570c01b3c66ffb73173.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">v0</span><span class="o">.</span><span class="n">embedding</span> <span class="o">=</span> <span class="n">tsnem</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_atlas_view</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span>  <span class="n">n_milestones</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">sc_labels</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">],</span> <span class="n">fontsize_labels</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">extra_title_text</span><span class="o">=</span><span class="s1">&#39;Atlas View colored by pseudotime&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>inside add sc embedding second if
</pre></div>
</div>
<img alt="../_images/734038af0479fd753bc9dfcd6fa758dbf965aa38ff63211b5558fb3f97cbe35f.png" src="../_images/734038af0479fd753bc9dfcd6fa758dbf965aa38ff63211b5558fb3f97cbe35f.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># edge plots can be made with different edge resolutions. Run hammerbundle_milestone_dict() to recompute the edges for plotting. Then provide the new hammerbundle as a parameter to plot_edge_bundle()</span>
<span class="c1"># it is better to compute the edges and save them to the via_object. this gives more control to the merging of edges</span>
<span class="n">decay</span> <span class="o">=</span> <span class="mf">0.6</span> <span class="c1">#increasing decay increasing merging of edges</span>
<span class="n">i_bw</span> <span class="o">=</span> <span class="mf">0.02</span> <span class="c1">#increasing bw increases merging of edges</span>
<span class="n">global_visual_pruning</span> <span class="o">=</span> <span class="mf">0.5</span> <span class="c1">#higher number retains more edges</span>
<span class="n">n_milestones</span> <span class="o">=</span> <span class="mi">200</span>

<span class="n">v0</span><span class="o">.</span><span class="n">hammerbundle_milestone_dict</span><span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">make_edgebundle_milestone</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">n_milestones</span><span class="o">=</span><span class="n">n_milestones</span><span class="p">,</span> <span class="n">decay</span><span class="o">=</span><span class="n">decay</span><span class="p">,</span> <span class="n">initial_bandwidth</span><span class="o">=</span><span class="n">i_bw</span><span class="p">,</span> <span class="n">global_visual_pruning</span><span class="o">=</span><span class="n">global_visual_pruning</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-30 15:03:28.888865	Start finding milestones
2023-10-30 15:03:33.107973	End milestones with 200
2023-10-30 15:03:33.130700	Recompute weights
2023-10-30 15:03:33.163942	pruning milestone graph based on recomputed weights
2023-10-30 15:03:33.165341	Graph has 1 connected components before pruning
2023-10-30 15:03:33.166303	Graph has 1 connected components after pruning
2023-10-30 15:03:33.166606	Graph has 1 connected components after reconnecting
2023-10-30 15:03:33.168703	regenerate igraph on pruned edges
2023-10-30 15:03:33.183040	Setting numeric label as single cell pseudotime for coloring edges
2023-10-30 15:03:33.235016	Making smooth edges
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_atlas_view</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span>  <span class="n">add_sc_embedding</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sc_labels_expression</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;jet&#39;</span><span class="p">,</span> <span class="n">sc_labels</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">],</span> <span class="n">text_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">extra_title_text</span><span class="o">=</span><span class="s1">&#39;Atlas View by Cell type&#39;</span><span class="p">,</span> <span class="n">fontsize_labels</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>inside add sc embedding second if
</pre></div>
</div>
<img alt="../_images/508ed8320492523c461bd0d63ef909675c73f14302fe0e0dc52ce49bcbf700d6.png" src="../_images/508ed8320492523c461bd0d63ef909675c73f14302fe0e0dc52ce49bcbf700d6.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># via_streamplot() requires you to either provide an ndarray as embedding as an input parameter OR for via to have an embedding attribute</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">via_streamplot</span><span class="p">(</span><span class="n">v0</span><span class="p">,</span> <span class="n">embedding</span><span class="o">=</span><span class="n">tsnem</span><span class="p">,</span> <span class="n">density_grid</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">scatter_size</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/49d6ea3ad9c053ebafde2db102ac486db54c38f6340f5308914d00b44c9d0e81.png" src="../_images/49d6ea3ad9c053ebafde2db102ac486db54c38f6340f5308914d00b44c9d0e81.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Colored by pseudotime</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">via_streamplot</span><span class="p">(</span><span class="n">v0</span><span class="p">,</span><span class="n">density_grid</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">scatter_size</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">color_scheme</span><span class="o">=</span><span class="s1">&#39;time&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> 
                             <span class="n">min_mass</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">cutoff_perc</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">marker_edgewidth</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> 
                             <span class="n">density_stream</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">smooth_transition</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">smooth_grid</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/e019a4479e01ab1a064f770667b8a17d9bb478441abf9ca95c959cf75eb0cf8d.png" src="../_images/e019a4479e01ab1a064f770667b8a17d9bb478441abf9ca95c959cf75eb0cf8d.png" />
</div>
</div>
</section>
</section>
<section id="probabilistic-pathways-and-memory">
<h2>Probabilistic pathways and Memory<a class="headerlink" href="#probabilistic-pathways-and-memory" title="Permalink to this heading"></a></h2>
<p>Visualize the probabilistic pathways from root to terminal state as indicated by the lineage likelihood. The higher the linage likelihood, the greater the potential of that particular cell to differentiate towards the terminal state of interest.
Changing the memory paramater will alter the specificity of the lineage pathway.
This can be visualized at the single-cell level but also combined with the Atlas View which visualizes cell-cell connectivity and pathways</p>
<section id="id2">
<h3>Key Parameters:<a class="headerlink" href="#id2" title="Permalink to this heading"></a></h3>
<p>marker_lineages (list) of terminal clusters</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span><span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_sc_lineage_probability</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">marker_lineages</span><span class="o">=</span><span class="p">[</span><span class="mi">7</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span><span class="mi">22</span><span class="p">],</span> <span class="n">embedding</span><span class="o">=</span><span class="n">tsnem</span><span class="p">)</span> <span class="c1">#marker_lineages=v0.terminal_clusters to plot all</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-30 14:42:00.472388	Marker_lineages: [7, 11, 12, 15, 20, 22]
2023-10-30 14:42:00.476814	The number of components in the original full graph is 1
2023-10-30 14:42:00.476862	For downstream visualization purposes we are also constructing a low knn-graph 
2023-10-30 14:42:05.171433	Check sc pb 0.9999999999999999 
f getting majority comp
2023-10-30 14:42:05.475235	Cluster path on clustergraph starting from Root Cluster 1 to Terminal Cluster 7: [1, 2, 0, 4, 16, 8, 12, 7]
2023-10-30 14:42:05.475274	Cluster path on clustergraph starting from Root Cluster 1 to Terminal Cluster 11: [1, 13, 3, 6, 17, 11]
2023-10-30 14:42:05.475293	Cluster path on clustergraph starting from Root Cluster 1 to Terminal Cluster 12: [1, 2, 0, 4, 16, 8, 12]
2023-10-30 14:42:05.475311	Cluster path on clustergraph starting from Root Cluster 1 to Terminal Cluster 15: [1, 2, 0, 4, 16, 5, 15]
2023-10-30 14:42:05.475328	Cluster path on clustergraph starting from Root Cluster 1 to Terminal Cluster 18: [1, 13, 3, 18]
2023-10-30 14:42:05.475345	Cluster path on clustergraph starting from Root Cluster 1 to Terminal Cluster 20: [1, 2, 0, 4, 16, 20]
2023-10-30 14:42:05.475362	Cluster path on clustergraph starting from Root Cluster 1 to Terminal Cluster 22: [1, 2, 0, 4, 22]
2023-10-30 14:42:05.605789	Revised Cluster level path on sc-knnGraph from Root Cluster 1 to Terminal Cluster 7 along path: [1, 1, 1, 8, 12, 12, 12]
2023-10-30 14:42:05.640568	Revised Cluster level path on sc-knnGraph from Root Cluster 1 to Terminal Cluster 11 along path: [1, 1, 2, 6, 11, 11, 11, 11]
2023-10-30 14:42:05.678481	Revised Cluster level path on sc-knnGraph from Root Cluster 1 to Terminal Cluster 12 along path: [1, 1, 1, 8, 12, 12, 12, 12]
2023-10-30 14:42:05.724775	Revised Cluster level path on sc-knnGraph from Root Cluster 1 to Terminal Cluster 15 along path: [1, 1, 2, 9, 15]
2023-10-30 14:42:05.761232	Revised Cluster level path on sc-knnGraph from Root Cluster 1 to Terminal Cluster 20 along path: [1, 1, 1, 8, 20, 20, 20, 20]
2023-10-30 14:42:05.805373	Revised Cluster level path on sc-knnGraph from Root Cluster 1 to Terminal Cluster 22 along path: [1, 1, 1, 8, 22, 22, 22]
</pre></div>
</div>
<img alt="../_images/e3a7d520f40fe204fb92ee319933bc87b6ecafc803a8269b6338e987a941c8d0.png" src="../_images/e3a7d520f40fe204fb92ee319933bc87b6ecafc803a8269b6338e987a941c8d0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span><span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_atlas_view</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">lineage_pathway</span><span class="o">=</span><span class="p">[</span><span class="mi">7</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span><span class="mi">22</span><span class="p">])</span> <span class="c1">#marker_lineages=v0.terminal_clusters to plot all</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>location of 7 is at [0] and 0
location of 11 is at [1] and 1
location of 12 is at [2] and 2
location of 15 is at [3] and 3
location of 20 is at [5] and 5
location of 22 is at [6] and 6
</pre></div>
</div>
<img alt="../_images/b91c125aec0be3c96f3426868f70872b97de3de6e100333d2e986d612e658519.png" src="../_images/b91c125aec0be3c96f3426868f70872b97de3de6e100333d2e986d612e658519.png" />
</div>
</div>
</section>
</section>
<section id="gene-dynamics">
<h2>Gene Dynamics<a class="headerlink" href="#gene-dynamics" title="Permalink to this heading"></a></h2>
<p>The gene dynamics along pseudotime for all detected lineages are automatically inferred by VIA. These can be interpreted as the change in gene expression along any given lineage.</p>
<section id="id3">
<h3>Key Parameters<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>n_splines</p></li>
<li><p>spline_order</p></li>
<li><p>gene_exp (dataframe) single-cell level gene expression of select genes (gene imputation is an optional pre-step)</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">marker_genes</span> <span class="o">=</span><span class="p">[</span><span class="s1">&#39;ITGA2B&#39;</span><span class="p">,</span> <span class="s1">&#39;IL3RA&#39;</span><span class="p">,</span> <span class="s1">&#39;IRF8&#39;</span><span class="p">,</span> <span class="s1">&#39;MPO&#39;</span><span class="p">,</span> <span class="s1">&#39;CSF1R&#39;</span><span class="p">,</span> <span class="s1">&#39;GATA2&#39;</span><span class="p">,</span> <span class="s1">&#39;CD79B&#39;</span><span class="p">,</span> <span class="s1">&#39;CD34&#39;</span><span class="p">]</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
<span class="n">df_magic</span> <span class="o">=</span> <span class="n">v0</span><span class="o">.</span><span class="n">do_impute</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">magic_steps</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list_magic</span><span class="p">)</span> <span class="c1">#optional</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span><span class="o">=</span><span class="n">via</span><span class="o">.</span><span class="n">get_gene_expression</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">gene_exp</span><span class="o">=</span><span class="n">df_magic</span><span class="p">[</span><span class="n">marker_genes</span><span class="p">])</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span><span class="mi">7</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>shape of transition matrix raised to power 3 (5780, 5780)
</pre></div>
</div>
<img alt="../_images/07f44b4ed0cf846566a108ba1e94bb7083afbfea9e0100c6655e6d377684465f.png" src="../_images/07f44b4ed0cf846566a108ba1e94bb7083afbfea9e0100c6655e6d377684465f.png" />
</div>
</div>
</section>
</section>
<section id="id4">
<h2>Gene Dynamics<a class="headerlink" href="#id4" title="Permalink to this heading"></a></h2>
<p>Heatmaps of genes along pseudotimetime</p>
<section id="id5">
<h3>Key Parameters<a class="headerlink" href="#id5" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>df_gene_exp (dataframe) single-cell level gene expression of selected genes</p></li>
<li><p>marker_lineages (list, optional) to specify which lineages to plot heatmaps for</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_gene_trend_heatmaps</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">df_gene_exp</span><span class="o">=</span><span class="n">df_magic</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;plasma&#39;</span><span class="p">,</span>
                             <span class="n">marker_lineages</span><span class="o">=</span><span class="p">[</span><span class="mi">7</span><span class="p">,</span><span class="mi">11</span><span class="p">])</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>branches [7, 11]
</pre></div>
</div>
<img alt="../_images/18beee593add4909323fb14b83aabddf46b31871c3c5d3eb484a195e7182e0a8.png" src="../_images/18beee593add4909323fb14b83aabddf46b31871c3c5d3eb484a195e7182e0a8.png" />
</div>
</div>
</section>
</section>
<section id="various-visualizations">
<h2>Various Visualizations<a class="headerlink" href="#various-visualizations" title="Permalink to this heading"></a></h2>
<section id="visualizations-of-the-trajectory-can-be-plotted-at-various-edge-resolutions">
<h3>Visualizations of the trajectory can be plotted at various edge resolutions<a class="headerlink" href="#visualizations-of-the-trajectory-can-be-plotted-at-various-edge-resolutions" title="Permalink to this heading"></a></h3>
</section>
</section>
<section id="edge-bundle-plot">
<h2>Edge bundle plot<a class="headerlink" href="#edge-bundle-plot" title="Permalink to this heading"></a></h2>
<section id="key-parameters-default">
<h3>Key Parameters (default):<a class="headerlink" href="#key-parameters-default" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>alpha_bundle_factor=1</p></li>
<li><p>linewidth_bundle=2</p></li>
<li><p>cmap:str = ‘plasma_r’</p></li>
<li><p>size_scatter:int=2</p></li>
<li><p>alpha_scatter:float = 0.2</p></li>
<li><p>headwidth_bundle=0.1</p></li>
</ul>
</section>
</section>
<section id="animated-stream-plot">
<h2>Animated Stream plot<a class="headerlink" href="#animated-stream-plot" title="Permalink to this heading"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">120</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
<span class="c1">#the streamlines load very slowly through the Notebook, so see the code below to open the file and view the animation properly</span>
<span class="n">via</span><span class="o">.</span><span class="n">animated_streamplot</span><span class="p">(</span><span class="n">v0</span><span class="p">,</span> <span class="n">embedding</span><span class="o">=</span><span class="n">tsnem</span><span class="p">,</span> <span class="n">cmap_stream</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span> <span class="n">scatter_size</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">marker_edgewidth</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> 
                        <span class="n">density_grid</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">segment_length</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">saveto</span><span class="o">=</span><span class="s1">&#39;/home/shobi/Trajectory/Datasets/HumanCD34/HumanCD34_animation_Jan3.gif&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-01-03 09:32:23.524989	Inside animated. File will be saved to location /home/shobi/Trajectory/Datasets/HumanCD34/HumanCD34_animation_Jan3.gif
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  0%|          | 0/250 [00:00&lt;?, ?it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>total number of stream lines 201
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<img alt="../_images/5b918a5c2b1098e5ab53dcc3bab20b24b3755887e20070fd99b034538fe1db3a.png" src="../_images/5b918a5c2b1098e5ab53dcc3bab20b24b3755887e20070fd99b034538fe1db3a.png" />
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&lt;Figure size 1200x720 with 1 Axes&gt;, &lt;AxesSubplot:&gt;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">Image</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;./Figures/HumanCD34_animation.gif&#39;</span><span class="p">,</span><span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">display</span><span class="p">(</span><span class="n">Image</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">read</span><span class="p">()))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="animated-edge-bundle-plot">
<h2>Animated edge bundle plot<a class="headerlink" href="#animated-edge-bundle-plot" title="Permalink to this heading"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">via</span><span class="o">.</span><span class="n">animate_edge_bundle</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">extra_title_text</span><span class="o">=</span><span class="s1">&#39;test animation&#39;</span><span class="p">,</span> <span class="n">n_milestones</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                        <span class="n">saveto</span><span class="o">=</span><span class="s1">&#39;./Figures/human_edgebundle_test.gif&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>making hammerbundle
2022-11-08 16:46:46.082215	Start kmeans milestone
2022-11-08 16:46:50.162267	End kmeans milestone
2022-11-08 16:46:50.230431	Recompute weights
2022-11-08 16:46:50.348908	pruning milestone graph based on recomputed weights
2022-11-08 16:46:50.358479	Graph has 1 connected components before pruning
2022-11-08 16:46:50.358479	Graph has 1 connected components before pruning
2022-11-08 16:46:50.368028	regenerate igraph on pruned edges
2022-11-08 16:46:50.456737	make node dataframe
2022-11-08 16:46:50.472879	Setting numeric label as time_series_labels or other sequential metadata for coloring edges
2022-11-08 16:46:50.592365	Hammer bundling
seg weight 0.1311759587041706
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>MovieWriter imagemagick unavailable; using Pillow instead.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2022-11-08 16:47:46.937861	Finished plotting edge bundle
times series order set [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
number cycles if no time_series labels given 40.0
complete animate
inside update 0 out of 120
inside update 0 out of 120
inside update 1 out of 120
inside update 2 out of 120
inside update 3 out of 120
inside update 4 out of 120
inside update 5 out of 120
inside update 6 out of 120
inside update 7 out of 120
inside update 8 out of 120
inside update 9 out of 120
inside update 10 out of 120
inside update 11 out of 120
inside update 12 out of 120
inside update 13 out of 120
inside update 14 out of 120
inside update 15 out of 120
inside update 16 out of 120
inside update 17 out of 120
inside update 18 out of 120
inside update 19 out of 120
inside update 20 out of 120
inside update 21 out of 120
inside update 22 out of 120
inside update 23 out of 120
inside update 24 out of 120
inside update 25 out of 120
inside update 26 out of 120
inside update 27 out of 120
inside update 28 out of 120
inside update 29 out of 120
inside update 30 out of 120
inside update 31 out of 120
inside update 32 out of 120
inside update 33 out of 120
inside update 34 out of 120
inside update 35 out of 120
inside update 36 out of 120
inside update 37 out of 120
inside update 38 out of 120
inside update 39 out of 120
inside update 40 out of 120
inside update 41 out of 120
inside update 42 out of 120
inside update 43 out of 120
inside update 44 out of 120
inside update 45 out of 120
inside update 46 out of 120
inside update 47 out of 120
inside update 48 out of 120
inside update 49 out of 120
inside update 50 out of 120
inside update 51 out of 120
inside update 52 out of 120
inside update 53 out of 120
inside update 54 out of 120
inside update 55 out of 120
inside update 56 out of 120
inside update 57 out of 120
inside update 58 out of 120
inside update 59 out of 120
inside update 60 out of 120
inside update 61 out of 120
inside update 62 out of 120
inside update 63 out of 120
inside update 64 out of 120
inside update 65 out of 120
inside update 66 out of 120
inside update 67 out of 120
inside update 68 out of 120
inside update 69 out of 120
inside update 70 out of 120
inside update 71 out of 120
inside update 72 out of 120
inside update 73 out of 120
inside update 74 out of 120
inside update 75 out of 120
inside update 76 out of 120
inside update 77 out of 120
inside update 78 out of 120
inside update 79 out of 120
inside update 80 out of 120
inside update 81 out of 120
inside update 82 out of 120
inside update 83 out of 120
inside update 84 out of 120
inside update 85 out of 120
inside update 86 out of 120
inside update 87 out of 120
inside update 88 out of 120
inside update 89 out of 120
inside update 90 out of 120
inside update 91 out of 120
inside update 92 out of 120
inside update 93 out of 120
inside update 94 out of 120
inside update 95 out of 120
inside update 96 out of 120
inside update 97 out of 120
inside update 98 out of 120
inside update 99 out of 120
inside update 100 out of 120
inside update 101 out of 120
inside update 102 out of 120
inside update 103 out of 120
inside update 104 out of 120
inside update 105 out of 120
inside update 106 out of 120
inside update 107 out of 120
inside update 108 out of 120
inside update 109 out of 120
inside update 110 out of 120
inside update 111 out of 120
inside update 112 out of 120
inside update 113 out of 120
inside update 114 out of 120
inside update 115 out of 120
inside update 116 out of 120
inside update 117 out of 120
inside update 118 out of 120
inside update 119 out of 120
saved animation
</pre></div>
</div>
<img alt="../_images/f88bb1efaa184ecf3c49aa8c77031cf12988aadc7f04a62aff45a2576696c772.png" src="../_images/f88bb1efaa184ecf3c49aa8c77031cf12988aadc7f04a62aff45a2576696c772.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">Image</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;/home/user/Github_local/VIA/Figures/human_edgebundle_test.gif&#39;</span><span class="p">,</span><span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">display</span><span class="p">(</span><span class="n">Image</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">read</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/f705d11425b733263e424e82ed3844dcb7381d4fbb9e25be463025dc9531cd1d.png" src="../_images/f705d11425b733263e424e82ed3844dcb7381d4fbb9e25be463025dc9531cd1d.png" />
</div>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="Imaging%20Cytometry%20%28cell%20cycle%29.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">4. Imaging cytometry</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="ViaJupyter_Toy_Disconnected.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">2. Disconnected Trajectories</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2022, shobana stassen
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">3. Via Human Hematopoiesis</a><ul>
<li><a class="reference internal" href="#load-data-from-directory">Load data from directory</a></li>
<li><a class="reference internal" href="#user-s-choice-of-embedding">User’s choice of embedding</a></li>
<li><a class="reference internal" href="#run-via">Run VIA</a><ul>
<li><a class="reference internal" href="#key-parameters">Key Parameters:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#cluster-level-trajectory-plots">Cluster Level Trajectory Plots</a></li>
<li><a class="reference internal" href="#visualise-gene-feature-graph">Visualise gene/feature graph</a></li>
<li><a class="reference internal" href="#trajectory-projection">Trajectory projection</a><ul>
<li><a class="reference internal" href="#id1">Key Parameters:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#probabilistic-pathways-and-memory">Probabilistic pathways and Memory</a><ul>
<li><a class="reference internal" href="#id2">Key Parameters:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#gene-dynamics">Gene Dynamics</a><ul>
<li><a class="reference internal" href="#id3">Key Parameters</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id4">Gene Dynamics</a><ul>
<li><a class="reference internal" href="#id5">Key Parameters</a></li>
</ul>
</li>
<li><a class="reference internal" href="#various-visualizations">Various Visualizations</a><ul>
<li><a class="reference internal" href="#visualizations-of-the-trajectory-can-be-plotted-at-various-edge-resolutions">Visualizations of the trajectory can be plotted at various edge resolutions</a></li>
</ul>
</li>
<li><a class="reference internal" href="#edge-bundle-plot">Edge bundle plot</a><ul>
<li><a class="reference internal" href="#key-parameters-default">Key Parameters (default):</a></li>
</ul>
</li>
<li><a class="reference internal" href="#animated-stream-plot">Animated Stream plot</a></li>
<li><a class="reference internal" href="#animated-edge-bundle-plot">Animated edge bundle plot</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/scripts/furo.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script defer="defer" src="https://unpkg.com/@popperjs/core@2"></script>
    <script defer="defer" src="https://unpkg.com/tippy.js@6"></script>
    <script defer="defer" src="../_static/tippy\notebooks\ViaJupyter_scRNA_Hematopoiesis.342ed35b-ac0a-4c16-9326-505cc190ffec.js"></script>
    </body>
</html>