<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="3. Via Human Hematopoiesis" href="ViaJupyter_scRNA_Hematopoiesis.html" /><link rel="prev" title="1. Basic workflow" href="ViaJupyter_Toy_Multifurcating.html" />

    <!-- Generated with Sphinx 5.1.1 and Furo 2023.03.27 -->
        <title>2. Disconnected Trajectories - pyvia documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?digest=fad236701ea90a88636c2a8c73b44ae642ed2a53" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">pyvia  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">pyvia  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">General</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../pyVia-home.html">StaVia - Multi-Omic Single-Cell Cartography for Spatial and Temporal Atlases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Release%20History.html">Release History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Tutorial%20Video.html">VIA Tutorial Videos</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Parameters%20and%20Attributes.html">Parameters and Attributes</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../api/index.html">API</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../api/pyVIA%20Core.html">pyVIA core</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/Plotting.html">Plotting</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.animate_atlas.html">pyVIA.plotting_via.animate_atlas</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.animate_streamplot.html">pyVIA.plotting_via.animate_streamplot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.get_gene_expression.html">pyVIA.plotting_via.get_gene_expression</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_atlas_view.html">pyVIA.plotting_via.plot_atlas_view</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_differentiation_flow.html">pyVIA.plotting_via.plot_differentiation_flow</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_gene_trend_heatmaps.html">pyVIA.plotting_via.plot_gene_trend_heatmaps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_piechart_viagraph.html">pyVIA.plotting_via.plot_piechart_viagraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_population_composition.html">pyVIA.plotting_via.plot_population_composition</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_sc_lineage_probability.html">pyVIA.plotting_via.plot_sc_lineage_probability</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_scatter.html">pyVIA.plotting_via.plot_scatter</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_trajectory_curves.html">pyVIA.plotting_via.plot_trajectory_curves</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.plot_viagraph.html">pyVIA.plotting_via.plot_viagraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_streamplot.html">pyVIA.plotting_via.via_streamplot</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/Utils.html">Util Functions for plotting</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.make_edgebundle_milestone.html">pyVIA.plotting_via.make_edgebundle_milestone</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_atlas_emb.html">pyVIA.plotting_via.via_atlas_emb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_forcelayout.html">pyVIA.plotting_via.via_forcelayout</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.plotting_via/pyVIA.plotting_via.via_mds.html">pyVIA.plotting_via.via_mds</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/Datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.cell_cycle.html">VIA.datasets_via.cell_cycle</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.cell_cycle_cyto_data.html">VIA.datasets_via.cell_cycle_cyto_data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.embryoid_body.html">VIA.datasets_via.embryoid_body</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.scATAC_hematopoiesis.html">VIA.datasets_via.scATAC_hematopoiesis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.scRNA_hematopoiesis.html">VIA.datasets_via.scRNA_hematopoiesis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.toy_disconnected.html">VIA.datasets_via.toy_disconnected</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/_autosummary/VIA.datasets_via/VIA.datasets_via.toy_multifurcating.html">VIA.datasets_via.toy_multifurcating</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../Basic%20Example%20Code.html">Examples for installation checking</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Via 2.0 Atlas View Gallery</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Atlas%20view%20examples.html">Via2.0 Atlas views</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mouse_to_pup_atlas.html">Large Mouse Embryo to Pup Developmental Atlas</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Via2%20Atlas%20Animation.html">Via 2.0 Atlas Animations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Via%20Atlas%20View%20for%20Spatial%20omics.html">Via2.0 Atlas views for Spatial Omics data</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Via 2.0 Cartography</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Via2.0%20Cartographic%20Mouse%20Gastrualation.html">1. Via 2.0 Cartography on Mouse Gastrulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Zebrahub%20TI%20tutorial.html">2. Via 2.0 Cartography on Zebrahub (Trajectory Inference)</a></li>
<li class="toctree-l1"><a class="reference internal" href="Zebrahub_tutorial_visualization.html">3. Via 2.0 Cartography on Zebrahub (Visualization)</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials Via</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_Toy_Multifurcating.html">1. Basic workflow</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">2. Disconnected Trajectories</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_scRNA_Hematopoiesis.html">3. Via Human Hematopoiesis</a></li>
<li class="toctree-l1"><a class="reference internal" href="Imaging%20Cytometry%20%28cell%20cycle%29.html">4. Imaging cytometry</a></li>
<li class="toctree-l1"><a class="reference internal" href="mESC_timeseries.html">5. Using time-series metadata (mESC Cytof)</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_Pancreas_RNAvelocity.html">6. Using RNA-velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="ViaJupyter_scRNAVelocity_hematopoiesis.html">7. Bone marrow with RNA-velocity</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section class="tex2jax_ignore mathjax_ignore" id="disconnected-trajectories">
<h1>2. Disconnected Trajectories<a class="headerlink" href="#disconnected-trajectories" title="Permalink to this heading"></a></h1>
<p>This vignette shows the basic workflow of Via for disconnected trajectories and how to access the different visualizations after running the computations. We use a toydataset with two disconnected components to show different ways to set the root conditions depending on the availability of labels and root level information apriori.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pyVIA.core</span> <span class="k">as</span> <span class="nn">via</span>
<span class="kn">import</span> <span class="nn">pyVIA.datasets_via</span> <span class="k">as</span> <span class="nn">datasets_via</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span> 

<span class="kn">import</span> <span class="nn">umap</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: <span class=" -Color -Color-Bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: <span class=" -Color -Color-Bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/umap/distances.py:1086: NumbaDeprecationWarning: <span class=" -Color -Color-Bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/umap/umap_.py:660: NumbaDeprecationWarning: <span class=" -Color -Color-Bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
</pre></div>
</div>
</div>
</div>
<section id="load-data">
<h2>Load data<a class="headerlink" href="#load-data" title="Permalink to this heading"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Since this is a toy dataset, we known the labels of each cell in terms of component and milestone. So T1_M1 is Trajectory1_Milestone1 (the milestones are not listed in any particular order, so M2 is not neccessarily earlier than M5)</span>

<span class="c1">#make anndata object with counts and group_ids </span>
<span class="n">adata_counts</span> <span class="o">=</span> <span class="n">datasets_via</span><span class="o">.</span><span class="n">toy_disconnected</span><span class="p">(</span><span class="n">foldername</span><span class="o">=</span><span class="s1">&#39;./&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_counts</span><span class="p">)</span>
<span class="n">true_label</span> <span class="o">=</span> <span class="n">adata_counts</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;group_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="c1">#define parameters</span>
<span class="c1">#for more generic purposes, dataset can be =&#39;group&#39; if root is a group level assignment corresponding to a group in true_label. or left as &#39;&#39;, in which case root_user = [cell index] where cell index is an int</span>
<span class="n">ncomps</span><span class="p">,</span> <span class="n">knn</span><span class="p">,</span> <span class="n">random_seed</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">root_user</span>  <span class="o">=</span><span class="mi">30</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span><span class="s1">&#39;toy&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;T1_M1&#39;</span><span class="p">,</span> <span class="s1">&#39;T2_M1&#39;</span><span class="p">]</span> 

<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_counts</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="n">ncomps</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/anndata/_core/anndata.py:117: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 1000 × 1000
    obs: &#39;group_id&#39;, &#39;true_time&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="initialize-and-run-via">
<h2>Initialize and run VIA<a class="headerlink" href="#initialize-and-run-via" title="Permalink to this heading"></a></h2>
<p>In this example we provide two group level root cells (T1_M1 and T2_M1). We therefore need to set dataset=’group’</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd"># knn: number of nearest neighbors for the KNN-graph </span>
<span class="sd"># clustergraph_pruning: typically 0-3, the number of standard deviations for clustergraph pruning. level of pruning applied to the clustergraph carried forward for pseudotime and lineage probability calculations. Larger means less pruning  </span>
<span class="sd"># true_label: list of annotations. if these are unavailable then leave as default None</span>
<span class="sd"># dist_std_local: local pruning on the singlecell graph used for clustering step, 1 is quite quite minimal level of pruning.  </span>
<span class="sd"># too_big_factor: default 0.3. Clusters that are bigger than 30% of the entire cell population will be re-clustered. If you only plan to do one iteration of VIA before visualizing TI, consider lowering this to 0.1-0.2.  </span>
<span class="sd"># preserve_disconnected: if you suspect disconnected trajectories, set this to False</span>
<span class="sd"># dataset = &#39;group&#39;: works for any dataset where the root is provided as a group ID in the true_label. e.g. if root = [&#39;Day0&#39;], then the root will be selected as a cluster which has root like properties and is largely made up of &#39;Day0&#39; cells. </span>
<span class="sd">            #If you wish to run a second finegrained iteration of VIA on a disconnected dataset, but uses the terminal clusters from the previous coarse run, then the dataset should be set as &#39;toy&#39; and the true_labels formatted as &quot;T1_M1&quot; where Tx is the x&#39;th component and My is the y&#39;th cluster.</span>
<span class="sd">            # otherwise for disconnected datasets, you can just run a single iteration of VIA where you vary too_big_factor and other parameters to change granularity and set the dataset as &#39;&#39; or &#39;group&#39; depending on the root_user type as [groupX, groupZ ...] or [cell indices...]  </span>
<span class="sd">            # dataset, can be left as default value &#39;&#39;, but then the root must be a cell index corresponding to a best guess. </span>
<span class="sd"># visual_cluster_graph_pruning: this impacts the level of pruning done to the clustergraph before visualization. Does not impact underlying computations  </span>
<span class="sd"># Max_visual_outgoing_edges: controls number of edges extending from each node</span>
<span class="sd"># Embedding_type can be set to &#39;via-umap&#39; or &#39;via-mds&#39;. via-mds is very fast and produces more continous plots. &#39;via-umap&#39; is initialized by the viagraph. To have more control of the plotting configurations/ </span>
<span class="sd"> and the generation of these embeddings, the functions via.run_umap_hnsw(via_object = v0)and via.via_mds(via_object=v0) can be called after running base case via v0=via.VIA()</span>

<span class="sd">&#39;&#39;&#39;</span>


<span class="n">v0</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">VIA</span><span class="p">(</span><span class="n">adata_counts</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="n">ncomps</span><span class="p">],</span> <span class="n">true_label</span><span class="p">,</span> <span class="n">edgepruning_clustering_resolution</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">edgepruning_clustering_resolution_local</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
             <span class="n">knn</span><span class="o">=</span><span class="n">knn</span><span class="p">,</span> <span class="n">cluster_graph_pruning</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">too_big_factor</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">root_user</span><span class="o">=</span><span class="n">root_user</span><span class="p">,</span> <span class="n">preserve_disconnected</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span><span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">)</span>
<span class="n">v0</span><span class="o">.</span><span class="n">run_VIA</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 16:14:16.244889	Running VIA over input data of 1000 (samples) x 30 (features)
2023-10-12 16:14:16.244949	Knngraph has 10 neighbors
2023-10-12 16:14:16.596490	Finished global pruning of 10-knn graph used for clustering at level of 0.15. Kept 50.2 % of edges. 
2023-10-12 16:14:16.602625	Number of connected components used for clustergraph  is 2
2023-10-12 16:14:16.621793	Commencing community detection
2023-10-12 16:14:16.637361	Finished running Leiden algorithm. Found 35 clusters.
2023-10-12 16:14:16.638715	Merging 20 very small clusters (&lt;10)
2023-10-12 16:14:16.639771	Finished detecting communities. Found 15 communities
2023-10-12 16:14:16.640069	Making cluster graph. Global cluster graph pruning level: 1
2023-10-12 16:14:16.644297	Graph has 2 connected components before pruning
2023-10-12 16:14:16.646449	Graph has 2 connected components after pruning
2023-10-12 16:14:16.646706	Graph has 2 connected components after reconnecting
2023-10-12 16:14:16.647272	0.0% links trimmed from local pruning relative to start
2023-10-12 16:14:16.649630	component number 0 out of  [0, 1]
2023-10-12 16:14:16.651277	group root method
2023-10-12 16:14:16.651293	for component 0, the root is T1_M1 and ri T1_M1
2023-10-12 16:14:16.651300	group root method
cluster 0 has majority T1_M3
cluster 4 has majority T1_M3
cluster 7 has majority T1_M1
2023-10-12 16:14:16.651866	New root is 7 and majority T1_M1
cluster 11 has majority T1_M2
cluster 13 has majority T1_M4
2023-10-12 16:14:16.651961	Computing lazy-teleporting expected hitting times
2023-10-12 16:14:16.875755	 ended all multiprocesses, will retrieve and reshape
try rw2 hitting times setup
start computing walks with rw2 method
g.indptr.size, 6
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pecanpy/graph.py:90: UserWarning: WARNING: Implicitly set node IDs to the canonical node ordering due to missing IDs field in the raw CSR npz file. This warning message can be suppressed by setting implicit_ids to True in the read_npz function call, or by setting the --implicit_ids flag in the CLI
  warnings.warn(
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "6d992c0c712a4c92a2f2c35b1a50375d"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>memory for rw2 hittings times  2. Using rw2 based pt
do scaling of pt
2023-10-12 16:14:22.958974	Identifying terminal clusters corresponding to unique lineages...
2023-10-12 16:14:22.958999	Closeness:[2, 3, 4]
2023-10-12 16:14:22.959013	Betweenness:[2, 3, 4]
2023-10-12 16:14:22.959022	Out Degree:[3, 4]
2023-10-12 16:14:22.959391	Terminal clusters corresponding to unique lineages in this component are [3, 4] 
TESTING rw2_lineage probability at memory 5
testing rw2 lineage probability at memory 5
g.indptr.size, 6
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pecanpy/graph.py:90: UserWarning: WARNING: Implicitly set node IDs to the canonical node ordering due to missing IDs field in the raw CSR npz file. This warning message can be suppressed by setting implicit_ids to True in the read_npz function call, or by setting the --implicit_ids flag in the CLI
  warnings.warn(
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "9687915126704e01885f7c218f400439"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 16:14:28.281297	 Cluster or terminal cell fate 3 is reached 944.0 times
2023-10-12 16:14:28.325328	 Cluster or terminal cell fate 4 is reached 342.0 times
2023-10-12 16:14:28.328641	component number 1 out of  [0, 1]
2023-10-12 16:14:28.331618	group root method
2023-10-12 16:14:28.331665	group root method
2023-10-12 16:14:28.331675	for component 1, the root is T2_M1 and ri T2_M1
cluster 1 has majority T2_M5
cluster 2 has majority T2_M2
cluster 3 has majority T2_M1
2023-10-12 16:14:28.333632	New root is 3 and majority T2_M1
cluster 5 has majority T2_M4
cluster 6 has majority T2_M3
cluster 8 has majority T2_M6
cluster 9 has majority T2_M2
cluster 10 has majority T2_M4
cluster 12 has majority T2_M5
cluster 14 has majority T2_M3
2023-10-12 16:14:28.334646	Computing lazy-teleporting expected hitting times
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pyVIA/core.py:2447: RuntimeWarning: invalid value encountered in divide
  bp_array = bp_array / bp_array.sum(axis=1)[:, None]  # row normalization at the cluster level
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 16:14:28.678470	 ended all multiprocesses, will retrieve and reshape
try rw2 hitting times setup
start computing walks with rw2 method
g.indptr.size, 11
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pecanpy/graph.py:90: UserWarning: WARNING: Implicitly set node IDs to the canonical node ordering due to missing IDs field in the raw CSR npz file. This warning message can be suppressed by setting implicit_ids to True in the read_npz function call, or by setting the --implicit_ids flag in the CLI
  warnings.warn(
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "ab5da424ed9e4afba397914ca4bffb19"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>memory for rw2 hittings times  2. Using rw2 based pt
do scaling of pt
2023-10-12 16:14:32.733637	Identifying terminal clusters corresponding to unique lineages...
2023-10-12 16:14:32.733667	Closeness:[1, 5, 6, 7]
2023-10-12 16:14:32.733683	Betweenness:[1, 4, 5, 6, 7]
2023-10-12 16:14:32.733696	Out Degree:[1, 5, 6, 7]
2023-10-12 16:14:32.734192	Terminal clusters corresponding to unique lineages in this component are [1, 5, 6, 7] 
TESTING rw2_lineage probability at memory 5
testing rw2 lineage probability at memory 5
g.indptr.size, 11
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pecanpy/graph.py:90: UserWarning: WARNING: Implicitly set node IDs to the canonical node ordering due to missing IDs field in the raw CSR npz file. This warning message can be suppressed by setting implicit_ids to True in the read_npz function call, or by setting the --implicit_ids flag in the CLI
  warnings.warn(
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "2de3481ea43d48389e2c183daa9bfe5b"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 16:14:36.444718	 Cluster or terminal cell fate 1 is reached 968.0 times
2023-10-12 16:14:36.504059	 Cluster or terminal cell fate 5 is reached 172.0 times
2023-10-12 16:14:36.553546	 Cluster or terminal cell fate 6 is reached 509.0 times
2023-10-12 16:14:36.606769	 Cluster or terminal cell fate 7 is reached 973.0 times
2023-10-12 16:14:36.613832	There are (6) terminal clusters corresponding to unique lineages {11: &#39;T1_M2&#39;, 13: &#39;T1_M4&#39;, 2: &#39;T2_M2&#39;, 8: &#39;T2_M6&#39;, 9: &#39;T2_M2&#39;, 10: &#39;T2_M4&#39;}
2023-10-12 16:14:36.613881	Begin projection of pseudotime and lineage likelihood
2023-10-12 16:14:36.787228	Cluster graph layout based on forward biasing
2023-10-12 16:14:36.789109	Starting make edgebundle viagraph...
2023-10-12 16:14:36.789135	Make via clustergraph edgebundle
2023-10-12 16:14:38.597314	Hammer dims: Nodes shape: (15, 2) Edges shape: (28, 3)
2023-10-12 16:14:38.598162	Graph has 2 connected components before pruning
2023-10-12 16:14:38.600122	Graph has 5 connected components after pruning
2023-10-12 16:14:38.602974	Graph has 2 connected components after reconnecting
2023-10-12 16:14:38.603480	7.1% links trimmed from local pruning relative to start
2023-10-12 16:14:38.603502	35.7% links trimmed from global pruning relative to start
2023-10-12 16:14:38.609519	Time elapsed 22.2 seconds
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/scipy/sparse/_index.py:103: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.
  self._set_intXint(row, col, x.flat[0])
</pre></div>
</div>
</div>
</div>
</section>
<section id="draw-the-viagraph">
<h2>Draw the viagraph<a class="headerlink" href="#draw-the-viagraph" title="Permalink to this heading"></a></h2>
<p>Left-hand side axis is colored by values in <code class="docutils literal notranslate"><span class="pre">via_object.true_label</span></code>. Coloring controlled by <code class="docutils literal notranslate"><span class="pre">cmap_piechart</span></code>. To color these piecharts by some other reference labels, pass a list into the argument <code class="docutils literal notranslate"><span class="pre">reference_labels</span></code> of the <code class="docutils literal notranslate"><span class="pre">plot_piechart_viagraph</span></code>
Right-hand side axis is colored by the pseudotime</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span><span class="n">ax</span><span class="p">,</span><span class="n">ax2</span><span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_piechart_viagraph</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">cmap_piechart</span><span class="o">=</span><span class="s1">&#39;rainbow&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;plasma&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pyVIA/plotting_via.py:3092: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  sct = ax.scatter(node_pos[:, 0], node_pos[:, 1],
</pre></div>
</div>
<img alt="../_images/67d0f5f09a823e40b3850f961ee7edf27bb38a2a94c688d0fbb2447af8e1aea2.png" src="../_images/67d0f5f09a823e40b3850f961ee7edf27bb38a2a94c688d0fbb2447af8e1aea2.png" />
</div>
</div>
</section>
<section id="initial-state">
<h2>Initial state<a class="headerlink" href="#initial-state" title="Permalink to this heading"></a></h2>
<section id="roots-can-also-be-provided-as-cell-indices">
<h3>Roots can also be provided as cell indices<a class="headerlink" href="#roots-can-also-be-provided-as-cell-indices" title="Permalink to this heading"></a></h3>
<p>If the roots are provided as indices, the dataset parameter should be set to dataset=’’. If the roots are group level labels, then dataset = ‘group’. This tells via which approach to take for identifying the root.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1">#consider providing the roots as indices corresponding to the desired start cells</span>
<span class="n">root_user</span><span class="p">,</span> <span class="n">dataset</span> <span class="o">=</span> <span class="p">[</span><span class="mi">136</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;&#39;</span>
<span class="n">v0</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">VIA</span><span class="p">(</span><span class="n">adata_counts</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="n">ncomps</span><span class="p">],</span> <span class="n">true_label</span><span class="p">,</span> <span class="n">edgepruning_clustering_resolution</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">edgepruning_clustering_resolution_local</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
             <span class="n">knn</span><span class="o">=</span><span class="n">knn</span><span class="p">,</span> <span class="n">cluster_graph_pruning</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>             <span class="n">too_big_factor</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">root_user</span><span class="o">=</span><span class="n">root_user</span><span class="p">,</span> <span class="n">preserve_disconnected</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">)</span>
<span class="n">v0</span><span class="o">.</span><span class="n">run_VIA</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 14:15:46.626276	Running VIA over input data of 1000 (samples) x 30 (features)
2023-10-12 14:15:46.626328	Knngraph has 10 neighbors
2023-10-12 14:15:46.993923	Finished global pruning of 10-knn graph used for clustering at level of 0.15. Kept 50.2 % of edges. 
2023-10-12 14:15:47.002606	Number of connected components used for clustergraph  is 2
2023-10-12 14:15:47.032302	Commencing community detection
2023-10-12 14:15:47.046669	Finished running Leiden algorithm. Found 35 clusters.
2023-10-12 14:15:47.047752	Merging 20 very small clusters (&lt;10)
2023-10-12 14:15:47.048800	Finished detecting communities. Found 15 communities
2023-10-12 14:15:47.049094	Making cluster graph. Global cluster graph pruning level: 1
2023-10-12 14:15:47.053235	Graph has 2 connected components before pruning
2023-10-12 14:15:47.055273	Graph has 2 connected components after pruning
2023-10-12 14:15:47.055491	Graph has 2 connected components after reconnecting
2023-10-12 14:15:47.056172	0.0% links trimmed from local pruning relative to start
2023-10-12 14:15:47.059469	component number 0 out of  [0, 1]
2023-10-12 14:15:47.061911	The root index, 136 provided by the user belongs to cluster number 7 and corresponds to cell type T1_M1
2023-10-12 14:15:47.063233	Computing lazy-teleporting expected hitting times
2023-10-12 14:15:47.270284	 ended all multiprocesses, will retrieve and reshape
try rw2 hitting times setup
start computing walks with rw2 method
g.indptr.size, 6
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pecanpy/graph.py:90: UserWarning: WARNING: Implicitly set node IDs to the canonical node ordering due to missing IDs field in the raw CSR npz file. This warning message can be suppressed by setting implicit_ids to True in the read_npz function call, or by setting the --implicit_ids flag in the CLI
  warnings.warn(
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "59356dbde79c447eafe00dc9a89decce"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>memory for rw2 hittings times  2. Using rw2 based pt
do scaling of pt
2023-10-12 14:15:51.415494	Identifying terminal clusters corresponding to unique lineages...
2023-10-12 14:15:51.415532	Closeness:[2, 3, 4]
2023-10-12 14:15:51.415551	Betweenness:[2, 3, 4]
2023-10-12 14:15:51.415573	Out Degree:[3, 4]
2023-10-12 14:15:51.416230	Terminal clusters corresponding to unique lineages in this component are [3, 4] 
TESTING rw2_lineage probability at memory 5
testing rw2 lineage probability at memory 5
g.indptr.size, 6
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pecanpy/graph.py:90: UserWarning: WARNING: Implicitly set node IDs to the canonical node ordering due to missing IDs field in the raw CSR npz file. This warning message can be suppressed by setting implicit_ids to True in the read_npz function call, or by setting the --implicit_ids flag in the CLI
  warnings.warn(
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "b90ad689a674490a9009ff3b031a15da"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 14:15:55.322206	 Cluster or terminal cell fate 3 is reached 946.0 times
2023-10-12 14:15:55.368623	 Cluster or terminal cell fate 4 is reached 374.0 times
2023-10-12 14:15:55.372102	component number 1 out of  [0, 1]
2023-10-12 14:15:55.374228	The root index, 4 provided by the user belongs to cluster number 3 and corresponds to cell type T2_M1
2023-10-12 14:15:55.377405	Computing lazy-teleporting expected hitting times
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pyVIA/core.py:2447: RuntimeWarning: invalid value encountered in divide
  bp_array = bp_array / bp_array.sum(axis=1)[:, None]  # row normalization at the cluster level
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 14:15:55.715180	 ended all multiprocesses, will retrieve and reshape
try rw2 hitting times setup
start computing walks with rw2 method
g.indptr.size, 11
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pecanpy/graph.py:90: UserWarning: WARNING: Implicitly set node IDs to the canonical node ordering due to missing IDs field in the raw CSR npz file. This warning message can be suppressed by setting implicit_ids to True in the read_npz function call, or by setting the --implicit_ids flag in the CLI
  warnings.warn(
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "eaeba3a9d30f48f4bfad4077379738c4"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>memory for rw2 hittings times  2. Using rw2 based pt
do scaling of pt
2023-10-12 14:15:59.380054	Identifying terminal clusters corresponding to unique lineages...
2023-10-12 14:15:59.380104	Closeness:[1, 5, 6, 7]
2023-10-12 14:15:59.380127	Betweenness:[1, 4, 5, 6, 7]
2023-10-12 14:15:59.380145	Out Degree:[1, 5, 6, 7]
2023-10-12 14:15:59.380804	Terminal clusters corresponding to unique lineages in this component are [1, 5, 6, 7] 
TESTING rw2_lineage probability at memory 5
testing rw2 lineage probability at memory 5
g.indptr.size, 11
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pecanpy/graph.py:90: UserWarning: WARNING: Implicitly set node IDs to the canonical node ordering due to missing IDs field in the raw CSR npz file. This warning message can be suppressed by setting implicit_ids to True in the read_npz function call, or by setting the --implicit_ids flag in the CLI
  warnings.warn(
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "3ea26b534e27467d8c4adc3f03f8d576"}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 14:16:03.077495	 Cluster or terminal cell fate 1 is reached 974.0 times
2023-10-12 14:16:03.133979	 Cluster or terminal cell fate 5 is reached 198.0 times
2023-10-12 14:16:03.181236	 Cluster or terminal cell fate 6 is reached 501.0 times
2023-10-12 14:16:03.245110	 Cluster or terminal cell fate 7 is reached 982.0 times
2023-10-12 14:16:03.253053	There are (6) terminal clusters corresponding to unique lineages {11: &#39;T1_M2&#39;, 13: &#39;T1_M4&#39;, 2: &#39;T2_M2&#39;, 8: &#39;T2_M6&#39;, 9: &#39;T2_M2&#39;, 10: &#39;T2_M4&#39;}
2023-10-12 14:16:03.253101	Begin projection of pseudotime and lineage likelihood
2023-10-12 14:16:03.451333	Cluster graph layout based on forward biasing
2023-10-12 14:16:03.454393	Starting make edgebundle viagraph...
2023-10-12 14:16:03.454426	Make via clustergraph edgebundle
2023-10-12 14:16:03.673289	Hammer dims: Nodes shape: (15, 2) Edges shape: (28, 3)
2023-10-12 14:16:03.674649	Graph has 2 connected components before pruning
2023-10-12 14:16:03.677158	Graph has 5 connected components after pruning
2023-10-12 14:16:03.681309	Graph has 2 connected components after reconnecting
2023-10-12 14:16:03.682096	7.1% links trimmed from local pruning relative to start
2023-10-12 14:16:03.682122	35.7% links trimmed from global pruning relative to start
2023-10-12 14:16:03.689517	Time elapsed 16.9 seconds
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/scipy/sparse/_index.py:103: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.
  self._set_intXint(row, col, x.flat[0])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span> <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_piechart_viagraph</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">cmap_piechart</span><span class="o">=</span><span class="s1">&#39;jet&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">pie_size_scale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">size_node_notpiechart</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/PycharmProjects/Via2_May2023_py310/plotting_via.py:3104: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  sct = ax.scatter(node_pos[:, 0], node_pos[:, 1],
</pre></div>
</div>
<img alt="../_images/8cb1219b62ccaf59a04c3c75d72faabb571f3ddf84c86eae0d4ceba5d0037d13.png" src="../_images/8cb1219b62ccaf59a04c3c75d72faabb571f3ddf84c86eae0d4ceba5d0037d13.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">e1</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">via_atlas_emb</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">n_epochs</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">init_pos</span><span class="o">=</span><span class="s1">&#39;via&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">e1</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">f</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span><span class="n">via</span><span class="o">.</span><span class="n">plot_scatter</span><span class="p">(</span><span class="n">embedding</span><span class="o">=</span><span class="n">e1</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">v0</span><span class="o">.</span><span class="n">single_cell_pt_markov</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;via init&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>X-input (1000, 30)
len membership and n_cells 1000 1000
n cell 1000
2023-10-12 16:15:10.827802	Computing embedding on sc-Viagraph
2023-10-12 16:15:10.828568	using via cluster graph to initialize embedding
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "875101d7b5b342d4a3d6e52309f80af5"}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(1000, 2)
</pre></div>
</div>
<img alt="../_images/3f3abd9e20d06b26cbfe24936965dea85390d1c3b5beec0cb95ac4164c409a08.png" src="../_images/3f3abd9e20d06b26cbfe24936965dea85390d1c3b5beec0cb95ac4164c409a08.png" />
</div>
</div>
</section>
</section>
<section id="plot-atlas-view">
<h2>Plot Atlas View<a class="headerlink" href="#plot-atlas-view" title="Permalink to this heading"></a></h2>
<p>Plot the directed connectivity combined with high resolution of spatial layout. The edges are directed by the inferred Pseudotime</p>
<ul class="simple">
<li><p>colored by pseudotime</p></li>
<li><p>by celltype</p></li>
<li><p>lineage pathways from root to terminal cell fate</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">v0</span><span class="o">.</span><span class="n">embedding</span> <span class="o">=</span> <span class="n">e1</span>
<span class="n">f</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_atlas_view</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">n_milestones</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">linewidth_bundle</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">headwidth_bundle</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">f</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 16:15:53.714670	Computing Edges
2023-10-12 16:15:53.715100	Start finding milestones
2023-10-12 16:15:54.570913	End milestones with 150
2023-10-12 16:15:54.573976	Recompute weights
2023-10-12 16:15:54.596576	pruning milestone graph based on recomputed weights
2023-10-12 16:15:54.597986	Graph has 2 connected components before pruning
2023-10-12 16:15:54.598929	Graph has 2 connected components after pruning
2023-10-12 16:15:54.599192	Graph has 2 connected components after reconnecting
2023-10-12 16:15:54.601260	regenerate igraph on pruned edges
2023-10-12 16:15:54.610036	Setting numeric label as time_series_labels or other sequential metadata for coloring edges
2023-10-12 16:15:54.622472	Making smooth edges
inside add sc embedding second if
</pre></div>
</div>
<img alt="../_images/4be420ef3cf0deb8ebb93d8521919ba133e62015a4577ac67ef9d891f53a267a.png" src="../_images/4be420ef3cf0deb8ebb93d8521919ba133e62015a4577ac67ef9d891f53a267a.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_atlas_view</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">linewidth_bundle</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">headwidth_bundle</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">n_milestones</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">sc_labels_expression</span><span class="o">=</span><span class="n">true_label</span><span class="p">,</span> <span class="n">text_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sc_labels</span><span class="o">=</span><span class="n">true_label</span><span class="p">,</span> <span class="n">fontsize_labels</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;jet&#39;</span><span class="p">,</span> <span class="n">colorbar_legend</span><span class="o">=</span><span class="s1">&#39;celltype&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 16:22:21.600289	Computing Edges
2023-10-12 16:22:21.600369	Start finding milestones
2023-10-12 16:22:22.458731	End milestones with 150
2023-10-12 16:22:22.460412	Recompute weights
2023-10-12 16:22:22.474079	pruning milestone graph based on recomputed weights
2023-10-12 16:22:22.475248	Graph has 2 connected components before pruning
2023-10-12 16:22:22.476176	Graph has 2 connected components after pruning
2023-10-12 16:22:22.476436	Graph has 2 connected components after reconnecting
2023-10-12 16:22:22.478479	regenerate igraph on pruned edges
2023-10-12 16:22:22.487920	Setting numeric label as time_series_labels or other sequential metadata for coloring edges
2023-10-12 16:22:22.504859	Making smooth edges
inside add sc embedding second if
</pre></div>
</div>
<img alt="../_images/0bf8f33b9687eeac6f2ec4abd48d30fda30165213e9c18021259660977492492.png" src="../_images/0bf8f33b9687eeac6f2ec4abd48d30fda30165213e9c18021259660977492492.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">via</span><span class="o">.</span><span class="n">plot_atlas_view</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">linewidth_bundle</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">headwidth_bundle</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">lineage_pathway</span><span class="o">=</span><span class="p">[</span><span class="mi">11</span><span class="p">,</span><span class="mi">13</span><span class="p">,</span><span class="mi">8</span><span class="p">])</span>
<span class="n">f</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 16:19:35.609089	Computing Edges
2023-10-12 16:19:35.609148	Start finding milestones
2023-10-12 16:19:36.431670	End milestones with 150
2023-10-12 16:19:36.433421	Recompute weights
2023-10-12 16:19:36.449687	pruning milestone graph based on recomputed weights
2023-10-12 16:19:36.450963	Graph has 2 connected components before pruning
2023-10-12 16:19:36.451904	Graph has 2 connected components after pruning
2023-10-12 16:19:36.452170	Graph has 2 connected components after reconnecting
2023-10-12 16:19:36.454188	regenerate igraph on pruned edges
2023-10-12 16:19:36.463885	Setting numeric label as time_series_labels or other sequential metadata for coloring edges
2023-10-12 16:19:36.471978	Making smooth edges
location of 11 is at [0] and 0
location of 13 is at [1] and 1
location of 8 is at [3] and 3
</pre></div>
</div>
<img alt="../_images/0a56be9269f726afbf1774ad48c3f4b39d413d74bc1298e3b508190e5e92bc90.png" src="../_images/0a56be9269f726afbf1774ad48c3f4b39d413d74bc1298e3b508190e5e92bc90.png" />
</div>
</div>
</section>
<section id="fine-grained-vector-field">
<h2>Fine-grained vector field<a class="headerlink" href="#fine-grained-vector-field" title="Permalink to this heading"></a></h2>
<p>Draw a stream plot that shows the directionality at a higher resolution</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1">#default parameters:</span>
<span class="c1"># scatter_size=500</span>
<span class="c1"># scatter_alpha=0.5</span>
<span class="c1"># marker_edgewidth=0.1, outline of scatter points</span>
<span class="c1"># density_stream = 2, spacing between field lines</span>
<span class="c1"># smooth_transition=1,smoothing of transition matrix over neighbors of neighbors</span>
<span class="c1"># color_scheme = &#39;annotation&#39;, corresponds to true_labels provided. Can also be set to &#39;cluster&#39; and &#39;time&#39;</span>
<span class="c1"># add_outline_clusters=False</span>
<span class="n">v0</span><span class="o">.</span><span class="n">embedding</span> <span class="o">=</span> <span class="n">e1</span>
<span class="n">f</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span><span class="n">via</span><span class="o">.</span><span class="n">via_streamplot</span><span class="p">(</span><span class="n">v0</span><span class="p">)</span>
<span class="n">f</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pyVIA/core.py:1318: RuntimeWarning: Mean of empty slice.
  V_emb[i] = probs.dot(dX) - probs.mean() * dX.sum(0)
/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/numpy/core/_methods.py:192: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
</pre></div>
</div>
<img alt="../_images/3fa11002680388ec349836c4b67b9052b90799db36b364dd53b76462a4df7434.png" src="../_images/3fa11002680388ec349836c4b67b9052b90799db36b364dd53b76462a4df7434.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">120</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
<span class="n">via</span><span class="o">.</span><span class="n">animated_streamplot</span><span class="p">(</span><span class="n">v0</span><span class="p">,</span> <span class="n">scatter_size</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">density_grid</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">saveto</span><span class="o">=</span><span class="s1">&#39;/home/user/Trajectory/Datasets/Toy4/toy4_animation_.gif&#39;</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/shobi/anaconda3/envs/ViaEnv/lib/python3.7/site-packages/pyVIA/core.py:1931: RuntimeWarning: Mean of empty slice.
  V_emb[i] = probs.dot(dX) - probs.mean() * dX.sum(0)
/home/shobi/anaconda3/envs/ViaEnv/lib/python3.7/site-packages/numpy/core/_methods.py:170: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2022-07-09 16:37:52.063276	Inside animated. File will be saved to location /home/shobi/Trajectory/Datasets/Toy4/toy4_animation.gif
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  0%|          | 0/27 [00:00&lt;?, ?it/s]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  0%|          | 0/27 [00:00&lt;?, ?it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>total number of stream lines 560
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<img alt="../_images/9bc251b1cd253ea2533398e086907afc204cb4e8177730f6fc5b73d323f6c20d.png" src="../_images/9bc251b1cd253ea2533398e086907afc204cb4e8177730f6fc5b73d323f6c20d.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">Image</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;/home/shobi/Trajectory/Datasets/Toy4/toy4_animation.gif&#39;</span><span class="p">,</span><span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">display</span><span class="p">(</span><span class="n">Image</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">read</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ff267974137f681df409ac2600b21c42b5c0b8d7ca9f1f8ea020da9790b647ea.png" src="../_images/ff267974137f681df409ac2600b21c42b5c0b8d7ca9f1f8ea020da9790b647ea.png" />
</div>
</div>
</section>
<section id="overall-trajectory">
<h2>Overall Trajectory<a class="headerlink" href="#overall-trajectory" title="Permalink to this heading"></a></h2>
<p>Draw high-level pseudotime and main trajectories by projecting clustergraph onto umap</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># draw overall pseudotime and main trajectories</span>
<span class="n">via</span><span class="o">.</span><span class="n">plot_trajectory_curves</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pyVIA/plotting_via.py:2641: RuntimeWarning: Couldn&#39;t reach some vertices at src/paths/unweighted.c:440
  path_orange = G_orange.get_shortest_paths(super_root, to=fst_i)[0]
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
100% (11 of 11) |########################| Elapsed Time: 0:00:00 Time:  0:00:00
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 14:26:49.451690	Super cluster 2 is a super terminal with sub_terminal cluster 11
2023-10-12 14:26:49.451787	Super cluster 8 is a super terminal with sub_terminal cluster 13
2023-10-12 14:26:49.451808	Super cluster 9 is a super terminal with sub_terminal cluster 2
2023-10-12 14:26:49.451825	Super cluster 10 is a super terminal with sub_terminal cluster 8
2023-10-12 14:26:49.451841	Super cluster 11 is a super terminal with sub_terminal cluster 9
2023-10-12 14:26:49.451858	Super cluster 13 is a super terminal with sub_terminal cluster 10
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&lt;Figure size 3000x1500 with 3 Axes&gt;,
 &lt;Axes: title={&#39;center&#39;: &#39;True Labels: ncomps:30. knn:10&#39;}&gt;,
 &lt;Axes: title={&#39;center&#39;: &#39;Pseudotime&#39;}&gt;)
</pre></div>
</div>
<img alt="../_images/74a2ae25ea7bf57d1be8554ae600e24ae2d3a3bb25cdbf6a47bd75c85fc49a2d.png" src="../_images/74a2ae25ea7bf57d1be8554ae600e24ae2d3a3bb25cdbf6a47bd75c85fc49a2d.png" />
</div>
</div>
</section>
<section id="lineage-probabilities">
<h2>Lineage Probabilities<a class="headerlink" href="#lineage-probabilities" title="Permalink to this heading"></a></h2>
<p>Plot the lineage probabilities for each lineage (associated with its terminal state, e.g. the leaf nodes)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">via</span><span class="o">.</span><span class="n">plot_sc_lineage_probability</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="n">v0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 14:27:39.714549	Marker_lineages: [11, 13, 2, 8, 9, 10]
2023-10-12 14:27:39.715041	Automatically setting embedding to via_object.embedding
2023-10-12 14:27:39.716301	The number of components in the original full graph is 2
2023-10-12 14:27:39.716333	For downstream visualization purposes we are also constructing a low knn-graph 
2023-10-12 14:27:39.889418	Check sc pb 1.0 
f getting majority comp
2023-10-12 14:27:39.915493	Cluster path on clustergraph starting from Root Cluster 7 to Terminal Cluster 11: [7, 0, 4, 11]
2023-10-12 14:27:39.915525	Cluster path on clustergraph starting from Root Cluster 7 to Terminal Cluster 13: [7, 0, 4, 13]
f getting majority comp
2023-10-12 14:27:39.916266	Cluster path on clustergraph starting from Root Cluster 3 to Terminal Cluster 2: [3, 14, 6, 9, 2]
2023-10-12 14:27:39.916290	Cluster path on clustergraph starting from Root Cluster 3 to Terminal Cluster 8: [3, 1, 8]
2023-10-12 14:27:39.916305	Cluster path on clustergraph starting from Root Cluster 3 to Terminal Cluster 9: [3, 14, 6, 9]
2023-10-12 14:27:39.916319	Cluster path on clustergraph starting from Root Cluster 3 to Terminal Cluster 10: [3, 1, 12, 5, 10]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pyVIA/plotting_via.py:1109: RuntimeWarning: Couldn&#39;t reach some vertices at src/paths/unweighted.c:440
  path_orange = G_orange.get_shortest_paths(via_object.root[ii], to=fst_i)[0]
/home/user/anaconda3/envs/Via2Env_py10/lib/python3.10/site-packages/pyVIA/plotting_via.py:1163: RuntimeWarning: Couldn&#39;t reach some vertices at src/paths/unweighted.c:440
  path = G.get_shortest_paths(root1_list[p1_cc[loc_ts_current]], to=labelsq1[0][0])  # weights=&#39;weight&#39;)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2023-10-12 14:27:40.055596	Revised Cluster level path on sc-knnGraph from Root Cluster 3 to Terminal Cluster 11 along path: [7, 7, 7, 0, 4, 11, 11, 11, 11, 11]
2023-10-12 14:27:40.095151	Revised Cluster level path on sc-knnGraph from Root Cluster 7 to Terminal Cluster 13 along path: []
2023-10-12 14:27:40.134602	Revised Cluster level path on sc-knnGraph from Root Cluster 3 to Terminal Cluster 2 along path: [3, 3, 3, 1, 12, 5, 10, 2, 2]
2023-10-12 14:27:40.170573	Revised Cluster level path on sc-knnGraph from Root Cluster 3 to Terminal Cluster 8 along path: [3, 3, 1, 8, 8, 8, 8]
2023-10-12 14:27:40.205726	Revised Cluster level path on sc-knnGraph from Root Cluster 3 to Terminal Cluster 9 along path: []
2023-10-12 14:27:40.247368	Revised Cluster level path on sc-knnGraph from Root Cluster 7 to Terminal Cluster 10 along path: [3, 3, 3, 1, 12, 5, 10, 10, 10, 10, 10]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&lt;Figure size 960x720 with 12 Axes&gt;,
 array([[&lt;Axes: title={&#39;center&#39;: &#39;Lineage: 11-T1_M2&#39;}&gt;,
         &lt;Axes: title={&#39;center&#39;: &#39;Lineage: 13-T1_M4&#39;}&gt;,
         &lt;Axes: title={&#39;center&#39;: &#39;Lineage: 2-T2_M2&#39;}&gt;],
        [&lt;Axes: title={&#39;center&#39;: &#39;Lineage: 8-T2_M6&#39;}&gt;,
         &lt;Axes: title={&#39;center&#39;: &#39;Lineage: 9-T2_M2&#39;}&gt;,
         &lt;Axes: title={&#39;center&#39;: &#39;Lineage: 10-T2_M4&#39;}&gt;]], dtype=object))
</pre></div>
</div>
<img alt="../_images/595817e14f0fffe56d036c5ebe441a797d421705b9a751ed27c066fa23d68ad9.png" src="../_images/595817e14f0fffe56d036c5ebe441a797d421705b9a751ed27c066fa23d68ad9.png" />
</div>
</div>
</section>
<section id="gene-dynamics">
<h2>Gene Dynamics<a class="headerlink" href="#gene-dynamics" title="Permalink to this heading"></a></h2>
<p>Plot the gene expression levels along each lineage for marker genes. Here we use dummy values for the Genes which are just the values of the initial PCs for illustrative purposes</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">df_subset</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_counts</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gene0&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene1&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene2&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene3&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene4&#39;</span><span class="p">])</span>
<span class="n">via</span><span class="o">.</span><span class="n">get_gene_expression</span><span class="p">(</span><span class="n">via0</span><span class="o">=</span><span class="n">v0</span><span class="p">,</span> <span class="n">gene_exp</span><span class="o">=</span><span class="n">df_subset</span><span class="p">,</span> <span class="n">marker_genes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gene0&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene1&#39;</span><span class="p">],</span> <span class="n">fontsize_</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span> <span class="c1"># since we dont want to plot all the genes in this dataframe, we select the two we wish to see trends for by setting marker_genes = [&#39;Gene0&#39;, &#39;Gene1&#39;]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/aaa4a243ced46f376756762fdded08c997e65cca957191456adfc79986919ec6.png" src="../_images/aaa4a243ced46f376756762fdded08c997e65cca957191456adfc79986919ec6.png" />
</div>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="ViaJupyter_scRNA_Hematopoiesis.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">3. Via Human Hematopoiesis</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="ViaJupyter_Toy_Multifurcating.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">1. Basic workflow</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2022, shobana stassen
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">2. Disconnected Trajectories</a><ul>
<li><a class="reference internal" href="#load-data">Load data</a></li>
<li><a class="reference internal" href="#initialize-and-run-via">Initialize and run VIA</a></li>
<li><a class="reference internal" href="#draw-the-viagraph">Draw the viagraph</a></li>
<li><a class="reference internal" href="#initial-state">Initial state</a><ul>
<li><a class="reference internal" href="#roots-can-also-be-provided-as-cell-indices">Roots can also be provided as cell indices</a></li>
</ul>
</li>
<li><a class="reference internal" href="#plot-atlas-view">Plot Atlas View</a></li>
<li><a class="reference internal" href="#fine-grained-vector-field">Fine-grained vector field</a></li>
<li><a class="reference internal" href="#overall-trajectory">Overall Trajectory</a></li>
<li><a class="reference internal" href="#lineage-probabilities">Lineage Probabilities</a></li>
<li><a class="reference internal" href="#gene-dynamics">Gene Dynamics</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/scripts/furo.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script defer="defer" src="https://unpkg.com/@popperjs/core@2"></script>
    <script defer="defer" src="https://unpkg.com/tippy.js@6"></script>
    <script defer="defer" src="../_static/tippy\notebooks\ViaJupyter_Toy_Disconnected.f8075d6f-460b-4770-ad6c-ca33db156244.js"></script>
    </body>
</html>