selector_to_html = {"a[href=\"#vector-field\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Vector field<a class=\"headerlink\" href=\"#vector-field\" title=\"Permalink to this heading\">\uf0c1</a></h3><p>The vector field of trajectories is projected onto the embedding</p>", "a[href=\"#fast-visualization-using-underlying-single-cell-via-graphs\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Fast visualization using underlying single-cell VIA-graphs<a class=\"headerlink\" href=\"#fast-visualization-using-underlying-single-cell-via-graphs\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>For more control after running VIA, simply call the <code class=\"docutils literal notranslate\"><span class=\"pre\">via_mds()</span></code> function. When using time-series data consider using via-mds and via-umap automatically for improved visualizations that are well integrated with the Trajectory.</p>", "a[href=\"#via-graph\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">VIA graph<a class=\"headerlink\" href=\"#via-graph\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>To visualize the results of the Trajectory inference in various ways. Via offers various plotting functions.We first show the cluster-graph level trajectory abstraction consisting of two subplots colored by annotated (true_label) composition and by pseudotime</p>", "a[href=\"#parameters\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Parameters<a class=\"headerlink\" href=\"#parameters\" title=\"Permalink to this heading\">\uf0c1</a></h2>", "a[href=\"#via-mds-parameters\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">VIA-mds Parameters<a class=\"headerlink\" href=\"#via-mds-parameters\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#overall-trajectory\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Overall trajectory<a class=\"headerlink\" href=\"#overall-trajectory\" title=\"Permalink to this heading\">\uf0c1</a></h3><p>Draw overall pseudotime and main trajectories on single cell embedding</p>", "a[href=\"#lineage-pathways-edge-views\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Lineage Pathways - edge views<a class=\"headerlink\" href=\"#lineage-pathways-edge-views\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#gene-dynamics\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Gene Dynamics<a class=\"headerlink\" href=\"#gene-dynamics\" title=\"Permalink to this heading\">\uf0c1</a></h3><p>Plot the gene expression levels along each lineage for marker genes. Here we use dummy values for the Genes which are just the values of the initial PCs for illustrative purposes</p>", "a[href=\"#lineage-probabilities\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Lineage probabilities<a class=\"headerlink\" href=\"#lineage-probabilities\" title=\"Permalink to this heading\">\uf0c1</a></h3><p>Plot the lineage probabilities for each lineage (associated with a terminal state)</p>", "a[href=\"#pre-defined-terminal-states-and-integrated-visualization\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Pre-defined terminal states and Integrated Visualization<a class=\"headerlink\" href=\"#pre-defined-terminal-states-and-integrated-visualization\" title=\"Permalink to this heading\">\uf0c1</a></h2>", "a[href=\"#basic-workflow\"]": "<h1 class=\"tippy-header\" style=\"margin-top: 0;\">1. Basic workflow<a class=\"headerlink\" href=\"#basic-workflow\" title=\"Permalink to this heading\">\uf0c1</a></h1><p>This vignette shows the basic workflow of Via and how to access the different visualizations after running the computations. The dataset is a toy multifurcation.\nWe start by importing modules.</p>", "a[href=\"#colored-by-true-label\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">colored by true_label<a class=\"headerlink\" href=\"#colored-by-true-label\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#expected-outputs\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Expected Outputs<a class=\"headerlink\" href=\"#expected-outputs\" title=\"Permalink to this heading\">\uf0c1</a></h2>", "a[href=\"#via-umap-based-on-underlying-via-single-cell-knn-graphs\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">VIA-umap based on underlying via single cell knn graphs<a class=\"headerlink\" href=\"#via-umap-based-on-underlying-via-single-cell-knn-graphs\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>This is encouraged especially when VIA is run with timeseries data <code class=\"docutils literal notranslate\"><span class=\"pre\">time_series_labels</span> <span class=\"pre\">=</span> <span class=\"pre\">[list</span> <span class=\"pre\">of</span> <span class=\"pre\">numeric</span> <span class=\"pre\">values]</span></code></p>", "a[href=\"#scatter-plots\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Scatter Plots<a class=\"headerlink\" href=\"#scatter-plots\" title=\"Permalink to this heading\">\uf0c1</a></h2><h3>colored by true_label<a class=\"headerlink\" href=\"#colored-by-true-label\" title=\"Permalink to this heading\">\uf0c1</a></h3>"}
skip_classes = ["headerlink", "sd-stretched-link"]

window.onload = function () {
    for (const [select, tip_html] of Object.entries(selector_to_html)) {
        const links = document.querySelectorAll(` ${select}`);
        for (const link of links) {
            if (skip_classes.some(c => link.classList.contains(c))) {
                continue;
            }

            tippy(link, {
                content: tip_html,
                allowHTML: true,
                arrow: true,
                placement: 'auto-start', maxWidth: 500, interactive: false,

            });
        };
    };
    console.log("tippy tips loaded!");
};
