selector_to_html = {"a[href=\"#load-the-data\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Load the data<a class=\"headerlink\" href=\"#load-the-data\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>Data has been filtered and library normalized. PCA has been computed and stored in the anndata object. PCA was performed on all genes remaining after filtering\nData is annotated with time/stage, cluster labels, and cell type labels (coarse and fine).\nThe anndata object is a large file due to the velocity matrices (18Gb) and can be downloaded <span class=\"xref myst\">here</span>.</p>", "a[href=\"#plot-via2-0-cluster-graph\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Plot Via2.0 cluster graph<a class=\"headerlink\" href=\"#plot-via2-0-cluster-graph\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>Examples showing how to control the colormaps, labeling and choice of attribute to plot on the viagraph</p>", "a[href=\"#take-a-look-at-the-data-using-the-stored-atlas-single-cell-embedding\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Take a look at the data using the stored Atlas single-cell embedding<a class=\"headerlink\" href=\"#take-a-look-at-the-data-using-the-stored-atlas-single-cell-embedding\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>We show the Atlas View (precomputed and saved here for convenience) colored by tissue type and gastrulation stage.\nNote that we have change the gastrulation time-stamps to numerical values. E.g. E6.5 becomes 650, E7.25 becoems 750 etc.\nContinue reading to see how the Atlas View is generated.</p>", "a[href=\"#compute-atlas-single-cell-embedding\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Compute Atlas single-cell embedding<a class=\"headerlink\" href=\"#compute-atlas-single-cell-embedding\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>We transfer the properties of the trajectory infused viagraph and the single-cell temporally adjusted graph to generate an embedding that visually conveys the structure of the trajectory and can be used as the basis of the Atlas View (described farther down)</p><p>If you want to plot the pre-computed atlas embedding set compute_atlas_embedding = False, otherwise compute the atlas embedding.\nWe can then plot the atlas view based on cell type or other attributes.</p>", "a[href=\"#plot-gene-trends\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Plot gene trends<a class=\"headerlink\" href=\"#plot-gene-trends\" title=\"Permalink to this heading\">\uf0c1</a></h2><p><em><strong>Memory</strong></em></p>", "a[href=\"#the-atlas-view\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">The Atlas View<a class=\"headerlink\" href=\"#the-atlas-view\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>Now we look at how to construct the atlas view which combines the single-cell spatial layout with a high-resoulution view of the edge connectivity (potential pathways).</p>", "a[href=\"#set-parameters-for-via-2-0\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Set parameters for Via 2.0<a class=\"headerlink\" href=\"#set-parameters-for-via-2-0\" title=\"Permalink to this heading\">\uf0c1</a></h2><p><em><strong>A quick note on how to (auto) set the root</strong></em></p>", "a[href=\"#initialize-and-run-the-via2-0\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Initialize and Run the Via2.0<a class=\"headerlink\" href=\"#initialize-and-run-the-via2-0\" title=\"Permalink to this heading\">\uf0c1</a></h2><p><strong>Optional: Predefined cluster labels and terminal cell fates</strong>\nBy default, Via will compute its own clusters and terminal cell fates. However, in some cases it may be desirable to provide pre-computed cluster groupings or set terminal cell fates. This can also be helfpul for reproducibility.\nIn the example below, we set the clusters to those that we have computed on a prior run of Via. <code class=\"docutils literal notranslate\"><span class=\"pre\">labels</span> <span class=\"pre\">=</span> <span class=\"pre\">[list</span> <span class=\"pre\">of</span> <span class=\"pre\">numeric</span> <span class=\"pre\">labels]</span></code> of length n_cells.\nWe also set the cell fates to those precomputed by Via. <code class=\"docutils literal notranslate\"><span class=\"pre\">user_defined_terminal_group</span> <span class=\"pre\">=</span> <span class=\"pre\">[list</span> <span class=\"pre\">of</span> <span class=\"pre\">cell</span> <span class=\"pre\">fates</span> <span class=\"pre\">that</span> <span class=\"pre\">corresponding</span> <span class=\"pre\">to</span> <span class=\"pre\">items</span> <span class=\"pre\">in</span> <span class=\"pre\">true_label]</span></code> This could also be provided as <code class=\"docutils literal notranslate\"><span class=\"pre\">user_defined_terminal_cell=[list</span> <span class=\"pre\">of</span> <span class=\"pre\">cell</span> <span class=\"pre\">indices</span> <span class=\"pre\">corresponding</span> <span class=\"pre\">to</span> <span class=\"pre\">cell</span> <span class=\"pre\">fates]</span></code></p>", "a[href=\"#plot-single-cell-lineage-probabilities\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Plot single-cell lineage probabilities<a class=\"headerlink\" href=\"#plot-single-cell-lineage-probabilities\" title=\"Permalink to this heading\">\uf0c1</a></h2><p><em><strong>Memory</strong></em></p>", "a[href=\"#plot-stream-plot-using-hybrid-pseudotime-rnavelocity-based-directionality\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Plot stream plot using hybrid pseudotime-RNAvelocity based directionality<a class=\"headerlink\" href=\"#plot-stream-plot-using-hybrid-pseudotime-rnavelocity-based-directionality\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>Using a combination of pseudotime and rna-velocity ensures biologically meaningful directionality</p>", "a[href=\"#via-2-0-cartography-on-mouse-gastrulation\"]": "<h1 class=\"tippy-header\" style=\"margin-top: 0;\">1. Via 2.0 Cartography on Mouse Gastrulation<a class=\"headerlink\" href=\"#via-2-0-cartography-on-mouse-gastrulation\" title=\"Permalink to this heading\">\uf0c1</a></h1><p>Time-series RNA-seq data with RNA velocity of E6.5-E8.5 murine gastrulation (Sala 2019).\nSince we have the time-series labels of the different stages (taken at quarter day intervals), we will show how to use these to adjust and guide the Cartography, both visually and for trajectory prediction.</p>"}
skip_classes = ["headerlink", "sd-stretched-link"]

window.onload = function () {
    for (const [select, tip_html] of Object.entries(selector_to_html)) {
        const links = document.querySelectorAll(` ${select}`);
        for (const link of links) {
            if (skip_classes.some(c => link.classList.contains(c))) {
                continue;
            }

            tippy(link, {
                content: tip_html,
                allowHTML: true,
                arrow: true,
                placement: 'auto-start', maxWidth: 500, interactive: false,

            });
        };
    };
    console.log("tippy tips loaded!");
};
