selector_to_html = {"a[href=\"#id1\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Key Parameters:<a class=\"headerlink\" href=\"#id1\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#various-visualizations\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Various Visualizations<a class=\"headerlink\" href=\"#various-visualizations\" title=\"Permalink to this heading\">\uf0c1</a></h2><h3>Visualizations of the trajectory can be plotted at various edge resolutions<a class=\"headerlink\" href=\"#visualizations-of-the-trajectory-can-be-plotted-at-various-edge-resolutions\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#user-s-choice-of-embedding\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">User\u2019s choice of embedding<a class=\"headerlink\" href=\"#user-s-choice-of-embedding\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>Three embedding exmaples: TSNE, UMAP, and PHATE\nAlternatively allow via.VIA() to compute an embedding using the underlying graph by setting: do_embedding = True AND embedding_type = \u2018via-umap\u2019 OR \u2018via-mds\u2019</p>", "a[href=\"#gene-dynamics\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Gene Dynamics<a class=\"headerlink\" href=\"#gene-dynamics\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>The gene dynamics along pseudotime for all detected lineages are automatically inferred by VIA. These can be interpreted as the change in gene expression along any given lineage.</p>", "a[href=\"#id2\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Key Parameters:<a class=\"headerlink\" href=\"#id2\" title=\"Permalink to this heading\">\uf0c1</a></h3><p>marker_lineages (list) of terminal clusters</p>", "a[href=\"#key-parameters\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Key Parameters:<a class=\"headerlink\" href=\"#key-parameters\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#id5\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Key Parameters<a class=\"headerlink\" href=\"#id5\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#id4\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Gene Dynamics<a class=\"headerlink\" href=\"#id4\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>Heatmaps of genes along pseudotimetime</p>", "a[href=\"#animated-stream-plot\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Animated Stream plot<a class=\"headerlink\" href=\"#animated-stream-plot\" title=\"Permalink to this heading\">\uf0c1</a></h2>", "a[href=\"#cluster-level-trajectory-plots\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Cluster Level Trajectory Plots<a class=\"headerlink\" href=\"#cluster-level-trajectory-plots\" title=\"Permalink to this heading\">\uf0c1</a></h2>", "a[href=\"#probabilistic-pathways-and-memory\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Probabilistic pathways and Memory<a class=\"headerlink\" href=\"#probabilistic-pathways-and-memory\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>Visualize the probabilistic pathways from root to terminal state as indicated by the lineage likelihood. The higher the linage likelihood, the greater the potential of that particular cell to differentiate towards the terminal state of interest.\nChanging the memory paramater will alter the specificity of the lineage pathway.\nThis can be visualized at the single-cell level but also combined with the Atlas View which visualizes cell-cell connectivity and pathways</p>", "a[href=\"#edge-bundle-plot\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Edge bundle plot<a class=\"headerlink\" href=\"#edge-bundle-plot\" title=\"Permalink to this heading\">\uf0c1</a></h2><h3>Key Parameters (default):<a class=\"headerlink\" href=\"#key-parameters-default\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#via-human-hematopoiesis\"]": "<h1 class=\"tippy-header\" style=\"margin-top: 0;\">3. Via Human Hematopoiesis<a class=\"headerlink\" href=\"#via-human-hematopoiesis\" title=\"Permalink to this heading\">\uf0c1</a></h1><p>This notebook uses VIA to interpret the CD34 Hematopoiesis dataset (Setty et al.,2019)</p>", "a[href=\"#visualizations-of-the-trajectory-can-be-plotted-at-various-edge-resolutions\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Visualizations of the trajectory can be plotted at various edge resolutions<a class=\"headerlink\" href=\"#visualizations-of-the-trajectory-can-be-plotted-at-various-edge-resolutions\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#key-parameters-default\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Key Parameters (default):<a class=\"headerlink\" href=\"#key-parameters-default\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#animated-edge-bundle-plot\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Animated edge bundle plot<a class=\"headerlink\" href=\"#animated-edge-bundle-plot\" title=\"Permalink to this heading\">\uf0c1</a></h2>", "a[href=\"#run-via\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Run VIA<a class=\"headerlink\" href=\"#run-via\" title=\"Permalink to this heading\">\uf0c1</a></h2><h3>Key Parameters:<a class=\"headerlink\" href=\"#key-parameters\" title=\"Permalink to this heading\">\uf0c1</a></h3>", "a[href=\"#load-data-from-directory\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Load data from directory<a class=\"headerlink\" href=\"#load-data-from-directory\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>We use annotations given by SingleR which uses Novershtern Hematopoietic cell data as the reference. These are in line with the annotations given by Setty et al., 2019 but offer a slightly more granular annotation.</p>", "a[href=\"#trajectory-projection\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Trajectory projection<a class=\"headerlink\" href=\"#trajectory-projection\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>Visualize the overall VIA trajectory projected onto a 2D embedding (UMAP, Phate, TSNE etc) in different ways.</p>", "a[href=\"#visualise-gene-feature-graph\"]": "<h2 class=\"tippy-header\" style=\"margin-top: 0;\">Visualise gene/feature graph<a class=\"headerlink\" href=\"#visualise-gene-feature-graph\" title=\"Permalink to this heading\">\uf0c1</a></h2><p>View the gene expression along the VIA graph. We use the computed HNSW small world graph in VIA to accelerate the gene imputation calculations (using similar approach to MAGIC) as follows:</p>", "a[href=\"#id3\"]": "<h3 class=\"tippy-header\" style=\"margin-top: 0;\">Key Parameters<a class=\"headerlink\" href=\"#id3\" title=\"Permalink to this heading\">\uf0c1</a></h3>"}
skip_classes = ["headerlink", "sd-stretched-link"]

window.onload = function () {
    for (const [select, tip_html] of Object.entries(selector_to_html)) {
        const links = document.querySelectorAll(` ${select}`);
        for (const link of links) {
            if (skip_classes.some(c => link.classList.contains(c))) {
                continue;
            }

            tippy(link, {
                content: tip_html,
                allowHTML: true,
                arrow: true,
                placement: 'auto-start', maxWidth: 500, interactive: false,

            });
        };
    };
    console.log("tippy tips loaded!");
};
